---
title: "KCinAD"
author: "Felicitas Pardow"
date: "2023-03-14"
output: html_document
---

```{r setup, include = FALSE}
#loading packages
# library(SingleCellExperiment)
# library(Seurat)
# library(ggplot2)
# library(dplyr)
# library(openxlsx)
# library(clusterProfiler)
# library(org.Hs.eg.db)
# library(ggvenn)
# library(RColorBrewer)
# library(progeny)
# library(pheatmap)
# library(SeuratObject)
# library(rBayesianOptimization)
# library(ggbeeswarm)
# library(slingshot)
# library(SummarizedExperiment)
# library(circlize)
# library(rmarkdown)
# library(tradeSeq)
# library(DelayedMatrixStats)
# library(ComplexHeatmap)
# library(SeuratDisk)
# library(gt)

set.seed(1234)

#setting directories
work.dir <- "C:/Users/spare/OneDrive - Radboudumc/Results/KCsinAD/FPcomp/KCinAD/" #working directory
data.dir <-  paste0(work.dir, "data/") # specify data location
save.dir <-  paste0(work.dir, "results_ownFilters/") # specify saving location
svm.data <- paste0(work.dir, "SVMannotation_ownFilters_ownAnnotation/Data/") #location to save svm data
svm.results <- paste0(work.dir, "SVMannotation_ownFilters_ownAnnotation/Results/")  #location to load svm results

#colors
colors_samples_state <- c("AD1" = "#FF6347", "AD2" = "#FF6347", "AD3" = "#FF6347", "AD4" = "#FF6347", "AD5" = "#FF6347", "AD6" = "#FF6347", "AD7" = "#FF6347", "AD8" = "#FF6347", "HC1" = "#228B22", "HC2" = "#228B22", "HC3" = "#228B22", "HC4" = "#228B22", "HC5" = "#228B22", "HC6" = "#228B22", "HC7" = "#228B22") 

colors_samples <- c("HC1" = "#228B22", "HC2" = "#228B22", "HC3" = "#228B22", "HC4" = "#228B22", "HC5" = "#228B22", "HC6" = "#F8766D", "HC7" = "#B79F00", "AD1" = "#FF6347", "AD2" = "#FF6347", "AD3" = "#FF6347", "AD4" = "#FF6347", "AD5" = "#00BA38", "AD6" = "#00BFC4", "AD7" = "#619CFF", "AD8" = "#F564E3") 

colors_state <- c("HC" = "#228B22", "AD" = "#FF6347") 
colors_cellcycle <- c("G1" = "darkorange", "S" = "#66C2A5", "G2M" = "#FFD92F")
colors_subtypes <- c("basal I" = "#FFD92F", "basal II" = "#66C2A5", "basal III" = "#8DA0CB", "basal IV" = "#A6D854", "stratum spinosum" = "#FC8D62", "stratum granulosum" = "#E78AC3")

#functions
seuratpipeline <- function(object = object, nfeat = 2000, npcs = 50, dims = 10, res = 0.1){ 
  
  #standard seurat pipeline with defaults, without normalizing
  object <- FindVariableFeatures(object, nfeatures = nfeat) #default: n = 2000, selection.method = vst
  object <- ScaleData(object)
  object <- RunPCA(object, features = VariableFeatures(object = object), npcs = npcs) #default: npcs = 50
  print(ElbowPlot(object) + ggtitle("Elbowplot of top PCAs"))
  object <- FindNeighbors(object, dims = 1:dims) #default: dims = 10
  object <- FindClusters(object, resolution = res) #default: resolution = 0.1
  object <- RunUMAP(object, dims = 1:dims) #default: dims = 10
    
  return(object)
}

percentage_df = function(seurat_object, cell_annotation = "seurat_clusters", variable = "orig.ident", facet = F){
  #variable = which columns
  #Change ID
  seurat_object = seurat_object@meta.data
  df_percentages = ""

  #list of variables
  my_vars = as.character(as.data.frame(table(seurat_object[[variable]]))[ , 1])
  count = 0
  #loop through each variable, add to dataframe
  for(var_x in my_vars){
    #Subset dataframe for specific variable
    seurat_object_tmp = seurat_object[seurat_object[[variable]] == var_x, ]
    
    percent_total = as.data.frame(table(seurat_object_tmp[[cell_annotation]])) #add cell count
    percent_total$Var1 = as.character(percent_total$Var1) #change type factor -> character
    percent_total["percentage"] = percent_total$Freq / sum(percent_total$Freq) * 100 #add percentage
    percent_total["variable"] = var_x #add variable
    if(count == 0){
    df_percentages = percent_total #create new df for first
    } else{
    df_percentages[(nrow(df_percentages) + 1):(nrow(df_percentages) + nrow(percent_total)), ] = percent_total
    }
  count = count + 1

  }
  
  #formatting
  df_percentages$percentage = signif(df_percentages$percentage, 3)
  colnames(df_percentages) = c("Celltype", "Freq" , "percentage", "Variable")
  #readable labels
  df_percentages = df_percentages[df_percentages$Freq != 0, ]
  
  if(facet){
    df_percentages$con = df_percentages$Variable
    list_m_a = strsplit(df_percentages$Variable, "_")
    for(p in 1:length(list_m_a)){
      var = paste0(list_m_a[[p]][1], list_m_a[[p]][2])
      df_percentages$con[p] = var
    }
    df_percentages$sample = df_percentages$Variable
  }
  df_percentages$ann = df_percentages$percentage
  df_percentages$ann[df_percentages$ann < 1] = "" #remove labels of small fractions
  return(df_percentages)
}

progeny_heatmap = function(object, nfeatures = 100, cell_order = x, clustering = F, pathway_order = c("JAK-STAT", "TNFa", "NFkB", "VEGF",  "TGFb", "Androgen", "PI3K", "EGFR", "MAPK", "Estrogen", "p53",  "WNT", "Hypoxia", "Trail")){
  
  require(progeny)
  require(pheatmap)
  require(tidyr)
  require(tibble)
  
  #data frame delineating cell annotations
  cell_annotation <- data.frame(cell = names(Idents(object)),
                                annotation = as.character(Idents(object)), stringsAsFactors = FALSE)
  
  #calculation progeny, added as an assay to the seurat object 
  object <- progeny(object, scale = F, top = nfeatures, perm = 1, return_assay = TRUE)
  # object@assays$progeny
  
  #We can now directly apply Seurat functions in our Progeny scores, like scaling the pathway activity scores
  object <- Seurat::ScaleData(object, assay = "progeny")
  
  #We transform Progeny scores into a data frame to better handling the results
  progeny_scores <- as.data.frame(t(object@assays$progeny$scale.data)) %>%
    tibble::rownames_to_column("cell") %>%
    tidyr::pivot_longer(names_to = "pathway", values_to = "activity", c(-"cell")) #transforms the table into a longer able format - assigning a separate row for every pathways + coordinating activity per cell
  
  #add information of cell annotation
  progeny_scores <- inner_join(progeny_scores, cell_annotation, by = c("cell"))
  
  #summarize progeny scores by population
  progeny_summarized <- progeny_scores %>%
    group_by(pathway, annotation) %>%
    summarise(average = mean(activity), std = sd(activity))
  
  #prepare dataframe for plotting
  progeny_summarized <- progeny_summarized %>% 
    dplyr::select(-std) %>%
    tidyr::spread(pathway, average) %>%
    data.frame(row.names = "annotation", check.names = FALSE, stringsAsFactors = FALSE)
  
  progeny_summarized <- progeny_summarized[cell_order, ]
  progeny_summarized <- progeny_summarized[ , pathway_order]
  
  #we prepare the plot
  paletteLength <- 100
  myColor <-  colorRampPalette(c("darkblue", "white", "indianred"))(paletteLength)
  
  progenyBreaks <- c(seq(min(progeny_summarized), 0, length.out = ceiling(paletteLength / 2) + 1),
                    seq(max(progeny_summarized) / paletteLength, max(progeny_summarized), length.out = floor(paletteLength / 2)))
  
  #we plot
  progeny_hmap <- pheatmap::pheatmap(t(progeny_summarized),
                                     cluster_cols = F, cluster_rows = clustering, row_order = cell_order,
                                     color = myColor, breaks = progenyBreaks, border_color = NA,
                                     angle_col = c("45")) + scale_x_discrete(labels = scales::label_wrap(10))
  
  return(progeny_hmap)
}

#graphics settings
options(ggrepel.max.overlaps = Inf)
```

# Setting up the workstream

This project was carried out using R version 4.4.1. To recreate the exact code, please install this R version. You can use the renv package to "lazy-load" my version of the installed packages. Alternatively, you can manually install the packages yourself. 

(Of note: the package SeuratDisk cannot be installed from github so i manually downloaded and installed it. I put my version in the /cellar of the renv package where it gets automatically installed during the init() command.)

```{r install packages, eval = F}
#install renv for package management and reproducibility
if (!requireNamespace("renv", quietly = TRUE)){ #installs package renv package if its not installed already
    install.packages("renv")
}

library(renv)
renv::init() #takes a while, press YES and later 1

#install the BiocManager version 3.19
# install.packages("remotes") 
# remotes::install_version("BiocManager", version = "3.19") 
#
#now use the exact same version of BiocManager to replicate my installed package versions 
# BiocManager::install("SeuratObject")
# BiocManager::install("Seurat")
# BiocManager::install("tidyverse")
# BiocManager::install("SingleCellExperiment") #update none
# BiocManager::install("clusterProfiler")
# BiocManager::install("openxlsx")
# BiocManager::install("org.Hs.eg.db")
# BiocManager::install("RColorBrewer")
# BiocManager::install("slingshot")
# BiocManager::install("ggvenn")
# BiocManager::install("limma")
# BiocManager::install("progeny")
# BiocManager::install("pheatmap")
# BiocManager::install("decoupleR")
# BiocManager::install("OmnipathR")
# BiocManager::install("ggpubr")
# BiocManager::install("tradeSeq")
# BiocManager::install("pathview")
# BiocManager::install("DelayedMatrixStats")
# BiocManager::install("rBayesianOptimization")
# BiocManager::install("ComplexHeatmap")
# 
# installing seurat-disk did not work from github so i downloaded the package via the windows prompt (git clone https://github.com/mojaveazure/seurat-disk) and installed it from source
# install.packages("seurat-disk", repos = NULL, type = "source") #make sure to specify the right path
# 
# BiocManager::install("gt")
# BiocManager::install("cytobox")
# BiocManager::install("GEOquery")
#
#you can choose to save a snapshot of your libraries using 
renv::snapshot() # initialize project environment
```

Here you can quick load important files during pre-processing and analysis. 

```{r quick load, eval = F}
wang <- readRDS(paste0(save.dir, "wang.rds"))
wang_filtered <- readRDS(paste0(save.dir, "wang_filtered.rds"))
wang_filtered.markers <- readRDS(paste0(save.dir, "wang_filtered.markers.rds")) 
wang_kcsubset <- readRDS(paste0(save.dir, "wang_kcsubset.rds"))
wang_kcfiltered <- readRDS(paste0(save.dir, "wang_kcfiltered.rds"))
wang_kc.markersAnnotation <- readRDS(paste0(save.dir, "wang_kc.markersAnnotation.rds"))
wang_kc.markersCluster <- readRDS(paste0(save.dir, "wang_kc.markersCluster.rds"))
wang_kc <- readRDS(paste0(save.dir, "wang_kc.rds"))

rojahn <- readRDS(paste0(save.dir, "rojahn.rds"))
rojahn_biopsy <- readRDS(paste0(save.dir, "rojahn_biopsy.rds"))
rojahn_filtered <- readRDS(paste0(save.dir, "rojahn_filtered.rds"))
rojahn_kcsubset <- readRDS(paste0(save.dir, "rojahn_kcsubset.rds"))
rojahn_kcfiltered <- readRDS(paste0(save.dir, "rojahn_kcfiltered.rds"))
rojahn_kcfiltered2 <- readRDS(paste0(save.dir, "rojahn_kcfiltered2.rds"))
rojahn_kcfiltered3 <- readRDS(paste0(save.dir, "rojahn_kcfiltered3.rds"))
rojahn_kc <- readRDS(paste0(save.dir, "rojahn_kc.rds"))
```

# Processing GSE147482 / Wang et al. 2019 dataset 
 
We first load the Wang et al. reference dataset which we use to train the SVM annotation model. In the next lines, we will load the complete dataset, filter for quality, as well as isolate and annotate the keratinocytes. 
 
## Loading 

please download the GSE147482_RAW.tar data file from https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE147482 into the data.dir. 

We start with reading all samples files and combine them into one seurat object.

For each sample, we have 
- a count slot, showing the counts of genes for all cells in the sample
- information about the percentage of mitochondrial genes / sample (we remove those genes from the count slot)
- metadata of the sample

```{r load and combine samples}
#obtain the sample metadata from GEO
geo <- GEOquery::getGEO("GSE147482")
info_geo <- pData(phenoData(geo[[1]]))[ , c(2, 46, 8)]
donors <- gsub("Human infant foreskin epidermis - ", "", pData(phenoData(geo[[1]]))[ , c(1)])
donors <- gsub(" ", "", donors)
rownames(info_geo) <- donors

#figure out to load without download, see code above
data10x <- scCustomize::Read10X_GEO(data_dir = paste0(data.dir, "GSE147482_RAW"))

#remove GSM numbers from the sample names
names(data10x) <- gsub("(GSM443190[0-9]_)|(_ForeskinEpidermis_10XGenomics_)", "", names(data10x)) # remove multiple regex

#create a list that contains the seurat object of all samples
seuratlist <- list()

for (i in names(data10x)){
  print(i)
  
  #prepare the metadata in the dataframe info
  info <- data.frame(row.names = colnames(as.data.frame(data10x[[i]]))) 
  info$sample <- i
  info$geo_accession <- info_geo[i, "geo_accession"]
  info$publication <- c("Wang 2019")
  info$age <- gsub("Infant / ", "", info_geo[i, "age:ch1"]) #adjust wording from age:ch1 column
  info$tissue <- info_geo[i, "source_name_ch1"]
  
  #create the seurat object with metadata + count table
  seurat <- CreateSeuratObject(counts = data10x[[i]],
                             meta.data = info,
                             min.cells = 3,
                             min.features = 100,
                             project = c("GSE147482"))
  seurat$orig.ident <- c("GSE147482")
  
  #add to the list
  seuratlist[[i]] <- seurat
}

#the cell names contain sample numbers -> remove sample number from the cell name
for (i in c(1:length(seuratlist))){
  print (i)
  colnames(seuratlist[[i]]) <- gsub(pattern = "^[0-9]+_", replacement = "", colnames(seuratlist[[i]]))
}
# tail(colnames(seuratlist[[5]])) #test if it worked

#merge individual seurat objects / samples into one 
wang <- merge(
  x = seuratlist[[1]], y = c(seuratlist[c(2:length(seuratlist))]), #seurat objects to be merged
  add.cell.ids = names(seuratlist[1:length(seuratlist)]), #makes cell names unique (required) by adding sample name
  project = "GSE147482"
)

#remove layers
# Layers(wang[['RNA']])
wang[['RNA']] <- JoinLayers(object = wang[['RNA']])
# wang[['RNA']] <- split(x = wang[['RNA']], f = wang$sample) #if you want to split them again, now using sample names as layer names
```

## Quality control

```{r find and remove mitochondrial genes}
#check content content of mito genes per cell
mito_genes_index <- grep(pattern = "^MT-", x = rownames(wang)) #find all mito genes (13 genes)
percent_mito_genes <- colSums(wang[["RNA"]]$counts[mito_genes_index, ]) /
                      colSums(wang[["RNA"]]$counts) * 100
# print(summary(percent_mito_genes))

#add mitochondrial gene percentage to the meta data 
wang$percent.mt <- percent_mito_genes

#remove mitochondrial genes 
wang <- wang[-c(mito_genes_index), ] 
```

Gene QC: When loading the count tables into the seurat object, we exclude genes detected in less than 3 cells

Cell QC: 
nCounts / library size: total sum of counts across all genes for each cell (total number of molecules / cell)
nFeature_RNA: total number of genes detected in each cell

```{r quality control filtering}
# wang <- readRDS(paste0(save.dir, "wang.rds"))
# wang_filtered <- readRDS(paste0(save.dir, "wang_filtered.rds"))

wang <- NormalizeData(wang) #performs log normalization

#visualization
Idents(wang) <- wang$sample

#Supplemental Figure 1a
p1 <- VlnPlot(wang, features = c("nCount_RNA"), flip = T, pt.size = 0, cols = c(rep("#919675", 5))) +
  geom_hline(yintercept = 45000) + 
  labs(x = element_blank(), y = c("total counts per cell"), title = element_blank()) + 
  theme(legend.position = "none")

p2 <- VlnPlot(wang, features = c("nFeature_RNA"), flip = T, pt.size = 0, cols = c(rep("#919675", 5))) +
  geom_hline(yintercept = 600) +  geom_hline(yintercept = 5500) + 
  labs(x = element_blank(), y = c("genes per cell"), title = element_blank()) + 
  theme(legend.position = "none")

p3 <- VlnPlot(wang, features = c("percent.mt"), flip = T, pt.size = 0, cols = c(rep("#919675", 5))) +
  geom_hline(yintercept = 1) + geom_hline(yintercept = 10) +
  scale_y_continuous(labels = scales::label_dollar(suffix = "%", prefix = "")) +
  labs(x = element_blank(), y = c("mitochondrial genes per cell"), title = element_blank()) + 
  theme(legend.position = "none")

patchwork::wrap_plots(p1, p2, p3) 

#quality control filtering
wang_filtered <- subset(wang, subset = nCount_RNA < 45000
                        & nFeature_RNA > 600 & nFeature_RNA < 5500 
                        & percent.mt > 1 & percent.mt < 10)

VlnPlot(wang_filtered, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"), pt.size = 0)

# saveRDS(wang, paste0(save.dir, "wang.rds"))
```

Sample 3 has a higher sequencing depth whereas sample 2 (in comparison) has more mitochondrial reads. We apply relatively loose filtering for the complete biopsy dataset with the goal to apply more stringent filters after isolating the keratinocytes. 

## Full biopsy cell clustering and annotation

We now cluster the samples with the goal to identify and isolate the keratinocytes. 

```{r cell clustering}
# wang_filtered <- readRDS(paste0(save.dir, "wang_filtered.rds"))

#clustering
wang_filtered <- seuratpipeline(wang_filtered, nfeat = 5000, res = 0.05) 

#visualize clustering
DefaultAssay(wang_filtered) <- "RNA"
DimPlot(wang_filtered, reduction = "umap")
DimPlot(wang_filtered, reduction = "umap", group.by = "sample")

FeaturePlot(wang_filtered, features = "nCount_RNA")
FeaturePlot(wang_filtered, features = "nFeature_RNA")
FeaturePlot(wang_filtered, features = "percent.mt")
```

We see sample differences (instead of cell types) determine the clustering also when picking lower number of dimensions. Donor 3 clusters separately, especially in cluster 1 (predominantly donor 3) and cluster 0 (predominantly all other donors). 
We therefore have to integrate the samples before continuing. For this, we use the seurat v4 workflow and the anchor-based RPCA integration method. We chose this method since it runs faster and integrates more conservatively (recommended to integrate data from the same species / generated on the same platform). 

```{r sample integration + clustering}
#split samples
wang_filtered.samples <- SplitObject(wang_filtered, split.by = "sample")

#normalize and identify variable features for each sample independently
wang_filtered.samples <- lapply(wang_filtered.samples, function(x) {
  print(x$sample[[1]]) # which element of the list get's processed
  x <- NormalizeData(x) 
  x <- FindVariableFeatures(x)
})

#select features that are repeatedly variable across sample for integration
integration_features <- SelectIntegrationFeatures(wang_filtered.samples)

#run PCA on each sample using these features
wang_filtered.samples <- lapply(wang_filtered.samples, function(x) {
  print(x$sample[[1]]) # which element of the list get's processed
  x <- ScaleData(x, features = integration_features) 
  x <- RunPCA(x, features = integration_features)
}) 

#identifies integration anchors
integration_anchors <- FindIntegrationAnchors(wang_filtered.samples, anchor.features = integration_features, reduction = "rpca") 

#integrates based on these anchors
wang_filtered <- IntegrateData(integration_anchors) #the resulting warning is normal behavior in the seurat pipeline

DefaultAssay(wang_filtered) <- "RNA"
wang_filtered <- JoinLayers(wang_filtered)

#now run clustering using the integrated data, the original unmodified data is still in the RNA assay
DefaultAssay(wang_filtered) <- "integrated"
wang_filtered <- seuratpipeline(wang_filtered, res = 0.1, dims = 20) 

#visualization clustering of integrated donors
DimPlot(wang_filtered, reduction = "umap", group.by = "seurat_clusters")
DimPlot(wang_filtered, reduction = "umap", group.by = "seurat_clusters", label = T)
DimPlot(wang_filtered, reduction = "umap", group.by = "sample")

FeaturePlot(wang_filtered, features = "nCount_RNA")
FeaturePlot(wang_filtered, features = "nFeature_RNA")
FeaturePlot(wang_filtered, features = "percent.mt")
```

Clustering now shows no dependence on the sample donors anymore. Number of counts, features or percentage of mitochondrial counts also does not majorly influence clustering. We therefore continue with identifying cell types in the biopsy data. 

First we check the most significant differentially expressed genes for each cluster. 

```{r marker analysis}
DefaultAssay(wang_filtered) <- "integrated"
wang_filtered.markers <- FindAllMarkers(wang_filtered, only.pos = TRUE, min.pct = 0.25)

# top10 <- wang_filtered.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)
# DoHeatmap(wang_filtered, features = top10$gene) + NoLegend()

# saveRDS(wang_filtered.markers, paste0(save.dir, "wang_filtered.markers.rds"))
# write.xlsx(wang_filtered.markers, file = paste0(save.dir, "wang_filtered.markers.xlsx"))
```

cluster 0: differentiated KCs
cluster 1: CC (basal I / II)
cluster 2: roughly basal III
cluster 3: melanocytes

```{r annotation}
# wang_filtered <- readRDS(paste0(save.dir, "wang_filtered.rds"))

Idents(wang_filtered) <- wang_filtered$seurat_clusters
DimPlot(wang_filtered, label = T) + NoLegend()
VlnPlot(wang_filtered, features = c("KRT1", "KRT10", #differentiated KCs
                                    "KRT5", "KRT14", #basal KCs
                                    "KRT15", #stem cells
                                    "PMEL", "MITF", "DCT", "TYRP1", "GPR143", "MLANA", #melanocytes
                                    "CD200", "KRT19", "SOX9", "RUNX1", "NFATC1", #hair follicle
                                    "LYZ", "CD68",  #monocyte/macrophage 
                                    "CD207", "CD86", #langerhans cells
                                    "LYVE1", "PROX1",  #lymphatic vessel
                                    "PECAM1", "CLDN5", #endothelial cells
                                    "MYL9", #smooth muscle
                                    "MBP", "SOX10", #oligodendrocyte
                                    "SOX9", #astrocyte / sertoli cell
                                    "HBB", "HBG2", "HBA1", #erythrocyte
                                    "COL1A1" #fibroblasts
                                    ),
        flip = T, stack = T) + theme(axis.title.x = element_blank()) + NoLegend()

# cluster3GO <-  as.data.frame(enrichGO(gene = wang_filtered.markers[wang_filtered.markers$cluster == "3", "gene"], OrgDb = org.Hs.eg.db, keyType = 'SYMBOL', ont = "ALL", readable = T))
# cluster4GO <-  as.data.frame(enrichGO(gene = wang_filtered.markers[wang_filtered.markers$cluster == "4", "gene"], OrgDb = org.Hs.eg.db, keyType = 'SYMBOL', ont = "ALL", readable = T))
# cluster5GO <-  as.data.frame(enrichGO(gene = wang_filtered.markers[wang_filtered.markers$cluster == "5", "gene"], OrgDb = org.Hs.eg.db, keyType = 'SYMBOL', ont = "ALL", readable = T))
# cluster6GO <-  as.data.frame(enrichGO(gene = wang_filtered.markers[wang_filtered.markers$cluster == "6", "gene"], OrgDb = org.Hs.eg.db, keyType = 'SYMBOL', ont = "ALL", readable = T))

wang_filtered <-  RenameIdents(wang_filtered,
                            '0' = "keratinocyte",
                            '1' = "keratinocyte",
                            '2' = "keratinocyte",
                            '3' = "melanocyte",
                            '4' = "hair follicle",
                            "5" = "endothelial cell",
                            "6" = "langerhans cell")

#Supplemental Figure 1c
VlnPlot(wang_filtered, features = c("KRT1", #differentiated KCs
                                    "KRT5", #basal KCs
                                    "KRT15", #stem cells
                                    "MITF", "MLANA", #melanocytes
                                    "SOX9", "RUNX1", #hair follicle
                                    "PECAM1", "CLDN5", #endothelial cells
                                    "CD207", "CD86" #langerhans cells
                                    ), flip = T, stack = T,
        cols = c(rep("#53B400", 3), rep("#F8766D", 2), rep("#E76BF3", 2), rep("#00B6EB", 2), rep("#A3A500", 2))) +
  scale_x_discrete(labels = scales::label_wrap(15)) +
  theme(axis.title.x = element_blank()) + NoLegend()

#Supplemental Figure 1b
DimPlot(wang_filtered, reduction = "umap", label = T, repel = T,
        cols = c("#53B400", "#F8766D", "#E76BF3", "#00B6EB", "#A3A500")) + NoLegend() 

# saveRDS(wang_filtered, paste0(save.dir, "wang_filtered.rds"))
```

cluster 0: fully differentiated KCs
cluster 1: differentiating KCs
cluster 2: basal KCs
cluster 3: melanocytes
cluster 4: hair follicle
cluster 5: endothelial / lymphatic vessle
cluster 6: monocyte lineage (langerhans cell / monocyte / macrophage)

## Keratinocye quality control, clustering and annotation

```{r extract + filter keratinocytes}
# wang_kcsubset <- readRDS(paste0(save.dir, "wang_kcsubset.rds"))
# wang_kcfiltered <- readRDS(paste0(save.dir, "wang_kcfiltered.rds"))

#isolate keratinocytes
wang_kcsubset <- subset(wang_filtered, idents = c("keratinocyte"))

#filter keratinocytes
Idents(wang_kcsubset) <- wang_kcsubset$sample

#Supplemental Figure 1d
p1 <- VlnPlot(wang_kcsubset, features = c("nCount_RNA"), flip = T, pt.size = 0, cols = c(rep("#53B400", 5))) +
  geom_hline(yintercept = 40000) + 
  labs(x = element_blank(), y = c("total counts per cell"), title = element_blank()) + 
  theme(legend.position = "none")

p2 <- VlnPlot(wang_kcsubset, features = c("nFeature_RNA"), flip = T, pt.size = 0, cols = c(rep("#53B400", 5))) +
  # geom_hline(yintercept = 900) +  geom_hline(yintercept = 5500) + 
  labs(x = element_blank(), y = c("genes per cell"), title = element_blank()) + 
  theme(legend.position = "none")

p3 <- VlnPlot(wang_kcsubset, features = c("percent.mt"), flip = T, pt.size = 0, cols = c(rep("#53B400", 5))) +
  geom_hline(yintercept = 1) + geom_hline(yintercept = 10) +
  scale_y_continuous(labels = scales::label_dollar(suffix = "%", prefix = "")) +
  labs(x = element_blank(), y = c("mitochondrial genes per cell"), title = element_blank()) + 
  theme(legend.position = "none")

patchwork::wrap_plots(p1, p2, p3) 

#KC quality control filtering
wang_kcfiltered <- subset(wang_kcsubset, subset = nCount_RNA < 40000
                        & percent.mt > 1 & percent.mt < 10)

VlnPlot(wang_kcfiltered, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"), pt.size = 0)

# saveRDS(wang_kcsubset, paste0(save.dir, "wang_kcsubset.rds"))
```

Filtering keratinocytes is not easy since donor 3 shows deeper sequencing than the other donors. More stringent filters on percentage (of mitochondrial genes) might treat the different donors more similar than applying too stringent gene / total count filters. 

```{r cluster keratinocytes}
# wang_kcfiltered <- readRDS(paste0(save.dir, "wang_kcfiltered.rds"))

DefaultAssay(wang_kcfiltered) <- "integrated"
wang_kcfiltered <- ScaleData(wang_kcfiltered)
wang_kcfiltered <- seuratpipeline(wang_kcfiltered, nfeat = 5000, npcs = 100, dims = 20, res = 0.8)

DimPlot(wang_kcfiltered, reduction = "umap", label = T) 
DimPlot(wang_kcfiltered, reduction = "umap", group.by = "sample")

FeaturePlot(wang_kcfiltered, features = "nCount_RNA")
FeaturePlot(wang_kcfiltered, features = "nFeature_RNA")
FeaturePlot(wang_kcfiltered, features = "percent.mt")
#counts, features, % mt RNA do not majorly influence clustering
```

Since the reposed data for Wang et al. 2019 does not contain annotations, we have to annotate the basal cell types ourselves based on the provided markers (mentioned in the paper) and differentially expressed genes (provided in Supplementary Data 1, integrative clustering of all samples).

```{r identify keratinocyte subtypes - by cell type resemblance}
# wang_kcfiltered <- readRDS(paste0(save.dir, "wang_kcfiltered.rds"))
# wang_kc.markersCluster <- readRDS(paste0(save.dir, "wang_kc.markersCluster.rds"))

#identify DE marker genes
Idents(wang_kcfiltered) <- wang_kcfiltered$seurat_clusters
DefaultAssay(wang_kcfiltered) <- "integrated"
wang_kc.markersCluster <- FindAllMarkers(wang_kcfiltered, only.pos = TRUE, min.pct = 0.25)
wang_kc.markersCluster2 <- wang_kc.markersCluster[wang_kc.markersCluster$p_val_adj < 0.05, ]

# saveRDS(wang_kc.markersCluster, paste0(save.dir, "wang_kc.markersCluster.rds"))
# write.xlsx(wang_kc.markersCluster, file = paste0(save.dir, "wang_kc.markersCluster.xlsx"))

#load reference markers (Wang et al. 2019, Supplementary Data 1, integrative clustering)
wang.orgmarkers <- readxl::read_excel(paste0(data.dir, "Wang2019_SData1_41467_2020_18075_MOESM2_ESM.xlsx"), range = "A2:D2028")
wang.orgmarkers <- as.data.frame(wang.orgmarkers)

#confusion matrix with all clusters
clusters <- c(0:15)
org_markers.basI <- vector()
org_markers.basII <- vector()
org_markers.basIII <- vector()
org_markers.basIV <- vector()

#check how many original markers are also DE cluster markers
for (clus in 1:(length(clusters))) { #iterate through all clusters
  print (clus)
  org_markers.basI[clus] <- table(wang.orgmarkers[wang.orgmarkers$Cluster == "BAS-I", "Row"] %in% wang_kc.markersCluster2[wang_kc.markersCluster2$cluster == clus-1, "gene"])["TRUE"]
  org_markers.basII[clus] <- table(wang.orgmarkers[wang.orgmarkers$Cluster == "BAS-II", "Row"] %in% wang_kc.markersCluster2[wang_kc.markersCluster2$cluster == clus-1, "gene"])["TRUE"]
  org_markers.basIII[clus] <- table(wang.orgmarkers[wang.orgmarkers$Cluster == "BAS-III", "Row"] %in% wang_kc.markersCluster2[wang_kc.markersCluster2$cluster == clus-1, "gene"])["TRUE"]
  org_markers.basIV[clus] <- table(wang.orgmarkers[wang.orgmarkers$Cluster == "BAS-IV", "Row"] %in% wang_kc.markersCluster2[wang_kc.markersCluster2$cluster == clus-1, "gene"])["TRUE"]
  }

#write a 0 in clusters where no original marker is present
org_markers.basI[is.na(org_markers.basI)] <- 0
org_markers.basII[is.na(org_markers.basII)] <- 0
org_markers.basIII[is.na(org_markers.basIII)] <- 0
org_markers.basIV[is.na(org_markers.basIV)] <- 0

#make the confusion matrix
cluster_names <- vector()
for (clus in 0:(length(clusters)-1)){
  cluster_names <- append(cluster_names, paste("markers_cluster", clus, sep = "."))
}

confusion_matrix <- data.frame(org_markers.basI = org_markers.basI,
                               org_markers.basII = org_markers.basII,
                               org_markers.basIII = org_markers.basIII,
                               org_markers.basIV = org_markers.basIV, row.names = cluster_names)

#transform into percentages
confusion_matrix <- data.frame(
  org_markers.basI = round(org_markers.basI / sum(org_markers.basI) * 100, digits = 2),
  org_markers.basII = round(org_markers.basII / sum(org_markers.basII) * 100, digits = 2),
  org_markers.basIII = round(org_markers.basIII / sum(org_markers.basIII) * 100, digits = 2),
  org_markers.basIV = round(org_markers.basIV / sum(org_markers.basIV) * 100, digits = 2), row.names = cluster_names)

confusion_matrix <- gt(confusion_matrix, rownames_to_stub = TRUE)
data_color(confusion_matrix, palette = c("white", "blue"))
```

```{r identify keratinocyte subtypes - by marker genes}
# wang_kcfiltered <- readRDS(paste0(save.dir, "wang_kcfiltered.rds"))

Idents(wang_kcfiltered) <- wang_kcfiltered$seurat_clusters
#basal I
VlnPlot(wang_kcfiltered, features = c("KRT5", "KRT14", "KRT1", "KRT10", #general markers
"PTTG1", "CDC20", "MKI67", "UBE2C", "TUBA1B", "CENPF", "TOP2A", "CCNB1", "CDKN3", "MT2A", "CALM2", "BIRC5", "CDK1"), stack = T, flip = T) + theme(axis.title.x = element_blank()) + NoLegend()

#basal II
VlnPlot(wang_kcfiltered, features = c("KRT5", "KRT14", "KRT1", "KRT10", #general markers
"RRM2", "HELLS", "UHRF1", "PCNA", "MT1X", "RANBP1", "MCM7", "MCM3", "DNMT1", "MCM5", "IDH2", "NUDT5", "MCM2", "MCM10"), stack = T, flip = T) + theme(axis.title.x = element_blank()) + NoLegend()

#basal III
VlnPlot(wang_kcfiltered, features = c("KRT5", "KRT14", "KRT1", "KRT10", #general markers
                                      "ASS1", "COL17A1", "POSTN", "KRT15", "CXCL14", "S100A2", "DCN", "DST", "SYT8"), stack = T, flip = T) + theme(axis.title.x = element_blank()) + NoLegend()

#basal IV
VlnPlot(wang_kcfiltered, features = c("KRT5", "KRT14", "KRT1", "KRT10", #general markers
                                      "KRT6A",  "KRT16", "KRT13", "S100A6", "MT1G"), stack = T, flip = T) + theme(axis.title.x = element_blank()) + NoLegend()

#stratum spinosum
VlnPlot(wang_kcfiltered, features = c("KRT5", "KRT14", "KRT1", "KRT10", #general markers
"CCND1", "DEFB1", "FXYD3", "RPS18", "DSC3", "CLEC2B", "RPS18", "RPS8", "RPL13", "IGF2"), stack = T, flip = T) + theme(axis.title.x = element_blank()) + NoLegend()

#stratum granulosum
VlnPlot(wang_kcfiltered, features = c("KRT5", "KRT14", "KRT1", "KRT10", #general markers
                                      "CALML5", "ZNF750", "SPINK5", "TGM3", "IVL", "FLG", "KRT2"), stack = T, flip = T) + theme(axis.title.x = element_blank()) + NoLegend()
```

```{r identify keratinocyte subtypes - by distribution percentage}
# wang_kcfiltered <- readRDS(paste0(save.dir, "wang_kcfiltered.rds"))

table(wang_kcfiltered$seurat_clusters)

# basal I: 720 (cluster 11 + 13)
# basal II: 1850
# basal III: 1550
# basal IV: 850 (cluster 6)
```

basal I: 11, 13 (11 based on resemblance, 13 based on markers, umap position + percentages of subtypes in Wang figure 1B)
basal II: 3, 10 (3 based on resemblance, 10 based on marker gene expression)
basal III: 2 (2 based on resemblance)
basal IV: 6 (6 based on resemblance + umap position)

spinosum: 0, 4, 5, 7, 8, 9, 14, 15 (based on (exclusion of other) marker gene expression)
granulosum: 1, 12 (1, 12 based on marker gene expression, 1 is early granulosum)

```{r annotate keratinocyte subtypes}
# wang_kcfiltered <- readRDS(paste0(save.dir, "wang_kcfiltered.rds"))

Idents(wang_kcfiltered) <- wang_kcfiltered$seurat_clusters
wang_kc <-  RenameIdents(wang_kcfiltered,
                            '0' = "stratum spinosum",
                            '1' = "stratum granulosum",
                            '2' = "basal III", 
                            '3' = "basal II", 
                            '4' = "stratum spinosum",
                            "5" = "stratum spinosum", 
                            "6" = "basal IV",
                            "7" = "stratum spinosum",
                            "8" = "stratum spinosum", 
                            "9" = "stratum spinosum",
                            "10" = "basal II", 
                            "11" = "basal I",
                            "12" = "stratum granulosum",
                            "13" = "basal I",
                            "14" = "stratum spinosum",
                            "15" = "stratum spinosum")

Idents(wang_kc) <- factor(Idents(wang_kc), levels = c("basal I", "basal II", "basal III", "basal IV", "stratum spinosum", "stratum granulosum")) #reorder 
wang_kc$annotation <- Idents(wang_kc)

# saveRDS(wang_kcfiltered, paste0(save.dir, "wang_kcfiltered.rds"))
# saveRDS(wang_kc, paste0(save.dir, "wang_kc.rds"))
```

```{r visualize annotation}
# wang_kc <- readRDS(paste0(save.dir, "wang_kc.rds"))

Idents(wang_kc) <- wang_kc$annotation

#Supplemental Figure 1e
DimPlot(wang_kc, reduction = "umap", pt.size = 0.5, label = T, repel = T, cols = colors_subtypes) + NoLegend() 

#visualize marker expression of Wang et al subtype markers, Supplemental Figure 1f
VlnPlot(wang_kc, features = c("MKI67", "TOP2A", #basal I
                              "MCM3", "MCM5",  #basal II
                              "ASS1", "POSTN", #basal III
                              "KRT16", "MT1G", #basal IV "KRT6A"
                              #spinosum "CCND1", "DEFB1" "SERPINB3", "HES1", "CLDN5",
                              "IVL", "FLG", #granulosum
                             "KRT5",  "KRT1" #general markers
                             ), 
        cols = c(rep("#FFD92F", 2), rep("#66C2A5", 2), rep("#8DA0CB", 2), c(rep("#A6D854", 2)), c(rep("#E78AC3", 2)), c(rep("#FC8D62", 2))), flip = T, stack = T) +
  theme(axis.title.x = element_blank()) +
  scale_x_discrete(labels = scales::label_wrap(15)) + NoLegend()

FeaturePlot(wang_kc, features = c("KRT5", "KRT1", "MKI67", "MCM3", "POSTN", "IVL"))

#visualize cell percentage, Supplemental Figure 1g
cell_percentage  <- percentage_df(wang_kc, cell_annotation = "annotation")
cell_percentage$Celltype <- factor(cell_percentage$Celltype, levels = c("basal I", "basal II", "basal III", "basal IV", "stratum spinosum", "stratum granulosum")) #reorder

ggplot(cell_percentage, aes(fill = Celltype, y = percentage, x = Variable)) + 
    geom_bar(position = "stack", stat = "identity") + 
    scale_y_continuous(labels = scales::label_dollar(suffix = "%", prefix = "")) +
    xlab(c("")) + ylab("relative cell number") + 
    geom_text(aes(label = percentage), position = position_stack(vjust = 0.5), size = 3) + 
    scale_fill_manual(values = colors_subtypes, name = "cell type") + 
    theme_classic() + theme(text = element_text(size = 10))
```

We can now cross-check the markers of our annotated basal subpopulations with the annotation of Wang et al. 2019 (Supplementary Data 1, integrative clustering). Be aware that the clustering in Wang et al. is made using a different integration and different clustering algorithm, therefore changes are normal.  

```{r check subtype annotation by marker gene resemblance}
# wang_kc <- readRDS(paste0(save.dir, "wang_kc.rds"))
# wang_kc.markersAnnotation <- readRDS(paste0(save.dir, "wang_kc.markersAnnotation.rds"))

#check marker expression of subtypes
DefaultAssay(wang_kc) <- "integrated"
Idents(wang_kc) <- wang_kc$annotation
wang_kc.markersAnnotation <- FindAllMarkers(wang_kc, only.pos = TRUE, min.pct = 0.25)
wang_kc.markersAnnotation2 <- wang_kc.markersAnnotation[wang_kc.markersAnnotation$p_val_adj < 0.05, ]

# saveRDS(wang_kc.markersAnnotation, paste0(save.dir, "wang_kc.markersAnnotation.rds"))
# write.xlsx(wang_kc.markersAnnotation, file = paste0(save.dir, "wang_kc.markersAnnotation.xlsx"))

#load published Wang et al. markers
wang_kc.orgmarkers <- readxl::read_excel(paste0(data.dir, "Wang2019_SData1_41467_2020_18075_MOESM2_ESM.xlsx"), range = "A2:D2028") #open supplementary data 1 of Wang et al. 2019
wang_kc.orgmarkers <- as.data.frame(wang_kc.orgmarkers)

#confusion matrix with all clusters
celltypes <- levels(wang_kc$annotation)[1:4]
org_markers.basI <- vector()
org_markers.basII <- vector()
org_markers.basIII <- vector()
org_markers.basIV <- vector()

#check how many original markers are also DE in subtypes
for (i in celltypes) { #iterate through all subtypes
  org_markers.basI[i] <- table(wang_kc.orgmarkers[wang_kc.orgmarkers$Cluster == "BAS-I", "Row"] %in% wang_kc.markersAnnotation2[wang_kc.markersAnnotation2$cluster == i, "gene"])["TRUE"]
  org_markers.basII[i] <- table(wang_kc.orgmarkers[wang_kc.orgmarkers$Cluster == "BAS-II", "Row"] %in% wang_kc.markersAnnotation2[wang_kc.markersAnnotation2$cluster == i, "gene"])["TRUE"]
  org_markers.basIII[i] <- table(wang_kc.orgmarkers[wang_kc.orgmarkers$Cluster == "BAS-III", "Row"] %in% wang_kc.markersAnnotation2[wang_kc.markersAnnotation2$cluster == i, "gene"])["TRUE"]
  org_markers.basIV[i] <- table(wang_kc.orgmarkers[wang_kc.orgmarkers$Cluster == "BAS-IV", "Row"] %in% wang_kc.markersAnnotation2[wang_kc.markersAnnotation2$cluster == i, "gene"])["TRUE"]
  }

#write a 0 in clusters where no original marker is present
org_markers.basI[is.na(org_markers.basI)] <- 0
org_markers.basII[is.na(org_markers.basII)] <- 0
org_markers.basIII[is.na(org_markers.basIII)] <- 0
org_markers.basIV[is.na(org_markers.basIV)] <- 0

#make the confusion matrix
cluster_names <- vector()
for (i in celltypes){
  cluster_names <- append(cluster_names, paste("assigned", i, sep = " "))
}

confusion_matrix <- data.frame(org_markers.basI = org_markers.basI,
                               org_markers.basII = org_markers.basII,
                               org_markers.basIII = org_markers.basIII,
                               org_markers.basIV = org_markers.basIV, row.names = cluster_names)

#transform into percentages
confusion_matrix <- data.frame(
  org_markers.basI = round(org_markers.basI / sum(org_markers.basI) * 100, digits = 2),
  org_markers.basII = round(org_markers.basII / sum(org_markers.basII) * 100, digits = 2),
  org_markers.basIII = round(org_markers.basIII / sum(org_markers.basIII) * 100, digits = 2),
  org_markers.basIV = round(org_markers.basIV / sum(org_markers.basIV) * 100, digits = 2),
  row.names = cluster_names)

#rename columns
colnames(confusion_matrix) <- c("original basal I", "original basal II", "original basal III", "original basal IV")

confusion_matrix <- tibble::rownames_to_column(confusion_matrix, var = "Cluster")

#create gt table with percentage formatting and color mapping
confusion_matrix_gt <- gt(confusion_matrix) %>%
  fmt_number(columns = -Cluster, decimals = 1) %>%  # Format numbers with 1 decimals
  fmt(columns = -Cluster, fns = function(x) paste0(x, "%")) %>%  # Append '%' sign
  data_color(columns = -Cluster, palette = c("white", "#08306b"))

#Supplemental Figure 1h
confusion_matrix_gt
```

Now we have confidently annotated the basal subtypes Wang dataset according to basal subtype annotation proposed in Wang et al. 2019. Later, we use these labelled transcriptomes as a training dataset for our machine learning model to annotate our test dataset (Rojahn et al. 2020 HC and AD biopsy keratinocytes). The training dataset has to have the exact same genes present as in the test dataset, so now we load the Rojahn et al. 2020 and then intersect the genes. Afterwards we export both datasets to perform SVM annotation in python.

# Loading GSE153760 / Rojahn et al. 2020 dataset 

please download the GSE153760_RAW.tar data file from https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE153760 into the data.dir. 

We start with reading all samples files and combine them into one seurat object.

For each sample, we have 
- a count slot, showing the counts of genes for all cells in the sample
- information about the percentage of mitochondrial genes / sample (we remove those genes from the count slot)
- metadata of the sample

```{r load and combine samples}
#obtain the sample metadata from GEO
geo <- GEOquery::getGEO("GSE153760")
info_geo <- pData(phenoData(geo[[1]]))[ , c(1, 2, 36, 37, 38, 40, 42)]
rownames(info_geo) <- info_geo$title

data10x <- scCustomize::Read10X_GEO(data_dir = paste0(data.dir, c("GSE153760_RAW")))

#remove GSM numbers from the sample names
names(data10x) <- gsub(pattern = "(GSM[0-9]+_)|(_$)", replacement = "", x = names(data10x))

#create a list that contains the seurat object of all samples
seuratlist <- list()

for (i in names(data10x)){
  print(i)
  
  #prepare the metadata in the dataframe info
  info <- data.frame(row.names = colnames(as.data.frame(data10x[[i]])))
  info$geo <- info_geo$geo_accession[info_geo$title == i]
  info$sample <- info_geo[i, "title"]
  info$state <-info_geo[i, "diagnosis:ch1"]
  info$method <- info_geo[i, "tissue:ch1"]
  
  #create the seurat object with metadata + count table
  seurat <- CreateSeuratObject(counts = data10x[[i]],
                             meta.data = info,
                             min.cells = 3,
                             project = c("GSE153760"))
  seurat$orig.ident <- c("GSE153760")
  
  #add to the list
  seuratlist[[i]] <- seurat
}

#the cell names contain sample numbers -> remove sample number from the cell name
for (i in c(1:length(seuratlist))){
  # print (i)
  colnames(seuratlist[[i]]) <- gsub(pattern = "^[0-9]+_", replacement = "", colnames(seuratlist[[i]]))
}
# tail(colnames(seuratlist[[5]])) #test if it worked

#merge individual seurat objects / samples into one 
rojahn <- merge(
  x = seuratlist[[1]], y = c(seuratlist[c(2:length(seuratlist))]), #seurat objects to be merged
  add.cell.ids = names(seuratlist[1:length(seuratlist)]), #makes cell names unique (required) by adding sample name
  project = "GSE153760"
)

#remove layers
rojahn[['RNA']] <- JoinLayers(object = rojahn[['RNA']])
```

```{r rename and reorder metadata}
#rename + reorder tissue state
Idents(rojahn) <- rojahn$state
rojahn <- RenameIdents(rojahn, "Healthy Control" = "HC", "Atopic Dermatitis" = "AD")
levels(rojahn) #check
rojahn@meta.data[["state"]] <- Idents(rojahn)

#reorder sample
Idents(rojahn) <- rojahn$sample
rojahn@meta.data[["sample"]] <- factor(rojahn@meta.data[["sample"]], levels = c("HC1", "HC2", "HC3", "HC4", "HC5", "HC6", "HC7", "AD1", "AD2", "AD3", "AD4", "AD5", "AD6", "AD7", "AD8")) #order cell identities
levels(rojahn$sample) #check
Idents(rojahn) <- rojahn$sample
```

# Preprocessing

```{r remove subction blister samples and mitochondrial genes}
# rojahn <- readRDS(paste0(save.dir, "rojahn.rds"))

rojahn_biopsy <- subset(rojahn, subset = method == "Skin Biopsy")

#check content content of mito genes per cell
mito_genes_index <- grep(pattern = "^MT-", x = rownames(rojahn)) #find all mito genes (13 genes)
percent_mito_genes <- colSums(rojahn_biopsy[["RNA"]]$counts[mito_genes_index, ]) /
                      colSums(rojahn_biopsy[["RNA"]]$counts) * 100
# print(summary(percent_mito_genes))

#add mitochondrial gene percentage to the meta data 
rojahn_biopsy$percent.mt <- percent_mito_genes

#remove mitochondrial genes 
rojahn_biopsy <- rojahn_biopsy[-c(mito_genes_index), ]

# saveRDS(rojahn, paste0(save.dir, "rojahn.rds"))
```

## Quality control

These light filters are good here because the purpose is to get a good umap of very different cell types and then when filtering for KCs apply more stringent filters  

```{r QC - visualization and filtering}
# rojahn_biopsy <- readRDS(paste0(save.dir, "rojahn_biopsy.rds"))
# rojahn_filtered <- readRDS(paste0(save.dir, "rojahn_filtered.rds"))

rojahn_biopsy <- NormalizeData(rojahn_biopsy) #performs log normalization

#visualization, Supplemental Figure 2a
Idents(rojahn_biopsy) <- rojahn_biopsy$sample
p1 <- VlnPlot(rojahn_biopsy, features = c("nCount_RNA"), flip = T, pt.size = 0) +
  geom_hline(yintercept = 500) +  geom_hline(yintercept = 15000) + 
  labs(x = element_blank(), y = c("total counts per cell"), title = element_blank()) + 
  scale_fill_manual(values = colors_samples_state) + theme(legend.position = "none")

p2 <- VlnPlot(rojahn_biopsy, features = c("nFeature_RNA"), flip = T, pt.size = 0) +
  geom_hline(yintercept = 100) +  geom_hline(yintercept = 3500) +
  labs(x = element_blank(), y = c("genes per cell"), title = element_blank()) + 
  scale_fill_manual(values = colors_samples_state) + theme(legend.position = "none")

p3 <- VlnPlot(rojahn_biopsy, features = c("percent.mt"), flip = T, pt.size = 0) +
  geom_hline(yintercept = 10) + 
  scale_y_continuous(labels = scales::label_dollar(suffix = "%", prefix = "")) +
  labs(x = element_blank(), y = c("mitochondrial genes per cell"), title = element_blank()) + 
  scale_fill_manual(values = colors_samples_state) + theme(legend.position = "none")

patchwork::wrap_plots(p1, p2, p3) 

#applying the filter
rojahn_filtered <- subset(rojahn_biopsy, subset = percent.mt < 10 &
                            nFeature_RNA > 100 & nFeature_RNA < 3500 &
                            nCount_RNA > 500 & nCount_RNA < 15000) 

VlnPlot(rojahn_filtered, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"), pt.size = 0)

# saveRDS(rojahn_biopsy, paste0(save.dir, "rojahn_biopsy.rds"))
```

## Clustering + annotating cell types

```{r clustering}
# rojahn_filtered <- readRDS(paste0(save.dir, "rojahn_filtered.rds"))

table(rojahn_filtered$state) #HC 10828, AD 13937

rojahn_filtered <- seuratpipeline(rojahn_filtered, nfeat = 5000, res = 0.05)
DimPlot(rojahn_filtered)
```

```{r visualize cell types}
# rojahn_filtered <- readRDS(paste0(save.dir, "rojahn_filtered.rds"))

Idents(rojahn_filtered) <- rojahn_filtered$seurat_clusters

VlnPlot(rojahn_filtered, features = c("CD3D", "CD3G", #T cell
                                      "LYZ", "CD68", #monocyte lineage 
                                      "KRT5", "KRT1", #keratinocytes
                                      "COL1A2", "FBLN1", #fibroblast
                                      "PMEL", "MLANA", #melanocytes
                                      "PECAM1", "PLVAP", #endothelial cell
                                      "CMA1", "KIT", #mast cell
                                      "ACTA2", "TAGLN" #smooth muscle cells 
                                      ),
        flip = T, stack = T) + theme(axis.title.x = element_blank()) + NoLegend() |
DimPlot(rojahn_filtered, label = T) + NoLegend()

# KC stem cell: lgr5, krt15
# KCs: KRT5, KRT1 
# Hair follicle: "LGR5", "KRT6", "CD34", "CD200", "KRT15", "KRT19", "LHX2", "SOX9", "RUNX1", "AXIN2", "FOXP1", "FOXC1", "NFATC1", "MSI2", "LGR6", "HOPX", "TCF3", "PCADH", "GLI1",
# Melanocytes: "PMEL", "CD117", "HMB45", "MEL5", "MITF", "S100", "DCT", "TYRP1", "GPR143", "MLANA", 
# fibroblast: col1a1 
# endothelial cell: PECAM1, cldn5, plvap, 
#lymphatic endothelial cell: "PROX1", "LYVE1"
# monocyte / macrophage: lyz, CD68, CD93, lyz2, "CD14" (macrophage)
# langerhans cell: cd207, "CD86", 
# T cell: cd3d, cd3g, cd4, CD7
# mast cell: "CMA1", "CD33", "CD45", "ENPP3", "CD203C", "CD32", "CD34", "CCL2", "CXCR2", "KIT", "VCAM1", "CD16", "CD32", "CD63", "FCER1", "ITGA4", "ITGB7", 
# smooth muscle cell: "MYL9", "MCAM", "SMA", "CDH5", "EMILIN2", "GPR38", "HEXIM1", "HRH2", "CALD1", "ACTA1", "ACTA2", "CNN1", "TAGLN", "TAGLN2"
# oligodendrocyte: "MBP", "SOX10"
# astrocyte, sertoli cell: "SOX9"
# erythrocyte" HBB", "HBG2", "HBA1"
```

```{r annotate cell types}
# rojahn_filtered <- readRDS(paste0(save.dir, "rojahn_filtered.rds"))

Idents(rojahn_filtered) <- rojahn_filtered$seurat_clusters
rojahn_filtered <- RenameIdents(rojahn_filtered,
                                '0' = "T cell", 
                                '1' = "monocyte lineage",
                                '2' = "keratinocyte",
                                '3' = "fibroblast",
                                '4' = "endothelial cell",
                                '5' = "mast cell",
                                '6' = "smooth muscle cell") 
rojahn_filtered$cell_type <- Idents(rojahn_filtered) 

Idents(rojahn_filtered) <- rojahn_filtered$cell_type

#Supplemental Figure 2b
VlnPlot(rojahn_filtered, features = c("CD3D", "CD3G", #T cell
                                      "LYZ", "CD68", #monocyte lineage 
                                      "KRT5", "KRT1", #keratinocytes
                                      "COL1A2", "FBLN1", #fibroblast
                                      "PECAM1", "PLVAP", #endothelial cell
                                      "CMA1", "KIT", #mast cell
                                      "ACTA2", "TAGLN" #smooth muscle cells 
                                      ),
        cols = c(rep("#F8766D", 2), rep("#C49A00", 2), rep("#53B400", 2), c(rep("#00C094", 2)), c(rep("#00B6EB", 2)), c(rep("#A58AFF", 2)), c(rep("#FB61D7", 2))), #scales::hue_pal()(7)  #shows 7 default ggplot2 colors
        flip = T, stack = T) + theme(axis.title.x = element_blank()) + NoLegend() +
  scale_x_discrete(labels = scales::label_wrap(13)) #breaks long labels         

#Figure 1a
DimPlot(rojahn_filtered, reduction = "umap", label = T, label.box = F, repel = T) + NoLegend() 

# saveRDS(rojahn_filtered, paste0(save.dir, "rojahn_filtered.rds"))
```

## Extract and filter keratinocytes

```{r extract KCs}
rojahn_kcsubset <- subset(rojahn_filtered, subset = method == "Skin Biopsy", idents = "keratinocyte")
# saveRDS(rojahn_kcsubset, paste0(save.dir, "rojahn_kcsubset.rds"))
```

```{r filter KCs}
# rojahn_kcsubset <- readRDS(paste0(save.dir, "rojahn_kcsubset.rds"))
# rojahn_kcfiltered <- readRDS(paste0(save.dir, "rojahn_kcfiltered.rds"))

#visualization, Supplemental Figure 2c
Idents(rojahn_kcsubset) <- rojahn_kcsubset$sample
p1 <- VlnPlot(rojahn_kcsubset, features = c("nCount_RNA"), flip = T, pt.size = 0) +
  # geom_hline(yintercept = 500) +  geom_hline(yintercept = 9000) + 
  labs(x = element_blank(), y = c("total counts per cell"), title = element_blank()) + 
  scale_fill_manual(values = colors_samples_state) + theme(legend.position = "none")

p2 <- VlnPlot(rojahn_kcsubset, features = c("nFeature_RNA"), flip = T, pt.size = 0) +
  # geom_hline(yintercept = 100) +  geom_hline(yintercept = 3000) +
  labs(x = element_blank(), y = c("genes per cell"), title = element_blank()) + 
  scale_fill_manual(values = colors_samples_state) + theme(legend.position = "none")

p3 <- VlnPlot(rojahn_kcsubset, features = c("percent.mt"), flip = T, pt.size = 0) +
  geom_hline(yintercept = 2) +
  scale_y_continuous(labels = scales::label_dollar(suffix = "%", prefix = "")) +
  labs(x = element_blank(), y = c("mitochondrial genes per cell"), title = element_blank()) + 
  scale_fill_manual(values = colors_samples_state) + theme(legend.position = "none")

patchwork::wrap_plots(p1, p2, p3) 

#applying the filter
rojahn_kcfiltered <- subset(rojahn_kcsubset, subset = percent.mt > 2)

VlnPlot(rojahn_kcfiltered, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"), pt.size = 0)
```

# Handling SVM keratinocyte annotation

SVM annotation requires the exact same genes to be present in both datasets and that genes have exactly the same order. 

```{r gene match wang and rojahn}
# wang_kc <- readRDS(paste0(save.dir, "wang_kc.rds"))
# rojahn_kcfiltered <- readRDS(paste0(save.dir, "rojahn_kcfiltered.rds"))

shared_genes <- intersect(rownames(wang_kc@assays$RNA$counts), rownames(rojahn_kcfiltered@assays$RNA$counts))
length(shared_genes) #16016

wang_kc2 <- DietSeurat(wang_kc, features = shared_genes)
rojahn_kcfiltered2 <- DietSeurat(rojahn_kcfiltered, features = shared_genes)

wang_export <- wang_kc2@assays$RNA$counts
rojahn_export <- rojahn_kcfiltered2@assays$RNA$counts

rojahn_export <- rojahn_export[match(rownames(wang_export), rownames(rojahn_export)), ] #match gene order - prerequirement for SVM

# saveRDS(rojahn_kcfiltered, paste0(save.dir, "rojahn_kcfiltered.rds"))
```

Rojahn: 16016 features across 4652 samples
Wang: 16016 features across 14479 samples 

Since updating to seurat v5 the SaveH5Seurat function / export to h5ad file is not working anymore. As a workaround we export the count matrix, gene names and metadata file separately and re-create the file in python. 

```{r export for SVM annotation}
identical(rownames(wang_export), rownames(rojahn_export)) #double check identical order

#create train dataset (wang et al 2019)
write.csv(as.character(wang_kc2$annotation), paste0(svm.data, "Wang_kc2_labels.csv"), row.names = F)

write.table(rownames(wang_export), file = paste0(svm.data, "Wang_kc2_features.tsv"), row.names = FALSE, col.names = FALSE, quote = FALSE, sep = "\t") #gene names
write.table(wang_kc2@meta.data, file = paste0(svm.data, "Wang_kc2_metadata.tsv"), quote = FALSE, sep = "\t") #metadata
Matrix::writeMM(obj = Matrix(wang_export, sparse = T), file = paste0(svm.data, "Wang_kc2_count.mtx")) #count matrix

# create 5- fold cross-validation data
CV <- function(labels.loc, col_Index = 1, OutputDir = "./"){
  # function adapted from Abdelaal et al. (2019)
  
  lab <- as.matrix(read.csv(labels.loc))
  lab <- as.vector(lab[ , col_Index])
  
  Removed_classes <- !(table(lab) > 10)
  Cells_to_Keep <- !(is.element(lab, names(Removed_classes)[Removed_classes]))
  lab <- lab[Cells_to_Keep]
  
  # Getting training and testing Folds
  library(rBayesianOptimization)
  n_folds = 5
  Folds <- KFold(lab, nfolds = n_folds, stratified = TRUE)
  Test_Folds <- c(n_folds:1)
  Train_Idx <- list()
  Test_Idx <- list()
  for (i in c(1:length(Folds))){
    Temp_Folds <- Folds
    Temp_Folds[Test_Folds[i]] <- NULL
    Train_Idx[i] <- list(unlist(Temp_Folds))
    Test_Idx[i] <- Folds[Test_Folds[i]]
  }
  remove(Temp_Folds,i,Folds)
  
  save(n_folds, Train_Idx, Test_Idx, col_Index, Cells_to_Keep, file = paste0(OutputDir, 'CV_folds.RData'))
}

CV(labels.loc = paste0(svm.data, "Wang_kc2_labels.csv"), OutputDir = svm.data)

#create test dataset (rojahn et al 2020)
write.table(rownames(rojahn_export), file = paste0(svm.data, "Rojahn_kcfiltered2_features.tsv"), row.names = FALSE, col.names = FALSE, quote = FALSE, sep = "\t") #gene names
write.table(rojahn_kcfiltered2@meta.data, file = paste0(svm.data, "Rojahn_kcfiltered2_metadata.tsv"), quote = FALSE, sep = "\t") #metadata
Matrix::writeMM(obj = Matrix(rojahn_export, sparse = T), file = paste0(svm.data, "Rojahn_kcfiltered2_count.mtx")) #count matrix
```

!!! STOP!!!

Now switch to python to perfom SVM annotation. Read the "How_to_run.txt" for further instructions. 

! IT IS CRUCIAL THAT BETWEEN EXPORT AND ANNOTATION NO CHANGES ARE MADE TO THE DATASET (the order of the cells has to stay the same)

```{r annotate and reorder annotation}
SVM <- read.csv(paste0(svm.results, "RojahnBiop/SVM_Pred_Labels.csv"))

if (length(colnames(rojahn_kcfiltered2)) == length(SVM[ , 1])){
  rojahn_kcfiltered2$svm_annotation <- SVM[ , 1]
  Idents(rojahn_kcfiltered2) <- rojahn_kcfiltered2$svm_annotation
  rojahn_kcfiltered2@meta.data[["svm_annotation"]] <- factor(rojahn_kcfiltered2@meta.data[["svm_annotation"]], levels = c("basal I", "basal II", "basal III", "basal IV", "stratum spinosum", "stratum granulosum")) #order cell identities
  # saveRDS(rojahn_kcfiltered2, paste0(save.dir, "rojahn_kcfiltered2.rds"))
  
} else {
  print("Data set and annotation need to have the same length. Data set needs to be in the same order as exported.")
}

table(rojahn_kcfiltered2$svm_annotation)

#basal I 164
#basal II 221
#basal III 1764
#basal IV 275
#stratum spinosum 2059
#stratum granulosum 169
```

# Filter and integrate keratinocytes

```{r cluster KCs}
# rojahn_kcfiltered2 <- readRDS(paste0(save.dir, "rojahn_kcfiltered2.rds"))

#perform clustering
##normalization already done
DefaultAssay(rojahn_kcfiltered2) <- "RNA"
rojahn_kcfiltered2 <- seuratpipeline(rojahn_kcfiltered2, nfeat = 5000, res = 0.5, dims = 15)
                                     
#visualize clusters
rojahn_kcfiltered2$sample <- factor(rojahn_kcfiltered2$sample, levels = c("HC6", "HC7", "AD5", "AD6", "AD7", "AD8")) #reorder

DimPlot(rojahn_kcfiltered2, label = T, group.by = "seurat_clusters") + NoLegend() + ggtitle(element_blank()) #Supplemental Figure 2d
FeaturePlot(rojahn_kcfiltered2, features = c("nFeature_RNA"))
FeaturePlot(rojahn_kcfiltered2, features = c("nCount_RNA"))
FeaturePlot(rojahn_kcfiltered2, features = c("percent.mt"))
DimPlot(rojahn_kcfiltered2, group.by = "state")
DimPlot(rojahn_kcfiltered2, group.by = "sample", cols = colors_samples) + labs(color = "sample") + ggtitle(element_blank())
DimPlot(rojahn_kcfiltered2, group.by = "svm_annotation", cols = colors_subtypes) + labs(color = "cell type") + ggtitle(element_blank())
```

We try to identify keratinocyte cells states but we observe donor differences and tissue state to drive clustering (also if we choose different number of features  and dimensions). We also see that distinct keratinocyte subtypes did not always cluster separately. We remove clusters of outliers. 

```{r remove outliers and recluster}
#remove clusters of outliers
Idents(rojahn_kcfiltered2) <- rojahn_kcfiltered2$seurat_clusters
rojahn_kcfiltered3 <- subset(rojahn_kcfiltered2, ident = c(3, 10, 12, 13), invert = T) #those mentioned are removed

#perform clustering again
# rojahn_kcfiltered3 <- readRDS(paste0(save.dir, "rojahn_kcfiltered3.rds"))
DefaultAssay(rojahn_kcfiltered3) <- "RNA"
rojahn_kcfiltered3 <- seuratpipeline(rojahn_kcfiltered3, nfeat = 2000, res = 0.5, dims = 11)

#reorder annotation
Idents(rojahn_kcfiltered3) <- rojahn_kcfiltered3$svm_annotation
rojahn_kcfiltered3@meta.data[["svm_annotation"]] <- factor(rojahn_kcfiltered3@meta.data[["svm_annotation"]], levels = c("basal I", "basal II", "basal III", "basal IV", "stratum spinosum", "stratum granulosum")) 
levels(rojahn_kcfiltered3$svm_annotation) #check
Idents(rojahn_kcfiltered3) <- rojahn_kcfiltered3$svm_annotation

#visualize clusters
DimPlot(rojahn_kcfiltered3, label = T, group.by = "seurat_clusters")
DimPlot(rojahn_kcfiltered3, group.by = "state")
DimPlot(rojahn_kcfiltered3, group.by = "sample", cols = colors_samples) + labs(color = "sample") + ggtitle(element_blank()) #Supplemental Figure 2g
DimPlot(rojahn_kcfiltered3, group.by = "svm_annotation", cols = colors_subtypes) + labs(color = "cell type") +ggtitle(element_blank()) #Supplemental Figure 2f

# saveRDS(rojahn_kcfiltered2, paste0(save.dir, "rojahn_kcfiltered2.rds"))
```

We still see a domination of tissue state and individual sample  in the clustering. We have to integrate samples solely for clustering/visualization purposes. Gene expression analysis will be performed based on the non-integrated "RNA" data assay.  

```{r integrate KCs}
#split samples
rojahn_kcfiltered3.samples <- SplitObject(rojahn_kcfiltered3, split.by = "sample")

#normalize and identify variable features for each sample independently
rojahn_kcfiltered3.samples <- lapply(rojahn_kcfiltered3.samples, function(x) {
  print(x$sample[[1]]) # which element of the list get's processed 
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, nfeatures = 5000)
  }
)

#select features that are repeatedly variable across sample for integration
integration_features <- SelectIntegrationFeatures(rojahn_kcfiltered3.samples)

rojahn_kcfiltered3.samples <- lapply(rojahn_kcfiltered3.samples, function(x) {
  print(x$sample[[1]]) # which element of the list get's processed
  x <- ScaleData(x, features = integration_features)
  x <- RunPCA(x, features = integration_features)
})

#identifies integration anchors, reduction method: canonical correlation analysis
rojahn_kcfiltered3.samples <- FindIntegrationAnchors(rojahn_kcfiltered3.samples, anchor.features = integration_features)

#integrates based on these anchors 
rojahn_kcfiltered3 <- IntegrateData(rojahn_kcfiltered3.samples) #the resulting warning is normal behaviour in the seurat pipeline

#join layers 
DefaultAssay(rojahn_kcfiltered3) <- "RNA"
rojahn_kcfiltered3 <- JoinLayers(rojahn_kcfiltered3)
```

```{r cluster integrated KCs}
# rojahn_kcfiltered3 <- readRDS(paste0(save.dir, "rojahn_kcfiltered3.rds"))

#now run clustering using the integrated data, the original unmodified data is still in the RNA assay
DefaultAssay(rojahn_kcfiltered3) <- "integrated"
rojahn_kcfiltered3 <- seuratpipeline(rojahn_kcfiltered3, nfeat = 5000, res = 0.3, dims = 11) 
DefaultAssay(rojahn_kcfiltered3) <- "RNA" #assay back to RNA -> gene expression analysis based on non-integrated data

DimPlot(rojahn_kcfiltered3, group.by = "sample")
FeaturePlot(rojahn_kcfiltered3, features = c("nFeature_RNA"))
FeaturePlot(rojahn_kcfiltered3, features = c("nCount_RNA"))
FeaturePlot(rojahn_kcfiltered3, features = c("percent.mt"))

DimPlot(rojahn_kcfiltered3, label = T, group.by = "seurat_clusters")
DimPlot(rojahn_kcfiltered3, group.by = "state")
DimPlot(rojahn_kcfiltered3, group.by = "svm_annotation")

rojahn_kc <- rojahn_kcfiltered3
# saveRDS(rojahn_kcfiltered3, paste0(save.dir, "rojahn_kcfiltered3.rds"))
```

```{r reorder}
#reorder state
Idents(rojahn_kc) <- rojahn_kc$state
rojahn_kc@meta.data[["state"]] <- factor(rojahn_kc@meta.data[["state"]], levels = c("HC", "AD")) 
levels(rojahn_kc$state) #check
Idents(rojahn_kc) <- rojahn_kc$state

#reorder sample
Idents(rojahn_kc) <- rojahn_kc$sample
rojahn_kc@meta.data[["sample"]] <- factor(rojahn_kc@meta.data[["sample"]], levels = c("HC6", "HC7", "AD5", "AD6", "AD7", "AD8")) #reorder sample
levels(rojahn_kc$sample) #check
Idents(rojahn_kc) <- rojahn_kc$sample

#reorder annotation
Idents(rojahn_kc) <- rojahn_kc$svm_annotation
rojahn_kc@meta.data[["svm_annotation"]] <- factor(rojahn_kc@meta.data[["svm_annotation"]], levels = c("basal I", "basal II", "basal III", "basal IV", "stratum spinosum", "stratum granulosum")) 
levels(rojahn_kc$svm_annotation) #check
Idents(rojahn_kc) <- rojahn_kc$svm_annotation

# saveRDS(rojahn_kc, paste0(save.dir, "rojahn_kc.rds"))
```

# Visualize dataset

```{r visualize samples}
table(rojahn_kc$state) #HC 2108, AD 1905

Idents(rojahn_kc) <- rojahn_kc$sample
VlnPlot(rojahn_kc, features = c("nCount_RNA"), flip = T, pt.size = 0) +
  labs(x = element_blank(), y = c("total counts per cell"), title = element_blank()) + 
  scale_fill_manual(values = colors_samples_state) + theme(legend.position = "none") |
VlnPlot(rojahn_kc, features = c("nFeature_RNA"), flip = T, pt.size = 0) +
  labs(x = element_blank(), y = c("genes per cell"), title = element_blank()) + 
  scale_fill_manual(values = colors_samples_state) + theme(legend.position = "none") |
VlnPlot(rojahn_kc, features = c("percent.mt"), flip = T, pt.size = 0) + 
  scale_y_continuous(labels = scales::label_dollar(suffix = "%", prefix = "")) +
  labs(x = element_blank(), y = c("mitochondrial genes per cell"), title = element_blank()) + 
  scale_fill_manual(values = colors_samples_state) + theme(legend.position = "none")
```

```{r visualize metrics subpopulations}
Idents(rojahn_kc) <- c(paste0(rojahn_kc$state, " ", rojahn_kc$svm_annotation))

#reorder
Idents(rojahn_kc) <- factor(Idents(rojahn_kc), levels = c(
  paste0(levels(rojahn_kc$state)[1], " ", levels(rojahn_kc$svm_annotation)),
  paste0(levels(rojahn_kc$state)[2], " ", levels(rojahn_kc$svm_annotation))
  )
)

#Supplemental Figure 2e
VlnPlot(rojahn_kc, features = "nCount_RNA", pt.size = 0, cols = c(rep(c("#FFD92F", "#66C2A5", "#8DA0CB", "#A6D854", "#FC8D62", "#E78AC3"), 2))) + 
  # scale_x_discrete(labels = scales::label_wrap(15)) +
  theme(axis.title.x = element_blank()) + ylab(c("genes per cell")) + ggtitle(element_blank()) + NoLegend()
```

```{r visualize non-integrated clustering}
DefaultAssay(rojahn_kc) <- "RNA"
rojahn_kc <- seuratpipeline(rojahn_kc, nfeat = 2000, res = 0.5, dims = 11)

#visualize clustering
FeaturePlot(rojahn_kc, features = c("nFeature_RNA")) + labs(color = "counts per cell") + ggtitle(element_blank())
FeaturePlot(rojahn_kc, features = c("nCount_RNA"))  + labs(color = "genes per cell") + ggtitle(element_blank())
FeaturePlot(rojahn_kc, features = c("percent.mt"))  + labs(color = "mitochondrial \ngenes per cell") +
  scale_colour_gradient(low = "lightgrey", high = "blue",
                        labels = scales::label_dollar(suffix = "%", prefix = "")) +
  ggtitle(element_blank())

DimPlot(rojahn_kc, group.by = "state", cols = colors_state) + labs(color = "state") + ggtitle(element_blank())
DimPlot(rojahn_kc, group.by = "sample", cols = colors_samples) + labs(color = "sample") + ggtitle(element_blank())
DimPlot(rojahn_kc, group.by = "svm_annotation") + labs(color = "cell type") + scale_colour_discrete(type = colors_subtypes, labels = scales::label_wrap(10)) + ggtitle(element_blank())
FeaturePlot(rojahn_kc, features = c("KRT5", "KRT1", "KRT15", "ASS1", "MCM3", "MKI67", "FLG"))
```

```{r visualize integrated clustering}
# rojahn_kc <- readRDS(paste0(save.dir, "rojahn_kc.rds"))

DefaultAssay(rojahn_kc) <- "integrated"
rojahn_kc <- seuratpipeline(rojahn_kc, nfeat = 5000, res = 0.3, dims = 11) 
  
#visualize clustering
FeaturePlot(rojahn_kc, features = c("nFeature_RNA")) + labs(color = "counts per cell") + ggtitle(element_blank())
FeaturePlot(rojahn_kc, features = c("nCount_RNA"))  + labs(color = "genes per cell") + ggtitle(element_blank())
FeaturePlot(rojahn_kc, features = c("percent.mt"))  + labs(color = "mitochondrial \ngenes per cell") +
  scale_colour_gradient(low = "lightgrey", high = "blue",
                        labels = scales::label_dollar(prefix = "", suffix = "%")) +
  ggtitle(element_blank())

DimPlot(rojahn_kc, group.by = "state", cols = colors_state) + labs(color = "state") + ggtitle(element_blank())
DimPlot(rojahn_kc, group.by = "sample") + labs(color = "sample") + ggtitle(element_blank()) #Supplemental figure 1b
DimPlot(rojahn_kc, group.by = "svm_annotation") + labs(color = "cell type") + scale_colour_discrete(type = colors_subtypes, labels = scales::label_wrap(10)) + ggtitle(element_blank()) #figure 1b

FeaturePlot(rojahn_kc, features = c("KRT5", "KRT1", "KRT15", "ASS1", "MCM3", "MKI67", "FLG"))
```

```{r Visualization genes of interest}
DefaultAssay(rojahn_kc) <- "RNA"

#Fig 1c
DefaultAssay(rojahn_kc) <- "RNA"
VlnPlot(rojahn_kc, group.by = "svm_annotation", split.by = "state",
        features = c("KRT5", "KRT14", "KRT1", "KRT10", #general markers
                    "CLDN1",  "DSC1", "CLDN4",  #"CLDN1" also possible
                     "IVL", "TGM3", "FLG", "LOR", # "TGM3", "KLK7", "KLK8", also possible
                     'KRT6A', 'KRT16', "S100A8", "SERPINB4", #activated epithelium, S100A7, S100A8,S100A9, KRT17,
                     "CA2" #AD specific
                     ), stack = T, flip = T, cols = colors_state) +
  scale_x_discrete(labels = scales::label_wrap(10)) + 
  theme(axis.title.x = element_blank(),
        axis.title.y.right = element_text(face = "italic")) + 
  labs(fill = "state")

# Fig 1d
Feat_KRT14 <- FeaturePlot(rojahn_kc, features = c("KRT14"), split.by = "state", coord.fixed = F) +
  theme(legend.position = "right", axis.title.y.right = element_blank()) #element_text(face = "bold.italic")

Feat_KRT10 <- FeaturePlot(rojahn_kc, features = c("KRT10"), split.by = "state", coord.fixed = F) + 
  theme(legend.position = "right", axis.title.y.right = element_blank()) 

Feat_KRT6A <- FeaturePlot(rojahn_kc, features = c("KRT6A"), split.by = "state", coord.fixed = F) +
  scale_color_continuous(low = "lightgray", high = "blue",  breaks = c(0, 2, 4, 6)) +
  theme(legend.position = "right", axis.title.y.right = element_blank()) 

Feat_CA2 <- FeaturePlot(rojahn_kc, features = c("CA2"), split.by = "state", coord.fixed = F) +
  theme(legend.position = "right", axis.title.y.right = element_blank())

Feat_KRT14 / Feat_KRT10 / Feat_KRT6A / Feat_CA2

#Figure 2b
m_bIII <- c("KRT15", "ASS1", "WNT3") #"POSTN", "COL7A1", "COL18A1", "JAG2", "DKK3", "ITGB4", "WNT3"
m_bII <- c("MCM3", "GINS2", "CDT1") # "DNMT1", "MCM5", "HELLS", "MCM4", "MCM7"
m_bI <- c("MKI67", "CENPF", "CCNB2")  #"CDKN3", "TOP2A", "CCNB1", "ASPM", "BIRC5", "UBE2C", "CDC20"
m_bIV <- c("MT1H", "MT1G") #"TKT", "KRT6A", "RRP15", "ITPRIPL2", "MAPK1IP1L"

Idents(rojahn_kc) <- rojahn_kc$state
VlnPlot(rojahn_kc, idents =  c("HC"), group.by = "svm_annotation",
        features = c(m_bI, m_bII, m_bIII, m_bIV),
        cols = c(rep("#FFD92F", 3), rep("#66C2A5", 3), rep("#8DA0CB", 3), c(rep("#A6D854", 7))),
        stack = T, flip = T) +
  scale_x_discrete(labels = scales::label_wrap(10)) + theme(axis.title.x = element_blank()) + NoLegend()

# DotPlot(rojahn_kc, idents =  c("HC"), group.by = "svm_annotation",
#         features = c(m_bI, m_bII, m_bIII, m_bIV)) +
#   scale_x_discrete(guide = guide_axis(angle = 45), labels = scales::label_wrap(10)) + 
#   scale_y_discrete(limits = rev) + 
#   theme(axis.title.x = element_blank(), axis.title.y = element_blank()) 

#Supplemental figure 3a
DefaultAssay(rojahn_kc) <- "RNA"
VlnPlot(rojahn_kc[ , rojahn_kc$state == "HC"], group.by = "svm_annotation", split.by = "state", features = c("COL17A1", "ITGB1", "ITGA6"), stack = T, flip = T, cols = rep("#8DA0CB", 3)) + theme(axis.title.x = element_blank()) + NoLegend() + scale_x_discrete(labels = scales::label_wrap(10))
#i also tried "LRIG1", "DLL1", "MCSP", "CD46", "FRMD4A", "OCT4", "MYC", "OCT3", "SOX2", "KLF4"

#Supplemental figure 3e
FeaturePlot(rojahn_kc[ ,rojahn_kc$state == "HC"], features = c("KRT15", "ASS1"), blend = T, cols = c("lightgrey", "#ff0000", "#00ff00"))

#Figure 3d
m_bIII <- c("KRT15", "ASS1", "WNT3") #"POSTN", "COL7A1", "COL18A1", "JAG2", "DKK3", "ITGB4", "WNT3"
m_bII <- c("MCM3", "GINS2", "CDT1") # "DNMT1", "MCM5", "HELLS", "MCM4", "MCM7"
m_bI <- c("MKI67", "CENPF", "CCNB2")  #"CDKN3", "TOP2A", "CCNB1", "ASPM", "BIRC5", "UBE2C", "CDC20"
m_bIV <- c("MT1H", "MT1G") #"TKT", "KRT6A", "RRP15", "ITPRIPL2", "MAPK1IP1L"

Idents(rojahn_kc) <- rojahn_kc$state
VlnPlot(rojahn_kc, group.by = "svm_annotation", split.by = "state",
        features = c(m_bI, m_bII, m_bIII, m_bIV),
        cols = c(colors_state),
        stack = T, flip = T) +
  scale_x_discrete(labels = scales::label_wrap(10)) + theme(axis.title.x = element_blank()) + labs(fill = "state")

#Supplemental figure 4a
FeaturePlot(rojahn_kc[ ,rojahn_kc$state == "AD"], features = c("KRT15", "ASS1"), blend = T, cols = c("lightgrey", "#ff0000", "#00ff00"))

#Supplemental Figure 6d
VlnPlot(rojahn_kc, group.by = "svm_annotation", features = c("WNT5A", "CAV1", "CD74", "HIF1A", "HSPA1A", "PTPN2", "PYCARD", "ROCK2", "TNF"), split.by = "state", stack = T, flip = T, pt.size = 0, cols = colors_state) + theme(axis.title.x = element_blank()) + scale_x_discrete(labels = scales::label_wrap(10)) 
```

```{r lineplot HC}
DefaultAssay(rojahn_kc) <- "RNA"

#extract the gene counts per pseudobulk cluster
counts_HCbasalIII <- rojahn_kc[ , rojahn_kc$svm_annotation == "basal III" & rojahn_kc$state == "HC"]@assays[["RNA"]]$data
counts_HCbasalII <- rojahn_kc[ , rojahn_kc$svm_annotation == "basal II" & rojahn_kc$state == "HC"]@assays[["RNA"]]$data
counts_HCbasalI <- rojahn_kc[ , rojahn_kc$svm_annotation == "basal I" & rojahn_kc$state == "HC"]@assays[["RNA"]]$data
counts_HCbasalIV <- rojahn_kc[ , rojahn_kc$svm_annotation == "basal IV" & rojahn_kc$state == "HC"]@assays[["RNA"]]$data
counts_HCspin <- rojahn_kc[ , rojahn_kc$svm_annotation == "stratum spinosum" & rojahn_kc$state == "HC"]@assays[["RNA"]]$data
counts_HCgran <- rojahn_kc[ , rojahn_kc$svm_annotation == "stratum granulosum" & rojahn_kc$state == "HC"]@assays[["RNA"]]$data

#filter the counts for the required marker and prepare the data frame
counts_KCmarkers <- data.frame(
  counts = median(counts_HCbasalIII[c("KRT5"), ]),
  gene = "KRT5",
  cell_type = "basal III"
  ) 

counts_KCmarkers <- counts_KCmarkers %>%
  add_row(counts = median(counts_HCbasalII[c("KRT5"), ]), gene = "KRT5", cell_type = "basal II") %>% 
  add_row(counts = median(counts_HCbasalI[c("KRT5"), ]), gene = "KRT5", cell_type = "basal I") %>% 
  add_row(counts = median(counts_HCbasalIV[c("KRT5"), ]), gene = "KRT5", cell_type = "basal IV") %>%
  add_row(counts = median(counts_HCspin[c("KRT5"), ]), gene = "KRT5", cell_type = "stratum spinosum") %>% 
  add_row(counts = median(counts_HCgran[c("KRT5"), ]), gene = "KRT5", cell_type = "stratum granulosum") 

counts_KCmarkers <- counts_KCmarkers %>%
  add_row(counts = median(counts_HCbasalIII[c("KRT14"), ]), gene = "KRT14", cell_type = "basal III") %>% 
  add_row(counts = median(counts_HCbasalII[c("KRT14"), ]), gene = "KRT14", cell_type = "basal II") %>% 
  add_row(counts = median(counts_HCbasalI[c("KRT14"), ]), gene = "KRT14", cell_type = "basal I") %>% 
  add_row(counts = median(counts_HCbasalIV[c("KRT14"), ]), gene = "KRT14", cell_type = "basal IV") %>%
  add_row(counts = median(counts_HCspin[c("KRT14"), ]), gene = "KRT14", cell_type = "stratum spinosum") %>% 
  add_row(counts = median(counts_HCgran[c("KRT14"), ]), gene = "KRT14", cell_type = "stratum granulosum") 

counts_KCmarkers <- counts_KCmarkers %>%
  add_row(counts = median(counts_HCbasalIII[c("KRT1"), ]), gene = "KRT1", cell_type = "basal III") %>% 
  add_row(counts = median(counts_HCbasalII[c("KRT1"), ]), gene = "KRT1", cell_type = "basal II") %>% 
  add_row(counts = median(counts_HCbasalI[c("KRT1"), ]), gene = "KRT1", cell_type = "basal I") %>% 
  add_row(counts = median(counts_HCbasalIV[c("KRT1"), ]), gene = "KRT1", cell_type = "basal IV") %>%
  add_row(counts = median(counts_HCspin[c("KRT1"), ]), gene = "KRT1", cell_type = "stratum spinosum") %>% 
  add_row(counts = median(counts_HCgran[c("KRT1"), ]), gene = "KRT1", cell_type = "stratum granulosum") 

counts_KCmarkers <- counts_KCmarkers %>%
  add_row(counts = median(counts_HCbasalIII[c("KRT10"), ]), gene = "KRT10", cell_type = "basal III") %>% 
  add_row(counts = median(counts_HCbasalII[c("KRT10"), ]), gene = "KRT10", cell_type = "basal II") %>% 
  add_row(counts = median(counts_HCbasalI[c("KRT10"), ]), gene = "KRT10", cell_type = "basal I") %>% 
  add_row(counts = median(counts_HCbasalIV[c("KRT10"), ]), gene = "KRT10", cell_type = "basal IV") %>%
  add_row(counts = median(counts_HCspin[c("KRT10"), ]), gene = "KRT10", cell_type = "stratum spinosum") %>% 
  add_row(counts = median(counts_HCgran[c("KRT10"), ]), gene = "KRT10", cell_type = "stratum granulosum") 

#factorize and reorder
counts_KCmarkers$cell_type <- factor(counts_KCmarkers$cell_type, levels = c("basal I", "basal II", "basal III", "basal IV", "stratum spinosum", "stratum granulosum")) #order cell identities
# counts_KCmarkers$cell_type <- factor(counts_KCmarkers$cell_type, levels = c("basal III", "basal II", "basal I", "basal IV", "stratum spinosum", "stratum granulosum")) #order cell identities
counts_KCmarkers$gene <- factor(counts_KCmarkers$gene, levels = c("KRT5", "KRT14", "KRT1", "KRT10"))

#Figure 2a
ggplot(data = counts_KCmarkers, aes(x = cell_type, y = counts, group = gene)) +
  geom_line(aes(color = gene), linewidth = 1) + geom_point(aes(color = gene)) +
  scale_x_discrete(expand = c(0.04, 0), #gets rid of empty space beginning of x axis
                   guide = guide_axis(angle = 45),
                   labels = scales::label_wrap(10))  + #breaks long labels
  ylab("log-normalized expression") +
  theme_classic() + theme(axis.title.x = element_blank()) +
  # scale_color_manual(values = c("#6573A0", "#B0B9E6", "orange1", "orange3"))
  scale_color_manual(values = c("#6573A0", "#B0B9E6", "#F4B6DA", "#C85A9C"))
  
  #loop did not work
# bla <- c("basal II", "basal I", "basal IV", "spinosum", "granulosum")
# names(bla) <- c("counts_ADbasalII", "counts_ADbasalI", "counts_ADbasalIV", "counts_ADspin", "counts_ADgran")
# 
# 
# bla[as.character(i)]
# 
# for (i in c(counts_ADbasalII, counts_ADbasalI, counts_ADbasalIV, counts_ADspin, counts_ADgran)) {
#   # print (i)
#   counts_KCmarkers <- counts_KCmarkers %>%
#   add_row(counts = median(i[c("KRT5"), ]), gene = "KRT5", cell_type = bla[as.character(i)])
# }
```

You need to look at relative percentage because different number of samples, different number of cells sequenced, different sample quality (-> affects filtering) etc. 

```{r stacked barplot cell type percentage - HC vs AD}
cell_percentage  <- percentage_df(rojahn_kc, cell_annotation = "svm_annotation", variable = "state", facet = F)

#reorder 
cell_percentage$Celltype <- factor(cell_percentage$Celltype, levels = c("basal I", "basal II", "basal III", "basal IV", "stratum spinosum", "stratum granulosum"))
cell_percentage$Variable <- factor(cell_percentage$Variable, levels = c("HC", "AD")) 
# factor(cell_percentage$Variable)

#Figure 3a
ggplot(cell_percentage, aes(fill = Celltype, y = percentage, x = Variable)) + 
    geom_bar(position = "stack", stat = "identity") +
    geom_text(aes(label = percentage), position = position_stack(vjust = 0.5), size = 3) +
    scale_fill_manual(values = colors_subtypes, name = "cell type") +
    scale_y_continuous(labels = scales::label_dollar(suffix = "%", prefix = "")) +
    ylab("relative cell number") + 
    theme_classic() + theme(text = element_text(size = 10), axis.title.x = element_blank())
```

```{r lineplot AD}
DefaultAssay(rojahn_kc) <- "RNA"

#extract the gene counts per pseudobulk cluster
counts_ADbasalIII <- rojahn_kc[ , rojahn_kc$svm_annotation == "basal III" & rojahn_kc$state == "AD"]@assays[["RNA"]]$data
counts_ADbasalII <- rojahn_kc[ , rojahn_kc$svm_annotation == "basal II" & rojahn_kc$state == "AD"]@assays[["RNA"]]$data
counts_ADbasalI <- rojahn_kc[ , rojahn_kc$svm_annotation == "basal I" & rojahn_kc$state == "AD"]@assays[["RNA"]]$data
counts_ADbasalIV <- rojahn_kc[ , rojahn_kc$svm_annotation == "basal IV" & rojahn_kc$state == "AD"]@assays[["RNA"]]$data
counts_ADspin <- rojahn_kc[ , rojahn_kc$svm_annotation == "stratum spinosum" & rojahn_kc$state == "AD"]@assays[["RNA"]]$data
counts_ADgran <- rojahn_kc[ , rojahn_kc$svm_annotation == "stratum granulosum" & rojahn_kc$state == "AD"]@assays[["RNA"]]$data

#filter the counts for the required marker and prepare the data frame
counts_KCmarkers <- data.frame(
  counts = median(counts_ADbasalIII[c("KRT5"), ]),
  gene = "KRT5",
  cell_type = "basal III"
  ) 

counts_KCmarkers <- counts_KCmarkers %>%
  add_row(counts = median(counts_ADbasalII[c("KRT5"), ]), gene = "KRT5", cell_type = "basal II") %>% 
  add_row(counts = median(counts_ADbasalI[c("KRT5"), ]), gene = "KRT5", cell_type = "basal I") %>% 
  add_row(counts = median(counts_ADbasalIV[c("KRT5"), ]), gene = "KRT5", cell_type = "basal IV") %>%
  add_row(counts = median(counts_ADspin[c("KRT5"), ]), gene = "KRT5", cell_type = "stratum spinosum") %>% 
  add_row(counts = median(counts_ADgran[c("KRT5"), ]), gene = "KRT5", cell_type = "stratum granulosum") 

counts_KCmarkers <- counts_KCmarkers %>%
  add_row(counts = median(counts_ADbasalIII[c("KRT14"), ]), gene = "KRT14", cell_type = "basal III") %>% 
  add_row(counts = median(counts_ADbasalII[c("KRT14"), ]), gene = "KRT14", cell_type = "basal II") %>% 
  add_row(counts = median(counts_ADbasalI[c("KRT14"), ]), gene = "KRT14", cell_type = "basal I") %>% 
  add_row(counts = median(counts_ADbasalIV[c("KRT14"), ]), gene = "KRT14", cell_type = "basal IV") %>%
  add_row(counts = median(counts_ADspin[c("KRT14"), ]), gene = "KRT14", cell_type = "stratum spinosum") %>% 
  add_row(counts = median(counts_ADgran[c("KRT14"), ]), gene = "KRT14", cell_type = "stratum granulosum") 

counts_KCmarkers <- counts_KCmarkers %>%
  add_row(counts = median(counts_ADbasalIII[c("KRT1"), ]), gene = "KRT1", cell_type = "basal III") %>% 
  add_row(counts = median(counts_ADbasalII[c("KRT1"), ]), gene = "KRT1", cell_type = "basal II") %>% 
  add_row(counts = median(counts_ADbasalI[c("KRT1"), ]), gene = "KRT1", cell_type = "basal I") %>% 
  add_row(counts = median(counts_ADbasalIV[c("KRT1"), ]), gene = "KRT1", cell_type = "basal IV") %>%
  add_row(counts = median(counts_ADspin[c("KRT1"), ]), gene = "KRT1", cell_type = "stratum spinosum") %>% 
  add_row(counts = median(counts_ADgran[c("KRT1"), ]), gene = "KRT1", cell_type = "stratum granulosum") 

counts_KCmarkers <- counts_KCmarkers %>%
  add_row(counts = median(counts_ADbasalIII[c("KRT10"), ]), gene = "KRT10", cell_type = "basal III") %>% 
  add_row(counts = median(counts_ADbasalII[c("KRT10"), ]), gene = "KRT10", cell_type = "basal II") %>% 
  add_row(counts = median(counts_ADbasalI[c("KRT10"), ]), gene = "KRT10", cell_type = "basal I") %>% 
  add_row(counts = median(counts_ADbasalIV[c("KRT10"), ]), gene = "KRT10", cell_type = "basal IV") %>%
  add_row(counts = median(counts_ADspin[c("KRT10"), ]), gene = "KRT10", cell_type = "stratum spinosum") %>% 
  add_row(counts = median(counts_ADgran[c("KRT10"), ]), gene = "KRT10", cell_type = "stratum granulosum") 

#factorize and reorder
counts_KCmarkers$cell_type <- factor(counts_KCmarkers$cell_type, levels = c("basal I", "basal II", "basal III", "basal IV", "stratum spinosum", "stratum granulosum")) #order cell identities
# counts_KCmarkers$cell_type <- factor(counts_KCmarkers$cell_type, levels = c("basal III", "basal II", "basal I", "basal IV", "stratum spinosum", "stratum granulosum")) #order cell identities
counts_KCmarkers$gene <- factor(counts_KCmarkers$gene, levels = c("KRT5", "KRT14", "KRT1", "KRT10"))

#the plot
ggplot(data = counts_KCmarkers, aes(x = cell_type, y = counts, group = gene)) +
  geom_line(aes(color = gene), linewidth = 1) + geom_point(aes(color = gene)) +
  scale_x_discrete(expand = c(0.04, 0), #gets rid of empty space beginning of x axis
                   labels = scales::label_wrap(10))  + #breaks long labels
  ylab("expression levels") +
  theme_classic() + theme(axis.title.x = element_blank()) +
  scale_color_manual(values = c("#6573A0", "#B0B9E6", "orange1", "orange3"))
```

# Differential expression analysis

```{r DEanalysis HC KC subtypes}
# diffex_hc <- readRDS(paste0(save.dir, "diffex_hc.rds"))

diffex_hc <- data.frame(gene = rownames(rojahn_kc))
diffex_hc$pct_global <- NA

for (i in levels(rojahn_kc$svm_annotation)){
  print(i)
  
  DefaultAssay(rojahn_kc) <- "RNA"
  Idents(rojahn_kc) <- rojahn_kc$svm_annotation
  markers <- FindMarkers(rojahn_kc[, rojahn_kc$state == "HC"], group.by = "svm_annotation", ident.1 = i) #DEanalysis of one subtype compared to the complete dataset, log2FC > 0: more expressed in subtype
  colnames(markers) <- paste("HC", i, colnames(markers), sep = "_")
  markers$gene <- rownames(markers)

  diffex_hc <- merge(x = diffex_hc, y = markers, by = "gene", all.x = TRUE, all.y = T)
  diffex_hc <- diffex_hc[ , -c(ncol(diffex_hc)-1, ncol(diffex_hc)-4)] #removes 2 columns
  
  # replaces all NAs in pct.global
  y <- diffex_hc[is.na(diffex_hc$pct_global), "gene"]
  for(gene_name in y){
    if(gene_name %in% markers$gene){ #if this gene is present in the differential expression comparison
      pct_global <- markers[markers$gene == gene_name, 4] #find in markers
      index <- which(diffex_hc$gene == gene_name) #find row in diffex_hc
      diffex_hc$pct_global[index] <- pct_global #replace NA there
    }
  }
}

# saveRDS(diffex_hc, file = paste0(save.dir, "diffex_hc.rds"))
# write.xlsx(diffex_hc, file = paste0(save.dir, "diffex_hc.xlsx"))
```

```{r DEanalysis AD KC subtypes}
# diffex_ad <- readRDS(paste0(save.dir, "diffex_ad.rds"))

diffex_ad <- data.frame(gene = rownames(rojahn_kc))
diffex_ad$pct_global <- NA

for (i in levels(rojahn_kc$svm_annotation)){
  print(i)
  
  DefaultAssay(rojahn_kc) <- "RNA"
  Idents(rojahn_kc) <- rojahn_kc$svm_annotation
  markers <- FindMarkers(rojahn_kc[, rojahn_kc$state == "AD"], group.by = "svm_annotation", ident.1 = i)
  # DEanalysis of one subtype compared to the complete dataset, log2FC > 0: more expressed in subtype
  colnames(markers) <- paste("AD", i, colnames(markers), sep = "_")
  markers$gene <- rownames(markers)

  diffex_ad <- merge(x = diffex_ad, y = markers, by = "gene", all.x = TRUE, all.y = T)
  diffex_ad <- diffex_ad[ , -c(ncol(diffex_ad)-1, ncol(diffex_ad)-4)] #removes 2 columns
  
  # replaces all NAs in pct.global
  y <- diffex_ad[is.na(diffex_ad$pct_global), "gene"]
  for(gene_name in y){
    if(gene_name %in% markers$gene){ #if this gene is present in the differential expression comparison
      pct_global <- markers[markers$gene == gene_name, 4] #find in markers
      index <- which(diffex_ad$gene == gene_name) #find row in diffex_hc
      diffex_ad$pct_global[index] <- pct_global #replace NA there
    }
  }
}

# saveRDS(diffex_ad, file = paste0(save.dir, "diffex_ad.rds"))
# write.xlsx(diffex_ad, file = paste0(save.dir, "diffex_ad.xlsx"))
```

```{r DEanalysis ADvsHC KC subtypes}
# diffex_ad_hc <- readRDS(paste0(save.dir, "diffex_ad_hc.rds"))

#first starting with a general is for (bulk) AD vs HC comparison
DefaultAssay(rojahn_kc) <- "RNA"
markers <- FindMarkers(rojahn_kc, group.by = "state", ident.1 = "AD", ident.2 = "HC") #comparison of changes in AD compared to HC, log2FC up means more expressed in first group (AD)

colnames(markers) <- paste("AD", colnames(markers), sep = "_")
markers$gene <- rownames(markers)

diffex_ad_hc <- data.frame(gene = rownames(markers))
diffex_ad_hc$pct_AD_global <- markers$AD_pct.1
diffex_ad_hc$pct_HC_global <- markers$AD_pct.2
diffex_ad_hc <- merge(x = diffex_ad_hc, y = markers, by = "gene", all.x = TRUE, all.y = T)
diffex_ad_hc <- diffex_ad_hc[ , -c(ncol(diffex_ad_hc)-1, ncol(diffex_ad_hc)-4)] #removes 2 columns
diffex_ad_hc <- diffex_ad_hc[ , -(5)] #removes AD.pct1 (because we have this already as AD global)

for (i in levels(rojahn_kc$svm_annotation)){
  print(i)
  
  DefaultAssay(rojahn_kc) <- "RNA"
  Idents(rojahn_kc) <- rojahn_kc$svm_annotation
  markers <- FindMarkers(rojahn_kc, group.by = "state", subset.ident = i, ident.1 = "AD", ident.2 = "HC") #comparison of changes in AD compared to HC, log2FC up means more expressed in first group (AD)
  colnames(markers) <- paste("ADvsHC", i, colnames(markers), sep = "_")
  names <- colnames(markers)
  names <- gsub(patter = "pct.1", replacement = "pct.AD", x = names) #replaces pattern in a string
  names <- gsub(patter = "pct.2", replacement = "pct.HC", x = names)
  colnames(markers) <- names
  markers$gene <- rownames(markers)
  
  diffex_ad_hc <- merge(x = diffex_ad_hc, y = markers, by = "gene", all.x = TRUE, all.y = T)
  diffex_ad_hc <- diffex_ad_hc[ , -c(ncol(diffex_ad_hc)-4)] #removes unadjusted p value column
}

# saveRDS(diffex_ad_hc, file = paste0(save.dir, "diffex_ad_hc.rds"))
# write.xlsx(diffex_ad_hc, file = paste0(save.dir, "diffex_ad_hc.xlsx"))
```

# Cell cycle analysis

```{r cell cycle analysis}
DefaultAssay(rojahn_kc) <- "RNA"
# cc.genes.updated.2019 #updated list of cell cycle marker genes
rojahn_kc <- CellCycleScoring(rojahn_kc, s.features = cc.genes.updated.2019$s.genes, g2m.features = cc.genes.updated.2019$g2m.genes, set.ident = TRUE) #adding CC identity to the object metaData
# the lists cc.genes.updated are part of the Seurat package, 2019 version is the latest update

Idents(rojahn_kc) <- "Phase"
DimPlot(rojahn_kc, group.by = "Phase", cols = colors_cellcycle, order = c("G2M", "S", "G1"))

#Figure 2c
DimPlot(rojahn_kc[ , rojahn_kc$state == "HC"], group.by = "Phase", cols = colors_cellcycle, order = c("G2M", "S", "G1")) + theme(plot.title = element_blank())

#Figure 3c
DimPlot(rojahn_kc[ , rojahn_kc$state == "AD"], group.by = "Phase", cols = colors_cellcycle, order = c("G2M", "S", "G1")) + theme(plot.title = element_blank()) + labs(color = "cell cycle\nphase")

# saveRDS(rojahn_kc, paste0(save.dir, "rojahn_kc.rds"))
```

Visualize the percentages of the cell cycle phases

```{r barplot CC percentages}
DefaultAssay(rojahn_kc) <- "RNA"
Idents(rojahn_kc) <- "Phase"

phase_percentage_HC  <- percentage_df(rojahn_kc[ , rojahn_kc$state == "HC"], cell_annotation = "Phase", variable = "svm_annotation", facet = F)

#reorder: 
phase_percentage_HC$Variable <- factor(phase_percentage_HC$Variable, levels = c("basal I", "basal II", "basal III", "basal IV", "stratum spinosum", "stratum granulosum"))
phase_percentage_HC$Celltype <- factor(phase_percentage_HC$Celltype, levels = c("G1", "S", "G2M")) 

ggplot(phase_percentage_HC, aes(fill = Celltype, y = percentage, x = Variable)) + 
    geom_bar(position = "stack", stat = "identity") +
    geom_text(aes(label = percentage), position = position_stack(vjust = 0.5), size = 3) + 
    scale_x_discrete(labels = scales::label_wrap(10)) + #breaks long labels
    scale_fill_manual(values = colors_cellcycle, name = "cell cycle\nphase") + 
    theme_classic() + ggtitle("HC") + 
    theme(text = element_text(size = 12),
          plot.title = element_text(hjust = 0.5),
          axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
          axis.title.x = element_blank())

#AD
phase_percentage_AD  <- percentage_df(rojahn_kc[, rojahn_kc$state == "AD"], cell_annotation = "Phase", variable = "svm_annotation", facet = F)

#reorder:
phase_percentage_AD$Variable <- factor(phase_percentage_AD$Variable, levels = c("basal I", "basal II", "basal III", "basal IV", "stratum spinosum", "stratum granulosum"))
phase_percentage_AD$Celltype <- factor(phase_percentage_AD$Celltype, levels = c("G1", "S", "G2M")) 

ggplot(phase_percentage_AD, aes(fill = Celltype, y = percentage, x = Variable)) + 
    geom_bar(position = "stack", stat = "identity") +
    geom_text(aes(label = percentage), position = position_stack(vjust = 0.5), size = 3) + 
    scale_x_discrete(labels = scales::label_wrap(10)) + #breaks long labels
    scale_fill_manual(values = colors_cellcycle, name = "cell cycle\nphase") + 
    theme_classic() + ggtitle("AD") + 
    theme(text = element_text(size = 10),
          plot.title = element_text(hjust = 0.5),
          axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
          axis.title.x = element_blank())
```

# Functional analysis 
## Functional analysis basal cell types HC

###Prepare gene lists

```{r annotate with entrezID}
# diffex_hc <- readRDS(paste0(save.dir, "diffex_hc.rds"))

#annotate all expressed background genes
background_symbol <- names(rowSums(rojahn_kc@assays$RNA$counts) > 3)

#annotate all expressed background genes
background <-  mapIds(
  org.Hs.eg.db,
  keys = background_symbol,
  column = "ENTREZID",
  keytype = "SYMBOL",
  multiVals = "first"
)

#annotate DE genes in the differential expression data frame
entrez_anno <- bitr(diffex_hc$gene, fromType = "SYMBOL",
        toType = c("ENTREZID"),
        OrgDb = org.Hs.eg.db)

diffex_hc <- merge(x = diffex_hc, y = entrez_anno, by.x = "gene", by.y = "SYMBOL", all.x = T)

diffex_hc <- diffex_hc %>% relocate(ENTREZID, .after = gene)
diffex_hc <- diffex_hc %>% rename(entrezID = ENTREZID)

#which genes do not have an entrezID
table(is.na(diffex_hc$entrezID))
diffex_hc$gene[is.na(diffex_hc$entrezID)]

# saveRDS(diffex_hc, file = paste0(save.dir, "diffex_hc.rds"))
```

```{r prepare input over-representation analysis}
#basal III
DEHC_basalIII <- diffex_hc[diffex_hc$`HC_basal III_p_val_adj` < 0.05 & !is.na(diffex_hc$`HC_basal III_p_val_adj`), "entrezID"]
DEHC_basalIIIup <- diffex_hc[diffex_hc$`HC_basal III_p_val_adj` < 0.05 & !is.na(diffex_hc$`HC_basal III_p_val_adj`) & diffex_hc$`HC_basal III_avg_log2FC` > 0, "entrezID"]
DEHC_basalIIIdown <- diffex_hc[diffex_hc$`HC_basal III_p_val_adj` < 0.05 & !is.na(diffex_hc$`HC_basal III_p_val_adj`) & diffex_hc$`HC_basal III_avg_log2FC` < 0, "entrezID"]

#basal II
DEHC_basalII <- diffex_hc[diffex_hc$`HC_basal II_p_val_adj` < 0.05 & !is.na(diffex_hc$`HC_basal II_p_val_adj`), "entrezID"]
DEHC_basalIIup <- diffex_hc[diffex_hc$`HC_basal II_p_val_adj` < 0.05 & !is.na(diffex_hc$`HC_basal II_p_val_adj`) & diffex_hc$`HC_basal II_avg_log2FC` > 0, "entrezID"]
DEHC_basalIIdown <- diffex_hc[diffex_hc$`HC_basal II_p_val_adj` < 0.05 & !is.na(diffex_hc$`HC_basal II_p_val_adj`) & diffex_hc$`HC_basal II_avg_log2FC` < 0, "entrezID"]

#basal I
DEHC_basalI <- diffex_hc[diffex_hc$`HC_basal I_p_val_adj` < 0.05 & !is.na(diffex_hc$`HC_basal I_p_val_adj`), "entrezID"]
DEHC_basalIup <- diffex_hc[diffex_hc$`HC_basal I_p_val_adj` < 0.05 & !is.na(diffex_hc$`HC_basal I_p_val_adj`) & diffex_hc$`HC_basal I_avg_log2FC` > 0, "entrezID"]
DEHC_basalIdown <- diffex_hc[diffex_hc$`HC_basal I_p_val_adj` < 0.05 & !is.na(diffex_hc$`HC_basal I_p_val_adj`) & diffex_hc$`HC_basal I_avg_log2FC` < 0, "entrezID"]

#basal IV
DEHC_basalIV <- diffex_hc[diffex_hc$`HC_basal IV_p_val_adj` < 0.05 & !is.na(diffex_hc$`HC_basal IV_p_val_adj`), "entrezID"]
DEHC_basalIVup <- diffex_hc[diffex_hc$`HC_basal IV_p_val_adj` < 0.05 & !is.na(diffex_hc$`HC_basal IV_p_val_adj`) & diffex_hc$`HC_basal IV_avg_log2FC` > 0, "entrezID"]
DEHC_basalIVdown <- diffex_hc[diffex_hc$`HC_basal IV_p_val_adj` < 0.05 & !is.na(diffex_hc$`HC_basal IV_p_val_adj`) & diffex_hc$`HC_basal IV_avg_log2FC` < 0, "entrezID"]

#spinosum
DEHC_spin <- diffex_hc[diffex_hc$`HC_stratum spinosum_p_val_adj` < 0.05 & !is.na(diffex_hc$`HC_stratum spinosum_p_val_adj`), "entrezID"]
DEHC_spinup <- diffex_hc[diffex_hc$`HC_stratum spinosum_p_val_adj` < 0.05 & !is.na(diffex_hc$`HC_stratum spinosum_p_val_adj`) & diffex_hc$`HC_stratum spinosum_avg_log2FC` > 0, "entrezID"]
DEHC_spindown <- diffex_hc[diffex_hc$`HC_stratum spinosum_p_val_adj` < 0.05 & !is.na(diffex_hc$`HC_stratum spinosum_p_val_adj`) & diffex_hc$`HC_stratum spinosum_avg_log2FC` < 0, "entrezID"]

#granulosum
DEHC_gran <- diffex_hc[diffex_hc$`HC_stratum granulosum_p_val_adj` < 0.05 & !is.na(diffex_hc$`HC_stratum granulosum_p_val_adj`), "entrezID"]
DEHC_granup <- diffex_hc[diffex_hc$`HC_stratum granulosum_p_val_adj` < 0.05 & !is.na(diffex_hc$`HC_stratum granulosum_p_val_adj`) & diffex_hc$`HC_stratum granulosum_avg_log2FC` > 0, "entrezID"]
DEHC_grandown <- diffex_hc[diffex_hc$`HC_stratum granulosum_p_val_adj` < 0.05 & !is.na(diffex_hc$`HC_stratum granulosum_p_val_adj`) & diffex_hc$`HC_stratum granulosum_avg_log2FC` < 0, "entrezID"]
```

```{r prepare input gsea}
Idents(rojahn_kc) <- rojahn_kc$state
DefaultAssay(rojahn_kc) <- "RNA"

#basal III
DEHC_basalIII <- FindMarkers(rojahn_kc, group.by = "svm_annotation", subset.ident = "HC", ident.1 = "basal III", logfc.threshold = 0.1) #logFC threshhold is 0.25 by default
DEHC_basalIII_gsea <- DEHC_basalIII[ , "avg_log2FC"]
names(DEHC_basalIII_gsea) <- mapIds(org.Hs.eg.db, keys = rownames(DEHC_basalIII), column = "ENTREZID", keytype = "SYMBOL", multiVals = "first")
DEHC_basalIII_gsea <- sort(DEHC_basalIII_gsea, decreasing = T)

#basal II
DEHC_basalII <- FindMarkers(rojahn_kc, group.by = "svm_annotation", subset.ident = "HC", ident.1 = "basal II", logfc.threshold = 0.1) #logFC threshhold is 0.25 by default
DEHC_basalII_gsea <- DEHC_basalII[ , "avg_log2FC"]
names(DEHC_basalII_gsea) <- mapIds(org.Hs.eg.db, keys = rownames(DEHC_basalII), column = "ENTREZID", keytype = "SYMBOL", multiVals = "first")
DEHC_basalII_gsea <- sort(DEHC_basalII_gsea, decreasing = T)

#basal I
DEHC_basalI <- FindMarkers(rojahn_kc, group.by = "svm_annotation", subset.ident = "HC", ident.1 = "basal I", logfc.threshold = 0.1) #logFC threshhold is 0.25 by default
DEHC_basalI_gsea <- DEHC_basalI[ , "avg_log2FC"]
names(DEHC_basalI_gsea) <- mapIds(org.Hs.eg.db, keys = rownames(DEHC_basalI), column = "ENTREZID", keytype = "SYMBOL", multiVals = "first")
DEHC_basalI_gsea <- sort(DEHC_basalI_gsea, decreasing = T)

#basal IV
DEHC_basalIV <- FindMarkers(rojahn_kc, group.by = "svm_annotation", subset.ident = "HC", ident.1 = "basal IV", logfc.threshold = 0.1) #logFC threshhold is 0.25 by default
DEHC_basalIV_gsea <- DEHC_basalIV[ , "avg_log2FC"]
names(DEHC_basalIV_gsea) <- mapIds(org.Hs.eg.db, keys = rownames(DEHC_basalIV), column = "ENTREZID", keytype = "SYMBOL", multiVals = "first")
DEHC_basalIV_gsea <- sort(DEHC_basalIV_gsea, decreasing = T)

#spinosum 
DEHC_spin <- FindMarkers(rojahn_kc, group.by = "svm_annotation", subset.ident = "HC", ident.1 = "stratum spinosum", logfc.threshold = 0.1) #logFC threshhold is 0.25 by default
DEHC_spin_gsea <- DEHC_spin[ , "avg_log2FC"]
names(DEHC_spin_gsea) <- mapIds(org.Hs.eg.db, keys = rownames(DEHC_spin), column = "ENTREZID", keytype = "SYMBOL", multiVals = "first")
DEHC_spin_gsea <- sort(DEHC_spin_gsea, decreasing = T)

#granulosum 
DEHC_gran <- FindMarkers(rojahn_kc, group.by = "svm_annotation", subset.ident = "HC", ident.1 = "stratum granulosum", logfc.threshold = 0.1) #logFC threshhold is 0.25 by default
DEHC_gran_gsea <- DEHC_gran[ , "avg_log2FC"]
names(DEHC_gran_gsea) <- mapIds(org.Hs.eg.db, keys = rownames(DEHC_gran), column = "ENTREZID", keytype = "SYMBOL", multiVals = "first")
DEHC_gran_gsea <- sort(DEHC_gran_gsea, decreasing = T)
```

###Gene ontology analysis

```{r GO over representation}
#read
DEHC_basalIII_GO <- readRDS(paste0(save.dir, "DEHC_basalIII_GO.rds"))
DEHC_basalII_GO <- readRDS(paste0(save.dir, "DEHC_basalII_GO.rds"))
DEHC_basalI_GO <- readRDS(paste0(save.dir, "DEHC_basalI_GO.rds"))
DEHC_basalIV_GO <- readRDS(paste0(save.dir, "DEHC_basalIV_GO.rds"))
DEHC_spin_GO <- readRDS(paste0(save.dir, "DEHC_spin_GO.rds"))
DEHC_gran_GO <- readRDS(paste0(save.dir, "DEHC_gran_GO.rds"))

DEHC_basalIIIup_GO <- readRDS(paste0(save.dir, "DEHC_basalIIIup_GO.rds"))
DEHC_basalIIup_GO <- readRDS(paste0(save.dir, "DEHC_basalIIup_GO.rds"))
DEHC_basalIup_GO <- readRDS(paste0(save.dir, "DEHC_basalIup_GO.rds"))
DEHC_basalIVup_GO <- readRDS(paste0(save.dir, "DEHC_basalIVup_GO.rds"))
DEHC_spinup_GO <- readRDS(paste0(save.dir, "DEHC_spinup_GO.rds"))
DEHC_granup_GO <- readRDS(paste0(save.dir, "DEHC_granup_GO.rds"))

DEHC_basalIIIdown_GO <- readRDS(paste0(save.dir, "DEHC_basalIIIdown_GO.rds"))
DEHC_basalIIdown_GO <- readRDS(paste0(save.dir, "DEHC_basalIIdown_GO.rds"))
DEHC_basalIdown_GO <- readRDS(paste0(save.dir, "DEHC_basalIdown_GO.rds"))
DEHC_basalIVdown_GO <- readRDS(paste0(save.dir, "DEHC_basalIVdown_GO.rds"))
DEHC_spindown_GO <- readRDS(paste0(save.dir, "DEHC_spindown_GO.rds"))
DEHC_grandown_GO <- readRDS(paste0(save.dir, "DEHC_grandown_GO.rds"))

#GO analysis
DEHC_basalIII_GO <-  enrichGO(gene = DEHC_basalIII, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEHC_basalIIIup_GO <-  enrichGO(gene = DEHC_basalIIIup, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEHC_basalIIIdown_GO <-  enrichGO(gene = DEHC_basalIIIdown, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)

DEHC_basalII_GO <-  enrichGO(gene = DEHC_basalII, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEHC_basalIIup_GO <-  enrichGO(gene = DEHC_basalIIup, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEHC_basalIIdown_GO <-  enrichGO(gene = DEHC_basalIIdown, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)

DEHC_basalI_GO <-  enrichGO(gene = DEHC_basalI, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEHC_basalIup_GO <-  enrichGO(gene = DEHC_basalIup, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEHC_basalIdown_GO <-  enrichGO(gene = DEHC_basalIdown, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)

DEHC_basalIV_GO <-  enrichGO(gene = DEHC_basalIV, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEHC_basalIVup_GO <-  enrichGO(gene = DEHC_basalIVup, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEHC_basalIVdown_GO <-  enrichGO(gene = DEHC_basalIVdown, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)

DEHC_spin_GO <-  enrichGO(gene = DEHC_spin, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEHC_spinup_GO <-  enrichGO(gene = DEHC_spinup, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEHC_spindown_GO <-  enrichGO(gene = DEHC_spindown, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)

DEHC_gran_GO <-  enrichGO(gene = DEHC_gran, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEHC_granup_GO <-  enrichGO(gene = DEHC_granup, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEHC_grandown_GO <-  enrichGO(gene = DEHC_grandown, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)

#save
# list <- list("GO DE HC basal III" = DEHC_basalIII_GO,
#              "GO DE HC basal IIIup" = DEHC_basalIIIup_GO,
#              "GO DE HC basal IIIdown" = DEHC_basalIIIdown_GO,
#              "GO DE HC basal II" = DEHC_basalII_GO,
#              "GO DE HC basal IIup" = DEHC_basalIIup_GO,
#              "GO DE HC basal IIdown" = DEHC_basalIIdown_GO,
#              "GO DE HC basal I" = DEHC_basalI_GO,
#              "GO DE HC basal Iup" = DEHC_basalIup_GO,
#              "GO DE HC basal Idown" = DEHC_basalIdown_GO,
#              "GO DE HC basal IV" = DEHC_basalIV_GO,
#              "GO DE HC basal IVup" = DEHC_basalIVup_GO,
#              "GO DE HC basal IVdown" = DEHC_basalIVdown_GO,
#              "GO DE HC spinosum" = DEHC_spin_GO,
#              "GO DE HC spinosumup" = DEHC_spinup_GO,
#              "GO DE HC spinosumdown" = DEHC_spindown_GO,
#              "GO DE HC granulosum" = DEHC_gran_GO,
#              "GO DE HC granulosumup" = DEHC_granup_GO,
#              "GO DE HC granulosumdown" = DEHC_grandown_GO)
# write.xlsx(list, file = paste0(save.dir, "GO DE HC.xlsx"))
# 
# saveRDS(DEHC_basalIII_GO, paste0(save.dir, "DEHC_basalIII_GO.rds"))
# saveRDS(DEHC_basalII_GO, paste0(save.dir, "DEHC_basalII_GO.rds"))
# saveRDS(DEHC_basalI_GO, paste0(save.dir, "DEHC_basalI_GO.rds"))
# saveRDS(DEHC_basalIV_GO, paste0(save.dir, "DEHC_basalIV_GO.rds"))
# saveRDS(DEHC_spin_GO, paste0(save.dir, "DEHC_spin_GO.rds"))
# saveRDS(DEHC_gran_GO, paste0(save.dir, "DEHC_gran_GO.rds"))
# 
# saveRDS(DEHC_basalIIIup_GO, paste0(save.dir, "DEHC_basalIIIup_GO.rds"))
# saveRDS(DEHC_basalIIup_GO, paste0(save.dir, "DEHC_basalIIup_GO.rds"))
# saveRDS(DEHC_basalIup_GO, paste0(save.dir, "DEHC_basalIup_GO.rds"))
# saveRDS(DEHC_basalIVup_GO, paste0(save.dir, "DEHC_basalIVup_GO.rds"))
# saveRDS(DEHC_spinup_GO, paste0(save.dir, "DEHC_spinup_GO.rds"))
# saveRDS(DEHC_granup_GO, paste0(save.dir, "DEHC_granup_GO.rds"))
# 
# saveRDS(DEHC_basalIIIdown_GO, paste0(save.dir, "DEHC_basalIIIdown_GO.rds"))
# saveRDS(DEHC_basalIIdown_GO, paste0(save.dir, "DEHC_basalIIdown_GO.rds"))
# saveRDS(DEHC_basalIdown_GO, paste0(save.dir, "DEHC_basalIdown_GO.rds"))
# saveRDS(DEHC_basalIVdown_GO, paste0(save.dir, "DEHC_basalIVdown_GO.rds"))
# saveRDS(DEHC_spindown_GO, paste0(save.dir, "DEHC_spindown_GO.rds"))
# saveRDS(DEHC_grandown_GO, paste0(save.dir, "DEHC_grandown_GO.rds"))

GO_HC_vis <- DEHC_basalIup_GO[c(5, 10, 44)] 
GO_HC_vis <- rbind(GO_HC_vis, DEHC_basalIIup_GO[c(1, 6, 32)])
GO_HC_vis <- rbind(GO_HC_vis, DEHC_basalIIIup_GO[c(23, 31, 33)])
GO_HC_vis <- rbind(GO_HC_vis, DEHC_basalIVup_GO[c(5, 12, 19)])

GO_HC_vis$celltype <- c(rep("basal I", 3), rep("basal II", 3), rep("basal III", 3), rep("basal IV", 3))
GO_HC_vis$celltype <- factor(GO_HC_vis$celltype, levels = c("basal I", "basal II", "basal III", "basal IV"))

#Figure 2d
ggplot(GO_HC_vis, aes(x = Count, y = Description)) +
        geom_col(aes(fill = p.adjust)) +
        facet_grid("celltype", scale = "free") +
        scale_fill_gradient(high = "#FAFAD2", low = "darkgoldenrod2", guide = guide_colourbar(reverse = TRUE)) +
        theme_bw(base_size = 13) + 
        ylab(NULL) + xlab("count") + labs(fill = "adjusted \np value") +
        theme(axis.text.y = element_text(size = 13))
```

```{r GO GSEA}
#read
DEHC_basalIII_gseaGO <- readRDS(paste0(save.dir, "DEHC_basalIII_gseaGO.rds"))
DEHC_basalIV_gseaGO <- readRDS(paste0(save.dir, "DEHC_basalIV_gseaGO.rds"))

#do GO GSEA analysis
#basal III
DEHC_basalIII_gseaGO <- gseGO(geneList = DEHC_basalIII_gsea, OrgDb = org.Hs.eg.db)
DEHC_basalIII_gseaGO <- setReadable(DEHC_basalIII_gseaGO, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#basal II
DEHC_basalII_gseaGO <- gseGO(geneList = DEHC_basalII_gsea, OrgDb = org.Hs.eg.db)
DEHC_basalII_gseaGO <- setReadable(DEHC_basalII_gseaGO, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#basal I
DEHC_basalI_gseaGO <- gseGO(geneList = DEHC_basalI_gsea, OrgDb = org.Hs.eg.db)
DEHC_basalI_gseaGO <- setReadable(DEHC_basalI_gseaGO, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#basal IV
DEHC_basalIV_gseaGO <- gseGO(geneList = DEHC_basalIV_gsea, OrgDb = org.Hs.eg.db)
DEHC_basalIV_gseaGO <- setReadable(DEHC_basalIV_gseaGO, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#spinosum
DEHC_spin_gseaGO <- gseGO(geneList = DEHC_spin_gsea, OrgDb = org.Hs.eg.db)
DEHC_spin_gseaGO <- setReadable(DEHC_spin_gseaGO, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#granulosum
DEHC_gran_gseaGO <- gseGO(geneList = DEHC_gran_gsea, OrgDb = org.Hs.eg.db)
DEHC_gran_gseaGO <- setReadable(DEHC_gran_gseaGO, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#save
# list <- list("GSEA GO HC basal III " = DEHC_basalIII_gseaGO,
#              "GSEA GO HC basal II" = DEHC_basalII_gseaGO,
#              "GSEA GO HC basal I" = DEHC_basalI_gseaGO,
#              "GSEA GO HC basal IV" = DEHC_basalIV_gseaGO,
#              "GSEA GO HC spinosum" = DEHC_spin_gseaGO,
#              "GSEA GO HC granulosum" = DEHC_gran_gseaGO)
# write.xlsx(list, file = paste0(save.dir, "GO GSEA HC logFC 0.1.xlsx"))
# 
# saveRDS(DEHC_basalIII_gseaGO, paste0(save.dir, "DEHC_basalIII_gseaGO.rds"))
# saveRDS(DEHC_basalII_gseaGO, paste0(save.dir, "DEHC_basalII_gseaGO.rds"))
# saveRDS(DEHC_basalI_gseaGO, paste0(save.dir, "DEHC_basalI_gseaGO.rds"))
# saveRDS(DEHC_basalIV_gseaGO, paste0(save.dir, "DEHC_basalIV_gseaGO.rds"))
# saveRDS(DEHC_spin_gseaGO, paste0(save.dir, "DEHC_spin_gseaGO.rds"))
# saveRDS(DEHC_gran_gseaGO, paste0(save.dir, "DEHC_gran_gseaGO"))

#visualization
#Supplemental figure 3b
GOgsea_HC_basalIII_vis <- DEHC_basalIII_gseaGO[c(1, 13, 21)] #be aware, excel has to be sorted with decreasing qvalue
GOgsea_HC_basalIII_vis <- rbind(GOgsea_HC_basalIII_vis, DEHC_basalIV_gseaGO[c(12)]) #be aware, excel has to be sorted with decreasing qvalue

GOgsea_HC_basalIII_vis$Description <- factor(GOgsea_HC_basalIII_vis$Description, levels = GOgsea_HC_basalIII_vis$Description)
GOgsea_HC_basalIII_vis$category <- c(rep("basal III", 3), rep("basal IV", 1)) 
GOgsea_HC_basalIII_vis$category <- factor(GOgsea_HC_basalIII_vis$category, levels = c("basal III", "basal IV"))

#for the labels
ggplot(GOgsea_HC_basalIII_vis, aes(x = NES, y = forcats::fct_rev(Description))) +
        geom_point(aes(color = p.adjust), size = 5) + #geom_dotplot stackratio = 1.5, dotsize = 3
        facet_grid("category", switch = "y", scale = "free_y", space = "free_y") +
        scale_color_gradient(high = "#FAFAD2", low = "darkgoldenrod2", guide = guide_colourbar(reverse = TRUE)) +
        theme_bw(base_size = 13) +  scale_y_discrete(labels = scales::label_wrap(38)) +
        ylab(NULL) + xlab("normalized enrichment score") + labs(fill = "adjusted \np value") + 
        scale_x_continuous(limits = c(1, 2.5)) + 
        theme(axis.text.y = element_text(size = 14), strip.text.y = element_text(size = 14), strip.placement = "outside")

#ridgeplot
gseaplot_basalIII <-
  ridgeplot(DEHC_basalIII_gseaGO, showCategory = DEHC_basalIII_gseaGO$Description[c(1, 13, 21)]) + 
  scale_x_continuous(limits = c(-9.5, 8)) + 
  scale_fill_gradient(limits = c(9.1e-06, 3.4e-08), high = c("#FAFAD2"), low = c("darkgoldenrod2"), trans = 'reverse', name = "adjusted\np value")

gseaplot_basalIV <- ridgeplot(DEHC_basalIV_gseaGO, showCategory = DEHC_basalIV_gseaGO$Description[c(12)]) +
  scale_x_continuous(limits = c(-9.5, 8)) +
  scale_fill_gradient(limits = c(9.1e-06, 3.4e-08), high = c("#FAFAD2"), low = c("darkgoldenrod2"), trans = 'reverse', name = "adjusted\np value") +
  xlab("normalized enrichment score")

ggpubr::ggarrange(gseaplot_basalIII, gseaplot_basalIV, ncol = 1, align = "v", heights = c(2, 1), common.legend = TRUE, legend = "right") #labels = c("basal III", "basal IV")
```

###KEGG pathway analysis

```{r KEGG over enrichment}
#read
DEHC_basalIII_kegg <- readRDS(paste0(save.dir, "DEHC_basalIII_kegg.rds"))
DEHC_basalII_kegg <- readRDS(paste0(save.dir, "DEHC_basalII_kegg.rds"))
DEHC_basalI_kegg <- readRDS(paste0(save.dir, "DEHC_basalI_kegg.rds"))
DEHC_basalIV_kegg <- readRDS(paste0(save.dir, "DEHC_basalIV_kegg.rds"))
DEHC_spin_kegg <- readRDS(paste0(save.dir, "DEHC_spin_kegg.rds"))
DEHC_gran_kegg <- readRDS(paste0(save.dir, "DEHC_gran_kegg.rds"))

DEHC_basalIIIup_kegg <- readRDS(paste0(save.dir, "DEHC_basalIIIup_kegg.rds"))
DEHC_basalIIup_kegg <- readRDS(paste0(save.dir, "DEHC_basalIIup_kegg.rds"))
DEHC_basalIup_kegg <- readRDS(paste0(save.dir, "DEHC_basalIup_kegg.rds"))
DEHC_basalIVup_kegg <- readRDS(paste0(save.dir, "DEHC_basalIVup_kegg.rds"))
DEHC_spinup_kegg <- readRDS(paste0(save.dir, "DEHC_spinup_kegg.rds"))
DEHC_granup_kegg <- readRDS(paste0(save.dir, "DEHC_granup_kegg.rds"))

DEHC_basalIIIdown_kegg <- readRDS(paste0(save.dir, "DEHC_basalIIIdown_kegg.rds"))
DEHC_basalIIdown_kegg <- readRDS(paste0(save.dir, "DEHC_basalIIdown_kegg.rds"))
DEHC_basalIdown_kegg <- readRDS(paste0(save.dir, "DEHC_basalIdown_kegg.rds"))
DEHC_basalIVdown_kegg <- readRDS(paste0(save.dir, "DEHC_basalIVdown_kegg.rds"))
DEHC_spindown_kegg <- readRDS(paste0(save.dir, "DEHC_spindown_kegg.rds"))
DEHC_grandown_kegg <- readRDS(paste0(save.dir, "DEHC_grandown_kegg.rds"))

#do KEGG enrichment analysis
DEHC_basalIII_kegg <- enrichKEGG(gene = DEHC_basalIII, universe = background, pvalueCutoff  = 0.05)
DEHC_basalIIIup_kegg <- enrichKEGG(gene = DEHC_basalIIIup, universe = background, pvalueCutoff  = 0.05)
DEHC_basalIIIdown_kegg <- enrichKEGG(gene = DEHC_basalIIIdown, universe = background, pvalueCutoff  = 0.05)

DEHC_basalII_kegg <- enrichKEGG(gene = DEHC_basalII, universe = background, pvalueCutoff  = 0.05)
DEHC_basalIIup_kegg <- enrichKEGG(gene = DEHC_basalIIup, universe = background, pvalueCutoff  = 0.05)
DEHC_basalIIdown_kegg <- enrichKEGG(gene = DEHC_basalIIdown, universe = background, pvalueCutoff  = 0.05)

DEHC_basalI_kegg <- enrichKEGG(gene = DEHC_basalI, universe = background, pvalueCutoff  = 0.05)
DEHC_basalIup_kegg <- enrichKEGG(gene = DEHC_basalIup, universe = background, pvalueCutoff  = 0.05)
DEHC_basalIdown_kegg <- enrichKEGG(gene = DEHC_basalIdown, universe = background, pvalueCutoff  = 0.05)

DEHC_basalIV_kegg <- enrichKEGG(gene = DEHC_basalIV, universe = background, pvalueCutoff  = 0.05)
DEHC_basalIVup_kegg <- enrichKEGG(gene = DEHC_basalIVup, universe = background, pvalueCutoff  = 0.05)
DEHC_basalIVdown_kegg <- enrichKEGG(gene = DEHC_basalIVdown, universe = background, pvalueCutoff  = 0.05)

DEHC_spin_kegg <- enrichKEGG(gene = DEHC_spin, universe = background, pvalueCutoff  = 0.05)
DEHC_spinup_kegg <- enrichKEGG(gene = DEHC_spinup, universe = background, pvalueCutoff  = 0.05)
DEHC_spindown_kegg <- enrichKEGG(gene = DEHC_spindown, universe = background, pvalueCutoff  = 0.05)

DEHC_gran_kegg <- enrichKEGG(gene = DEHC_gran, universe = background, pvalueCutoff  = 0.05)
DEHC_granup_kegg <- enrichKEGG(gene = DEHC_granup, universe = background, pvalueCutoff  = 0.05)
DEHC_grandown_kegg <- enrichKEGG(gene = DEHC_grandown, universe = background, pvalueCutoff  = 0.05)

# make result read.table
DEHC_basalIII_kegg <- setReadable(DEHC_basalIII_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEHC_basalIIIup_kegg <- setReadable(DEHC_basalIIIup_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEHC_basalIIIdown_kegg <- setReadable(DEHC_basalIIIdown_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

DEHC_basalII_kegg <- setReadable(DEHC_basalII_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEHC_basalIIup_kegg <- setReadable(DEHC_basalIIup_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEHC_basalIIdown_kegg <- setReadable(DEHC_basalIIdown_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

DEHC_basalI_kegg <- setReadable(DEHC_basalI_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEHC_basalIup_kegg <- setReadable(DEHC_basalIup_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEHC_basalIdown_kegg <- setReadable(DEHC_basalIdown_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

DEHC_basalIV_kegg <- setReadable(DEHC_basalIV_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEHC_basalIVup_kegg <- setReadable(DEHC_basalIVup_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEHC_basalIVdown_kegg <- setReadable(DEHC_basalIVdown_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

DEHC_spin_kegg <- setReadable(DEHC_spin_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEHC_spinup_kegg <- setReadable(DEHC_spinup_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEHC_spindown_kegg <- setReadable(DEHC_spindown_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

DEHC_gran_kegg <- setReadable(DEHC_gran_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEHC_granup_kegg <- setReadable(DEHC_granup_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEHC_grandown_kegg <- setReadable(DEHC_grandown_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#save
# list <- list("KEGG DE HC basal III" = DEHC_basalIII_kegg,
#              "KEGG DE HC basal IIIup" = DEHC_basalIIIup_kegg,
#              "KEGG DE HC basal IIIdown" = DEHC_basalIIIdown_kegg,
#              "KEGG DE HC basal II" = DEHC_basalII_kegg,
#              "KEGG DE HC basal IIup" = DEHC_basalIIup_kegg,
#              "KEGG DE HC basal IIdown" = DEHC_basalIIdown_kegg,
#              "KEGG DE HC basal I" = DEHC_basalI_kegg,
#              "KEGG DE HC basal Iup" = DEHC_basalIup_kegg,
#              "KEGG DE HC basal Idown" = DEHC_basalIdown_kegg,
#              "KEGG DE HC basal IV" = DEHC_basalIV_kegg,
#              "KEGG DE HC basal IVup" = DEHC_basalIVup_kegg,
#              "KEGG DE HC basal IVdown" = DEHC_basalIVdown_kegg,
#              "KEGG DE HC spinosum" = DEHC_spin_kegg,
#              "KEGG DE HC spinosum up" = DEHC_spinup_kegg,
#              "KEGG DE HC spinosum down" = DEHC_spindown_kegg,
#              "KEGG DE HC granulosum" = DEHC_gran_kegg,
#              "KEGG DE HC granulosum up" = DEHC_granup_kegg,
#              "KEGG DE HC granulosum down" = DEHC_grandown_kegg)
# write.xlsx(list, file = paste0(save.dir, "KEGG DE HC.xlsx"))
# 
# saveRDS(DEHC_basalIII_kegg, paste0(save.dir, "DEHC_basalIII_kegg.rds"))
# saveRDS(DEHC_basalII_kegg, paste0(save.dir, "DEHC_basalII_kegg.rds"))
# saveRDS(DEHC_basalI_kegg, paste0(save.dir, "DEHC_basalI_kegg.rds"))
# saveRDS(DEHC_basalIV_kegg, paste0(save.dir, "DEHC_basalIV_kegg.rds"))
# saveRDS(DEHC_spin_kegg, paste0(save.dir, "DEHC_spin_kegg.rds"))
# saveRDS(DEHC_gran_kegg, paste0(save.dir, "DEHC_gran_kegg.rds"))
# 
# saveRDS(DEHC_basalIIIup_kegg, paste0(save.dir, "DEHC_basalIIIup_kegg.rds"))
# saveRDS(DEHC_basalIIup_kegg, paste0(save.dir, "DEHC_basalIIup_kegg.rds"))
# saveRDS(DEHC_basalIup_kegg, paste0(save.dir, "DEHC_basalIup_kegg.rds"))
# saveRDS(DEHC_basalIVup_kegg, paste0(save.dir, "DEHC_basalIVup_kegg.rds"))
# saveRDS(DEHC_spinup_kegg, paste0(save.dir, "DEHC_spinup_kegg.rds"))
# saveRDS(DEHC_granup_kegg, paste0(save.dir, "DEHC_granup_kegg.rds"))
# 
# saveRDS(DEHC_basalIIIdown_kegg, paste0(save.dir, "DEHC_basalIIIdown_kegg.rds"))
# saveRDS(DEHC_basalIIdown_kegg, paste0(save.dir, "DEHC_basalIIdown_kegg.rds"))
# saveRDS(DEHC_basalIdown_kegg, paste0(save.dir, "DEHC_basalIdown_kegg.rds"))
# saveRDS(DEHC_basalIVdown_kegg, paste0(save.dir, "DEHC_basalIVdown_kegg.rds"))
# saveRDS(DEHC_spindown_kegg, paste0(save.dir, "DEHC_spindown_kegg.rds"))
# saveRDS(DEHC_grandown_kegg, paste0(save.dir, "DEHC_grandown_kegg.rds"))
```

```{r KEGG GSEA}
#read
DEHC_basalIII_gseaKEGG <- readRDS(paste0(save.dir, "DEHC_basalIII_gseaKEGG.rds"))
DEHC_basalIV_gseaKEGG <- readRDS(paste0(save.dir, "DEHC_basalIV_gseaKEGG.rds"))

#do kegg GSEA analysis
#basal III
DEHC_basalIII_gseaKEGG <- gseKEGG(geneList = DEHC_basalIII_gsea, organism = 'hsa', keyType = "kegg", pvalueCutoff = 0.05)
DEHC_basalIII_gseaKEGG <- setReadable(DEHC_basalIII_gseaKEGG, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#basal II
DEHC_basalII_gseaKEGG <- gseKEGG(geneList = DEHC_basalII_gsea, organism = 'hsa', keyType = "kegg", pvalueCutoff = 0.05)
DEHC_basalII_gseaKEGG <- setReadable(DEHC_basalII_gseaKEGG, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#basal I
DEHC_basalI_gseaKEGG <- gseKEGG(geneList = DEHC_basalI_gsea, organism = 'hsa', keyType = "kegg", pvalueCutoff = 0.05)
DEHC_basalI_gseaKEGG <- setReadable(DEHC_basalI_gseaKEGG, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#basal IV
DEHC_basalIV_gseaKEGG <- gseKEGG(geneList = DEHC_basalIV_gsea, organism = 'hsa', keyType = "kegg", pvalueCutoff = 0.05)
DEHC_basalIV_gseaKEGG <- setReadable(DEHC_basalIV_gseaKEGG, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#spinosum
DEHC_spin_gseaKEGG <- gseKEGG(geneList = DEHC_spin_gsea, organism = 'hsa', keyType = "kegg", pvalueCutoff = 0.05)
DEHC_spin_gseaKEGG <- setReadable(DEHC_spin_gseaKEGG, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#granulosum
DEHC_gran_gseaKEGG <- gseKEGG(geneList = DEHC_gran_gsea, organism = 'hsa', keyType = "kegg", pvalueCutoff = 0.05)
DEHC_gran_gseaKEGG <- setReadable(DEHC_gran_gseaKEGG, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#save
# list <- list("GSEA KEGG HC basal III " = DEHC_basalIII_gseaKEGG,
#              "GSEA KEGG HC basal II" = DEHC_basalII_gseaKEGG,
#              "GSEA KEGG HC basal I" = DEHC_basalI_gseaKEGG,
#              "GSEA KEGG HC basal IV" = DEHC_basalIV_gseaKEGG,
#              "GSEA KEGG HC spinosum" = DEHC_spin_gseaKEGG,
#              "GSEA KEGG HC granulosum" = DEHC_gran_gseaKEGG)
# write.xlsx(list, file = paste0(save.dir, "KEGG GSEA HC logFC 0.1.xlsx"))
# 
# saveRDS(DEHC_basalIII_gseaKEGG, paste0(save.dir, "DEHC_basalIII_gseaKEGG.rds"))
# saveRDS(DEHC_basalII_gseaKEGG, paste0(save.dir, "DEHC_basalII_gseaKEGG.rds"))
# saveRDS(DEHC_basalI_gseaKEGG, paste0(save.dir, "DEHC_basalI_gseaKEGG.rds"))
# saveRDS(DEHC_basalIV_gseaKEGG, paste0(save.dir, "DEHC_basalIV_gseaKEGG.rds"))
# saveRDS(DEHC_spin_gseaKEGG, paste0(save.dir, "DEHC_spin_gseaKEGG.rds"))
# saveRDS(DEHC_gran_gseaKEGG, paste0(save.dir, "DEHC_gran_gseaKEGG.rds"))

#visualization, Supplemental figure 3c
#the the labels
KEGGgsea_HC_basalIII_vis <- DEHC_basalIII_gseaKEGG[c(4, 7, 8)] #be aware, excel has to be sorted with decreasing qvalue
KEGGgsea_HC_basalIII_vis <- rbind(KEGGgsea_HC_basalIII_vis, DEHC_basalIV_gseaKEGG[c(3)]) #be aware, excel has to be sorted with decreasing qvalue

KEGGgsea_HC_basalIII_vis$Description <- factor(KEGGgsea_HC_basalIII_vis$Description)
KEGGgsea_HC_basalIII_vis$category <- c(rep("basal III", 3), rep("basal IV", 1)) 
KEGGgsea_HC_basalIII_vis$category <- factor(KEGGgsea_HC_basalIII_vis$category, levels = c("basal III", "basal IV"))

ggplot(KEGGgsea_HC_basalIII_vis, aes(x = NES, y = forcats::fct_rev(Description))) +
        geom_point(aes(color = p.adjust), size = 5) + #geom_dotplot stackratio = 1.5, dotsize = 3
        facet_grid("category", switch = "y", scale = "free_y", space = "free_y") +
        scale_color_gradient(high = "#FAFAD2", low = "darkgoldenrod2", guide = guide_colourbar(reverse = TRUE)) +
        theme_bw(base_size = 13) +  scale_y_discrete(labels = scales::label_wrap(38)) +
        ylab(NULL) + xlab("normalized enrichment score") + labs(fill = "adjusted \np value") + 
        scale_x_continuous(limits = c(1, 2.5)) + 
        theme(axis.text.y = element_text(size = 14), strip.text.y = element_text(size = 14), strip.placement = "outside")

#ridgeplot
gseaplot_basalIII <-
  ridgeplot(DEHC_basalIII_gseaKEGG, showCategory = DEHC_basalIII_gseaKEGG$Description[c(4, 7, 8)]) + 
  scale_x_continuous(limits = c(-6.7, 6.9)) + 
  scale_fill_gradient(high = c("#FAFAD2"), low = c("darkgoldenrod2"), trans = 'reverse', name = "adjusted\np value")

gseaplot_basalIV <- ridgeplot(DEHC_basalIV_gseaKEGG, showCategory = DEHC_basalIV_gseaKEGG$Description[c(3)]) +
  scale_x_continuous(limits = c(-6.7, 6.9)) +
  scale_fill_gradient(high = c("#FAFAD2"), low = c("darkgoldenrod2"), trans = 'reverse', name = "adjusted\np value") +
  xlab("normalized enrichment score")

ggpubr::ggarrange(gseaplot_basalIII, gseaplot_basalIV, ncol = 1, align = "v", heights = c(2, 1), common.legend = TRUE, legend = "right") #labels = c("basal III", "basal IV")
```

###Progeny pathway activity interference

We first perform progeny pathway activity interference only on HC samples. 

```{r progeny HC isolated}
rojahn_kc_hc <- rojahn_kc[ , rojahn_kc$state == "HC"]
Idents(rojahn_kc_hc) <-  paste0("HC_", rojahn_kc_hc$svm_annotation)

cell_order <- paste0("HC_", levels(rojahn_kc_hc$svm_annotation)) #cell_order has to be matching Idents()
# pathway_order <- c("JAK-STAT", "TNFa", "NFkB", "WNT", "Estrogen", "MAPK", "VEGF", "EGFR", "PI3K", "Trail", "Hypoxia", "p53","TGFb", "Androgen")
pathway_order <- c("TNFa", "NFkB", "TGFb","JAK-STAT",  "WNT", "Estrogen", "MAPK", "PI3K", "EGFR", "Trail", "VEGF", "Androgen", "Hypoxia", "p53")

progeny_heatmap(rojahn_kc_hc, nfeatures = 500, cell_order = cell_order, clustering = F, pathway_order = pathway_order)
```

The results look valid, yet we have to keep in mind that they are normalized within themselves e.g. we see JAK-STAT and other inflammatory pathways activated in HC basal cells. 
For interpretation, it might be necessary to put the results into a bigger context, e.g. by including AD samples in the calculation. Which is what we do in the following chunk but for now we visualize only HC cells.

```{r progeny HC in relation}
Idents(rojahn_kc) <-  paste0(rojahn_kc$state,"_",rojahn_kc$svm_annotation)

cell_order <- c(paste0(levels(rojahn_kc$state)[1], "_", levels(rojahn_kc$svm_annotation)))
pathway_order <- c("JAK-STAT", "TNFa", "NFkB", "WNT", "Estrogen", "MAPK", "EGFR", "PI3K", "TGFb", "Trail", "VEGF", "Androgen", "Hypoxia", "p53")

#supplemental figure 3d
progeny_heatmap(rojahn_kc, nfeatures = 500, cell_order = cell_order, clustering = F, pathway_order = pathway_order)
```

## Functional analysis basal cell types AD

```{r annotate with entrezID}
# diffex_ad <- readRDS(paste0(save.dir, "diffex_ad.rds"))

#annotate all expressed background genes
background_symbol <- names(rowSums(rojahn_kc@assays$RNA$counts) > 3)

#annotate all expressed background genes
background <-  mapIds(
  org.Hs.eg.db,
  keys = background_symbol,
  column = "ENTREZID",
  keytype = "SYMBOL",
  multiVals = "first"
)

#annotate DE genes in the differential expression data frame
entrez_anno <- bitr(diffex_ad$gene, fromType = "SYMBOL",
        toType = c("ENTREZID"),
        OrgDb = org.Hs.eg.db)

diffex_ad <- merge(x = diffex_ad, y = entrez_anno, by.x = "gene", by.y = "SYMBOL", all.x = T)

diffex_ad <- diffex_ad %>% relocate(ENTREZID, .after = gene)
diffex_ad <- diffex_ad %>% rename(entrezID = ENTREZID)

#which genes do not have an entrezID
# table(is.na(diffex_ad$entrezID))
# diffex_ad$gene[is.na(diffex_ad$entrezID)]

# saveRDS(diffex_ad, file = paste0(save.dir, "diffex_ad.rds"))
```

```{r prepare input over-representation analysis}
#basal III
DEAD_basalIII <- diffex_ad[diffex_ad$`AD_basal III_p_val_adj` < 0.05 & !is.na(diffex_ad$`AD_basal III_p_val_adj`), "entrezID"]
DEAD_basalIIIup <- diffex_ad[diffex_ad$`AD_basal III_p_val_adj` < 0.05 & !is.na(diffex_ad$`AD_basal III_p_val_adj`) & diffex_ad$`AD_basal III_avg_log2FC` > 0, "entrezID"]
DEAD_basalIIIdown <- diffex_ad[diffex_ad$`AD_basal III_p_val_adj` < 0.05 & !is.na(diffex_ad$`AD_basal III_p_val_adj`) & diffex_ad$`AD_basal III_avg_log2FC` < 0, "entrezID"]

#basal II
DEAD_basalII <- diffex_ad[diffex_ad$`AD_basal II_p_val_adj` < 0.05 & !is.na(diffex_ad$`AD_basal II_p_val_adj`), "entrezID"]
DEAD_basalIIup <- diffex_ad[diffex_ad$`AD_basal II_p_val_adj` < 0.05 & !is.na(diffex_ad$`AD_basal II_p_val_adj`) & diffex_ad$`AD_basal II_avg_log2FC` > 0, "entrezID"]
DEAD_basalIIdown <- diffex_ad[diffex_ad$`AD_basal II_p_val_adj` < 0.05 & !is.na(diffex_ad$`AD_basal II_p_val_adj`) & diffex_ad$`AD_basal II_avg_log2FC` < 0, "entrezID"]

#basal I
DEAD_basalI <- diffex_ad[diffex_ad$`AD_basal I_p_val_adj` < 0.05 & !is.na(diffex_ad$`AD_basal I_p_val_adj`), "entrezID"]
DEAD_basalIup <- diffex_ad[diffex_ad$`AD_basal I_p_val_adj` < 0.05 & !is.na(diffex_ad$`AD_basal I_p_val_adj`) & diffex_ad$`AD_basal I_avg_log2FC` > 0, "entrezID"]
DEAD_basalIdown <- diffex_ad[diffex_ad$`AD_basal I_p_val_adj` < 0.05 & !is.na(diffex_ad$`AD_basal I_p_val_adj`) & diffex_ad$`AD_basal I_avg_log2FC` < 0, "entrezID"]

#basal IV
DEAD_basalIV <- diffex_ad[diffex_ad$`AD_basal IV_p_val_adj` < 0.05 & !is.na(diffex_ad$`AD_basal IV_p_val_adj`), "entrezID"]
DEAD_basalIVup <- diffex_ad[diffex_ad$`AD_basal IV_p_val_adj` < 0.05 & !is.na(diffex_ad$`AD_basal IV_p_val_adj`) & diffex_ad$`AD_basal IV_avg_log2FC` > 0, "entrezID"]
DEAD_basalIVdown <- diffex_ad[diffex_ad$`AD_basal IV_p_val_adj` < 0.05 & !is.na(diffex_ad$`AD_basal IV_p_val_adj`) & diffex_ad$`AD_basal IV_avg_log2FC` < 0, "entrezID"]

#spinosum
DEAD_spin <- diffex_ad[diffex_ad$`AD_stratum spinosum_p_val_adj` < 0.05 & !is.na(diffex_ad$`AD_stratum spinosum_p_val_adj`), "entrezID"]
DEAD_spinup <- diffex_ad[diffex_ad$`AD_stratum spinosum_p_val_adj` < 0.05 & !is.na(diffex_ad$`AD_stratum spinosum_p_val_adj`) & diffex_ad$`AD_stratum spinosum_avg_log2FC` > 0, "entrezID"]
DEAD_spindown <- diffex_ad[diffex_ad$`AD_stratum spinosum_p_val_adj` < 0.05 & !is.na(diffex_ad$`AD_stratum spinosum_p_val_adj`) & diffex_ad$`AD_stratum spinosum_avg_log2FC` < 0, "entrezID"]

#granulosum
DEAD_gran <- diffex_ad[diffex_ad$`AD_stratum granulosum_p_val_adj` < 0.05 & !is.na(diffex_ad$`AD_stratum granulosum_p_val_adj`), "entrezID"]
DEAD_granup <- diffex_ad[diffex_ad$`AD_stratum granulosum_p_val_adj` < 0.05 & !is.na(diffex_ad$`AD_stratum granulosum_p_val_adj`) & diffex_ad$`AD_stratum granulosum_avg_log2FC` > 0, "entrezID"]
DEAD_grandown <- diffex_ad[diffex_ad$`AD_stratum granulosum_p_val_adj` < 0.05 & !is.na(diffex_ad$`AD_stratum granulosum_p_val_adj`) & diffex_ad$`AD_stratum granulosum_avg_log2FC` < 0, "entrezID"]
```

```{r GO over-representation}
DEAD_basalIII_GO <-  enrichGO(gene = DEAD_basalIII, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEAD_basalIIIup_GO <-  enrichGO(gene = DEAD_basalIIIup, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEAD_basalIIIdown_GO <-  enrichGO(gene = DEAD_basalIIIdown, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)

DEAD_basalII_GO <-  enrichGO(gene = DEAD_basalII, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEAD_basalIIup_GO <-  enrichGO(gene = DEAD_basalIIup, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEAD_basalIIdown_GO <-  enrichGO(gene = DEAD_basalIIdown, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)

DEAD_basalI_GO <-  enrichGO(gene = DEAD_basalI, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEAD_basalIup_GO <-  enrichGO(gene = DEAD_basalIup, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEAD_basalIdown_GO <-  enrichGO(gene = DEAD_basalIdown, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)

DEAD_basalIV_GO <-  enrichGO(gene = DEAD_basalIV, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEAD_basalIVup_GO <-  enrichGO(gene = DEAD_basalIVup, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEAD_basalIVdown_GO <-  enrichGO(gene = DEAD_basalIVdown, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)

DEAD_spin_GO <-  enrichGO(gene = DEAD_spin, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEAD_spinup_GO <-  enrichGO(gene = DEAD_spinup, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEAD_spindown_GO <-  enrichGO(gene = DEAD_spindown, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)

DEAD_gran_GO <-  enrichGO(gene = DEAD_gran, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEAD_granup_GO <-  enrichGO(gene = DEAD_granup, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEAD_grandown_GO <-  enrichGO(gene = DEAD_grandown, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)

#save
# list <- list("GO DE AD basal III" = DEAD_basalIII_GO,
#              "GO DE AD basal IIIup" = DEAD_basalIIIup_GO,
#              "GO DE AD basal IIIdown" = DEAD_basalIIIdown_GO,
#              "GO DE AD basal II" = DEAD_basalII_GO,
#              "GO DE AD basal IIup" = DEAD_basalIIup_GO,
#              "GO DE AD basal IIdown" = DEAD_basalIIdown_GO,
#              "GO DE AD basal I" = DEAD_basalI_GO,
#              "GO DE AD basal Iup" = DEAD_basalIup_GO,
#              "GO DE AD basal Idown" = DEAD_basalIdown_GO,
#              "GO DE AD basal IV" = DEAD_basalIV_GO,
#              "GO DE AD basal IVup" = DEAD_basalIVup_GO,
#              "GO DE AD basal IVdown" = DEAD_basalIVdown_GO,
#              "GO DE AD spinosum" = DEAD_spin_GO,
#              "GO DE AD spinosumup" = DEAD_spinup_GO,
#              "GO DE AD spinosumdown" = DEAD_spindown_GO,
#              "GO DE AD granulosum" = DEAD_gran_GO,
#              "GO DE AD granulosumup" = DEAD_granup_GO,
#              "GO DE AD granulosumdown" = DEAD_grandown_GO)
# write.xlsx(list, file = paste0(save.dir, "GO DE AD.xlsx"))
```

We first perform progeny pathway activity interference only on AD samples. 

```{r progeny AD isolated}
rojahn_kc_ad <- rojahn_kc[ , rojahn_kc$state == "AD"]
Idents(rojahn_kc_ad) <-  paste0("AD_", rojahn_kc_ad$svm_annotation)

cell_order <- paste0("AD_", levels(rojahn_kc_ad$svm_annotation)) #cell_order has to be matching Idents()
pathway_order <- c("JAK-STAT", "TNFa", "NFkB", "WNT", "Estrogen", "MAPK", "VEGF", "EGFR", "PI3K", "Trail", "Hypoxia", "p53","TGFb", "Androgen")

progeny_heatmap(rojahn_kc_ad, nfeatures = 500, cell_order = cell_order, clustering = T, pathway_order = pathway_order)
```

Similarly here, the results look valid but are out of context e.g. JAK-STAT and inflammatory pathways are low in most cells. We therefore also therefore perform pathway activity interference together with the HC samples but look at only AD samples for now. 

```{r progeny AD in relation}
Idents(rojahn_kc) <-  paste0(rojahn_kc$state,"_",rojahn_kc$svm_annotation)

cell_order <- c(paste0(levels(rojahn_kc$state)[2], "_", levels(rojahn_kc$svm_annotation)))
pathway_order <- c("JAK-STAT", "TNFa", "NFkB", "WNT", "Estrogen", "MAPK", "EGFR", "PI3K", "TGFb", "Trail", "VEGF", "Androgen", "Hypoxia", "p53")

progeny_heatmap(rojahn_kc, nfeatures = 500, cell_order = cell_order, clustering = F, pathway_order = pathway_order)
```

## Functional analysis AD vs HC
### Prepare gene lists

```{r annotate with entrezID}
# diffex_ad_hc <- readRDS(paste0(save.dir, "diffex_ad_hc.rds"))

background_symbol <- names(rowSums(rojahn_kc@assays$RNA$counts) > 3)

#annotate all expressed background genes
background <-  mapIds(
  org.Hs.eg.db,
  keys = background_symbol,
  column = "ENTREZID",
  keytype = "SYMBOL",
  multiVals = "first"
)

#annotate DE genes in the differential expression data frame
entrez_anno <- bitr(diffex_ad_hc$gene, fromType = "SYMBOL",
        toType = c("ENTREZID"),
        OrgDb = org.Hs.eg.db)

diffex_ad_hc <- merge(x = diffex_ad_hc, y = entrez_anno, by.x = "gene", by.y = "SYMBOL", all.x = T)

diffex_ad_hc <- diffex_ad_hc %>% relocate(ENTREZID, .after = gene)
diffex_ad_hc <- diffex_ad_hc %>% rename(entrezID = ENTREZID)

table(is.na(diffex_ad_hc$entrezID))

# saveRDS(diffex_ad_hc, file = paste0(save.dir, "diffex_ad_hc.rds"))
```

```{r prepare input over-representation analysis}
#basal III
DEADvsHC_basalIII <- diffex_ad_hc[diffex_ad_hc$`ADvsHC_basal III_p_val_adj` < 0.05 & !is.na(diffex_ad_hc$`ADvsHC_basal III_p_val_adj`), "entrezID"]
DEADvsHC_basalIIIup <- diffex_ad_hc[diffex_ad_hc$`ADvsHC_basal III_p_val_adj` < 0.05 & !is.na(diffex_ad_hc$`ADvsHC_basal III_p_val_adj`) & diffex_ad_hc$`ADvsHC_basal III_avg_log2FC` > 0, "entrezID"]
DEADvsHC_basalIIIdown <- diffex_ad_hc[diffex_ad_hc$`ADvsHC_basal III_p_val_adj` < 0.05 & !is.na(diffex_ad_hc$`ADvsHC_basal III_p_val_adj`) & diffex_ad_hc$`ADvsHC_basal III_avg_log2FC` < 0, "entrezID"]

#basal II
DEADvsHC_basalII <- diffex_ad_hc[diffex_ad_hc$`ADvsHC_basal II_p_val_adj` < 0.05 & !is.na(diffex_ad_hc$`ADvsHC_basal II_p_val_adj`), "entrezID"]
DEADvsHC_basalIIup <- diffex_ad_hc[diffex_ad_hc$`ADvsHC_basal II_p_val_adj` < 0.05 & !is.na(diffex_ad_hc$`ADvsHC_basal II_p_val_adj`) & diffex_ad_hc$`ADvsHC_basal II_avg_log2FC` > 0, "entrezID"]
DEADvsHC_basalIIdown <- diffex_ad_hc[diffex_ad_hc$`ADvsHC_basal II_p_val_adj` < 0.05 & !is.na(diffex_ad_hc$`ADvsHC_basal II_p_val_adj`) & diffex_ad_hc$`ADvsHC_basal II_avg_log2FC` < 0, "entrezID"]

#basal I
DEADvsHC_basalI <- diffex_ad_hc[diffex_ad_hc$`ADvsHC_basal I_p_val_adj` < 0.05 & !is.na(diffex_ad_hc$`ADvsHC_basal I_p_val_adj`), "entrezID"]
DEADvsHC_basalIup <- diffex_ad_hc[diffex_ad_hc$`ADvsHC_basal I_p_val_adj` < 0.05 & !is.na(diffex_ad_hc$`ADvsHC_basal I_p_val_adj`) & diffex_ad_hc$`ADvsHC_basal I_avg_log2FC` > 0, "entrezID"]
DEADvsHC_basalIdown <- diffex_ad_hc[diffex_ad_hc$`ADvsHC_basal I_p_val_adj` < 0.05 & !is.na(diffex_ad_hc$`ADvsHC_basal I_p_val_adj`) & diffex_ad_hc$`ADvsHC_basal I_avg_log2FC` < 0, "entrezID"]

#basal IV
DEADvsHC_basalIV <- diffex_ad_hc[diffex_ad_hc$`ADvsHC_basal IV_p_val_adj` < 0.05 & !is.na(diffex_ad_hc$`ADvsHC_basal IV_p_val_adj`), "entrezID"]
DEADvsHC_basalIVup <- diffex_ad_hc[diffex_ad_hc$`ADvsHC_basal IV_p_val_adj` < 0.05 & !is.na(diffex_ad_hc$`ADvsHC_basal IV_p_val_adj`) & diffex_ad_hc$`ADvsHC_basal IV_avg_log2FC` > 0, "entrezID"]
DEADvsHC_basalIVdown <- diffex_ad_hc[diffex_ad_hc$`ADvsHC_basal IV_p_val_adj` < 0.05 & !is.na(diffex_ad_hc$`ADvsHC_basal IV_p_val_adj`) & diffex_ad_hc$`ADvsHC_basal IV_avg_log2FC` < 0, "entrezID"]

#spinosum
DEADvsHC_spin <- diffex_ad_hc[diffex_ad_hc$`ADvsHC_stratum spinosum_p_val_adj` < 0.05 & !is.na(diffex_ad_hc$`ADvsHC_stratum spinosum_p_val_adj`), "entrezID"]
DEADvsHC_spinup <- diffex_ad_hc[diffex_ad_hc$`ADvsHC_stratum spinosum_p_val_adj` < 0.05 & !is.na(diffex_ad_hc$`ADvsHC_stratum spinosum_p_val_adj`) & diffex_ad_hc$`ADvsHC_stratum spinosum_avg_log2FC` > 0, "entrezID"]
DEADvsHC_spindown <- diffex_ad_hc[diffex_ad_hc$`ADvsHC_stratum spinosum_p_val_adj` < 0.05 & !is.na(diffex_ad_hc$`ADvsHC_stratum spinosum_p_val_adj`) & diffex_ad_hc$`ADvsHC_stratum spinosum_avg_log2FC` < 0, "entrezID"]

#granulosum
DEADvsHC_gran <- diffex_ad_hc[diffex_ad_hc$`ADvsHC_stratum granulosum_p_val_adj` < 0.05 & !is.na(diffex_ad_hc$`ADvsHC_stratum granulosum_p_val_adj`), "entrezID"]
DEADvsHC_granup <- diffex_ad_hc[diffex_ad_hc$`ADvsHC_stratum granulosum_p_val_adj` < 0.05 & !is.na(diffex_ad_hc$`ADvsHC_stratum granulosum_p_val_adj`) & diffex_ad_hc$`ADvsHC_stratum granulosum_avg_log2FC` > 0, "entrezID"]
DEADvsHC_grandown <- diffex_ad_hc[diffex_ad_hc$`ADvsHC_stratum granulosum_p_val_adj` < 0.05 & !is.na(diffex_ad_hc$`ADvsHC_stratum granulosum_p_val_adj`) & diffex_ad_hc$`ADvsHC_stratum granulosum_avg_log2FC` < 0, "entrezID"]
```

```{r prepare input gsea}
Idents(rojahn_kc) <- rojahn_kc$svm_annotation

#basal III
DEADvsHC_basalIII <- FindMarkers(rojahn_kc, group.by = "state", subset.ident = "basal III", ident.1 = "AD", ident.2 = "HC", logfc.threshold = 0.1) #logFC threshhold is 0.25 by default
DEADvsHC_basalIII_gsea <- DEADvsHC_basalIII[ , "avg_log2FC"]
names(DEADvsHC_basalIII_gsea) <- mapIds(org.Hs.eg.db, keys = rownames(DEADvsHC_basalIII), column = "ENTREZID", keytype = "SYMBOL", multiVals = "first")
DEADvsHC_basalIII_gsea <- sort(DEADvsHC_basalIII_gsea, decreasing = T)

#basal II
DEADvsHC_basalII <- FindMarkers(rojahn_kc, group.by = "state", subset.ident = "basal II", ident.1 = "AD", ident.2 = "HC", logfc.threshold = 0.1) #logFC threshhold is 0.25 by default
DEADvsHC_basalII_gsea <- DEADvsHC_basalII[ , "avg_log2FC"]
names(DEADvsHC_basalII_gsea) <- mapIds(org.Hs.eg.db, keys = rownames(DEADvsHC_basalII), column = "ENTREZID", keytype = "SYMBOL", multiVals = "first")
DEADvsHC_basalII_gsea <- sort(DEADvsHC_basalII_gsea, decreasing = T)

#basal I
DEADvsHC_basalI <- FindMarkers(rojahn_kc, group.by = "state", subset.ident = "basal I", ident.1 = "AD", ident.2 = "HC", logfc.threshold = 0.1) #logFC threshhold is 0.25 by default
DEADvsHC_basalI_gsea <- DEADvsHC_basalI[ , "avg_log2FC"]
names(DEADvsHC_basalI_gsea) <- mapIds(org.Hs.eg.db, keys = rownames(DEADvsHC_basalI), column = "ENTREZID", keytype = "SYMBOL", multiVals = "first")
DEADvsHC_basalI_gsea <- sort(DEADvsHC_basalI_gsea, decreasing = T)

#basal IV
DEADvsHC_basalIV <- FindMarkers(rojahn_kc, group.by = "state", subset.ident = "basal IV", ident.1 = "AD", ident.2 = "HC", logfc.threshold = 0.1) #logFC threshhold is 0.25 by default
DEADvsHC_basalIV_gsea <- DEADvsHC_basalIV[ , "avg_log2FC"]
names(DEADvsHC_basalIV_gsea) <- mapIds(org.Hs.eg.db, keys = rownames(DEADvsHC_basalIV), column = "ENTREZID", keytype = "SYMBOL", multiVals = "first")
DEADvsHC_basalIV_gsea <- sort(DEADvsHC_basalIV_gsea, decreasing = T)

#spinosum 
DEADvsHC_spin <- FindMarkers(rojahn_kc, group.by = "state", subset.ident = "stratum spinosum", ident.1 = "AD", ident.2 = "HC", logfc.threshold = 0.1) #logFC threshhold is 0.25 by default
DEADvsHC_spin_gsea <- DEADvsHC_spin[ , "avg_log2FC"]
names(DEADvsHC_spin_gsea) <- mapIds(org.Hs.eg.db, keys = rownames(DEADvsHC_spin), column = "ENTREZID", keytype = "SYMBOL", multiVals = "first")
DEADvsHC_spin_gsea <- sort(DEADvsHC_spin_gsea, decreasing = T)

#granulosum 
DEADvsHC_gran <- FindMarkers(rojahn_kc, group.by = "state", subset.ident = "stratum granulosum", ident.1 = "AD", ident.2 = "HC", logfc.threshold = 0.1) #logFC threshhold is 0.25 by default
DEADvsHC_gran_gsea <- DEADvsHC_gran[ , "avg_log2FC"]
names(DEADvsHC_gran_gsea) <- mapIds(org.Hs.eg.db, keys = rownames(DEADvsHC_gran), column = "ENTREZID", keytype = "SYMBOL", multiVals = "first")
DEADvsHC_gran_gsea <- sort(DEADvsHC_gran_gsea, decreasing = T)
```

### GO analysis

```{r GO over-representation}
####read####
DEADvsHC_basalIII_GO <- readRDS(paste0(save.dir, "DEADvsHC_basalIII_GO.rds"))
DEADvsHC_basalIIIup_GO <- readRDS(paste0(save.dir, "DEADvsHC_basalIIIup_GO.rds"))
DEADvsHC_basalIIIdown_GO <- readRDS(paste0(save.dir, "DEADvsHC_basalIIIdown_GO.rds"))
DEADvsHC_basalII_GO <- readRDS(paste0(save.dir, "DEADvsHC_basalII_GO.rds"))
DEADvsHC_basalIIup_GO <- readRDS(paste0(save.dir, "DEADvsHC_basalIIup_GO.rds"))
DEADvsHC_basalIIdown_GO <- readRDS(paste0(save.dir, "DEADvsHC_basalIIdown_GO.rds"))
DEADvsHC_basalI_GO <- readRDS(paste0(save.dir, "DEADvsHC_basalI_GO.rds"))
DEADvsHC_basalIup_GO <- readRDS(paste0(save.dir, "DEADvsHC_basalIup_GO.rds"))
DEADvsHC_basalIdown_GO <- readRDS(paste0(save.dir, "DEADvsHC_basalIdown_GO.rds"))
DEADvsHC_basalIV_GO <- readRDS(paste0(save.dir, "DEADvsHC_basalIV_GO.rds"))
DEADvsHC_basalIVup_GO <- readRDS(paste0(save.dir, "DEADvsHC_basalIVup_GO.rds"))
DEADvsHC_basalIVdown_GO <- readRDS(paste0(save.dir, "DEADvsHC_basalIVdown_GO.rds"))
DEADvsHC_spin_GO <- readRDS(paste0(save.dir, "DEADvsHC_spin_GO.rds"))
DEADvsHC_spinup_GO <- readRDS(paste0(save.dir, "DEADvsHC_spinup_GO.rds"))
DEADvsHC_spindown_GO <- readRDS(paste0(save.dir, "DEADvsHC_spindown_GO.rds"))
DEADvsHC_gran_GO <- readRDS(paste0(save.dir, "DEADvsHC_gran_GO.rds"))
DEADvsHC_granup_GO <- readRDS(paste0(save.dir, "DEADvsHC_granup_GO.rds"))
DEADvsHC_grandown_GO <- readRDS(paste0(save.dir, "DEADvsHC_grandown_GO.rds"))

####calculate####
DEADvsHC_basalIII_GO <- enrichGO(gene = DEADvsHC_basalIII, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEADvsHC_basalIIIup_GO <- enrichGO(gene = DEADvsHC_basalIIIup, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEADvsHC_basalIIIdown_GO <- enrichGO(gene = DEADvsHC_basalIIIdown, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)

DEADvsHC_basalII_GO <-  enrichGO(gene = DEADvsHC_basalII, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEADvsHC_basalIIup_GO <-  enrichGO(gene = DEADvsHC_basalIIup, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEADvsHC_basalIIdown_GO <-  enrichGO(gene = DEADvsHC_basalIIdown, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)

DEADvsHC_basalI_GO <-  enrichGO(gene = DEADvsHC_basalI, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEADvsHC_basalIup_GO <-  enrichGO(gene = DEADvsHC_basalIup, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEADvsHC_basalIdown_GO <-  enrichGO(gene = DEADvsHC_basalIdown, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)

DEADvsHC_basalIV_GO <-  enrichGO(gene = DEADvsHC_basalIV, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEADvsHC_basalIVup_GO <-  enrichGO(gene = DEADvsHC_basalIVup, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEADvsHC_basalIVdown_GO <-  enrichGO(gene = DEADvsHC_basalIVdown, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)

DEADvsHC_spin_GO <-  enrichGO(gene = DEADvsHC_spin, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEADvsHC_spinup_GO <-  enrichGO(gene = DEADvsHC_spinup, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEADvsHC_spindown_GO <-  enrichGO(gene = DEADvsHC_spindown, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)

DEADvsHC_gran_GO <-  enrichGO(gene = DEADvsHC_gran, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEADvsHC_granup_GO <-  enrichGO(gene = DEADvsHC_granup, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)
DEADvsHC_grandown_GO <-  enrichGO(gene = DEADvsHC_grandown, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "BP", readable = T)

####save####
# list <- list("GO DE ADvsHC basal III" = DEADvsHC_basalIII_GO,
#              "GO DE ADvsHC basal IIIup" = DEADvsHC_basalIIIup_GO,
#              "GO DE ADvsHC basal IIIdown" = DEADvsHC_basalIIIdown_GO,
#              "GO DE ADvsHC basal II" = DEADvsHC_basalII_GO,
#              "GO DE ADvsHC basal IIup" = DEADvsHC_basalIIup_GO,
#              "GO DE ADvsHC basal IIdown" = DEADvsHC_basalIIdown_GO,
#              "GO DE ADvsHC basal I" = DEADvsHC_basalI_GO,
#              "GO DE ADvsHC basal Iup" = DEADvsHC_basalIup_GO,
#              "GO DE ADvsHC basal Idown" = DEADvsHC_basalIdown_GO,
#              "GO DE ADvsHC basal IV" = DEADvsHC_basalIV_GO,
#              "GO DE ADvsHC basal IVup" = DEADvsHC_basalIVup_GO,
#              "GO DE ADvsHC basal IVdown" = DEADvsHC_basalIVdown_GO,
#              "GO DE ADvsHC spinosum" = DEADvsHC_spin_GO,
#              "GO DE ADvsHC spinosumup" = DEADvsHC_spinup_GO,
#              "GO DE ADvsHC spinosumdown" = DEADvsHC_spindown_GO,
#              "GO DE ADvsHC granulosum" = DEADvsHC_gran_GO,
#              "GO DE ADvsHC granulosumup" = DEADvsHC_granup_GO,
#              "GO DE ADvsHC granulosumdown" = DEADvsHC_grandown_GO)
# write.xlsx(list, file = paste0(save.dir, "GO DE ADvsHC.xlsx"))

# saveRDS(DEADvsHC_basalIII_GO, paste0(save.dir, "DEADvsHC_basalIII_GO.rds"))
# saveRDS(DEADvsHC_basalIIIup_GO, paste0(save.dir, "DEADvsHC_basalIIIup_GO.rds"))
# saveRDS(DEADvsHC_basalIIIdown_GO, paste0(save.dir, "DEADvsHC_basalIIIdown_GO.rds"))
# saveRDS(DEADvsHC_basalII_GO, paste0(save.dir, "DEADvsHC_basalII_GO.rds"))
# saveRDS(DEADvsHC_basalIIup_GO, paste0(save.dir, "DEADvsHC_basalIIup_GO.rds"))
# saveRDS(DEADvsHC_basalIIdown_GO, paste0(save.dir, "DEADvsHC_basalIIdown_GO.rds"))
# saveRDS(DEADvsHC_basalI_GO, paste0(save.dir, "DEADvsHC_basalI_GO.rds"))
# saveRDS(DEADvsHC_basalIup_GO, paste0(save.dir, "DEADvsHC_basalIup_GO.rds"))
# saveRDS(DEADvsHC_basalIdown_GO, paste0(save.dir, "DEADvsHC_basalIdown_GO.rds"))
# saveRDS(DEADvsHC_basalIV_GO, paste0(save.dir, "DEADvsHC_basalIV_GO.rds"))
# saveRDS(DEADvsHC_basalIVup_GO, paste0(save.dir, "DEADvsHC_basalIVup_GO.rds"))
# saveRDS(DEADvsHC_basalIVdown_GO, paste0(save.dir, "DEADvsHC_basalIVdown_GO.rds"))
# saveRDS(DEADvsHC_spin_GO, paste0(save.dir, "DEADvsHC_spin_GO.rds"))
# saveRDS(DEADvsHC_spinup_GO, paste0(save.dir, "DEADvsHC_spinup_GO.rds"))
# saveRDS(DEADvsHC_spindown_GO, paste0(save.dir, "DEADvsHC_spindown_GO.rds"))
# saveRDS(DEADvsHC_gran_GO, paste0(save.dir, "DEADvsHC_gran_GO.rds"))
# saveRDS(DEADvsHC_granup_GO, paste0(save.dir, "DEADvsHC_granup_GO.rds"))
# saveRDS(DEADvsHC_grandown_GO, paste0(save.dir, "DEADvsHC_grandown_GO.rds"))

####visualize####
GO_ADvsHC_basalIIIup_vis <- DEADvsHC_basalIIIup_GO[c(75, 7, 170, 13, 43, 82, 108, 123, 188, 10, 28, 147, 168, 132, 93)] 
GO_ADvsHC_basalIIIup_vis$Description <- factor(GO_ADvsHC_basalIIIup_vis$Description, levels = GO_ADvsHC_basalIIIup_vis$Description) #change order manually
GO_ADvsHC_basalIIIup_vis$category <- c(rep("inflammation", 5), rep("cellular stress", 4), rep("cell death", 3), rep("dysregulated\nlineage programs", 3))
GO_ADvsHC_basalIIIup_vis$category <- factor(GO_ADvsHC_basalIIIup_vis$category, levels = c("inflammation", "cellular stress", "cell death", "dysregulated\nlineage programs"))

#Figure 4a
ggplot(GO_ADvsHC_basalIIIup_vis, aes(x = Count, y = forcats::fct_rev(Description))) +
        geom_col(aes(fill = p.adjust)) +
        facet_grid("category", switch = "y", scale = "free_y", space = "free_y") +
        scale_fill_gradient(high = "#FAFAD2", low = "darkgoldenrod2", guide = guide_colourbar(reverse = TRUE)) +
        theme_bw(base_size = 13) +  scale_y_discrete(labels = scales::label_wrap(30)) +
        ylab(NULL) + xlab("gene count") + labs(fill = "adjusted \np value") +
        theme(axis.text.y = element_text(size = 13), strip.text.y = element_text(size = 13), strip.placement = "outside")

# cellular stress: 82 response to unfolded protein,
                # 108 intrinsic apoptotic signaling pathway in response to endoplasmic reticulum stress,
                # 123 positive regulation of release of cytochrome c from mitochondrium,
                # 188 apoptotic mitochondrial changes (possible addition: , mitophagy, autophagy of mitochondrion) 
# cell death: 10 intrinsic apoptotic signaling, 28 extrinsic apoptotic signaling, 147 necroptotic process
#             (possible addition: 3 regulation of apoptotic signaling pathway, 42 macroautophagy programmed necrotic cell death, regulation of autophagy, autophagosome assembly, autophagosome maturation)
# dysregulated lineage programs: 132, 93 epithelial cell migration (possible addition: cell-cell junction organization, regulation of protein-containing complex assembly, establishment or maintenance of cell polarity)

#Supplemental Figure 6c
# pdf(file = paste0(save.dir, "Supplemental Figure 6c.pdf"))
cnetplot(DEADvsHC_basalIIIup_GO, showCategory = DEADvsHC_basalIIIup_GO$Description[c(132)]) + guides(size = "none")
# dev.off()

#Figure 4d
View(DEADvsHC_basalIIIup_GO[c(54, 240, 190, 3, 61, 132, 227)])

#extracting the genes of the hub categories
hub_genes <- DEADvsHC_basalIIIup_GO$geneID[c(54, 240, 190, 3, 61, 132, 227)]
hub_genes <- strsplit(hub_genes, "/") # Split the string by "/"
hub_genes <- unlist(hub_genes) # unlist to get a single vector

count <- table(hub_genes)
count <- sort(count, decreasing = T)
View(count)
# write.xlsx(count, file = paste0(save.dir, "hub_genes.xlsx"))

#supplemental figure 6d
VlnPlot(rojahn_kc, group.by = "svm_annotation", features = c("WNT5A", "CAV1", "CD74", "HIF1A", "HSPA1A", "PTPN2", "PYCARD", "ROCK2", "TNF"), split.by = "state", stack = T, flip = T, pt.size = 0, cols = colors_state) + theme(axis.title.x = element_blank()) + scale_x_discrete(labels = scales::label_wrap(10)) 

#Figure 5d
# pdf(file = paste0(save.dir, "Figure 5d.pdf"), height = 17, width = 25)
cnetplot(DEADvsHC_basalIIIup_GO, showCategory = DEADvsHC_basalIIIup_GO$Description[c(54, 240, 190, 3, 61, 132, 227)], label_format = 30) + guides(size = "none") + theme(text = element_text(size = 4))
# dev.off()

# pdf(file = paste0(save.dir, "Figure 5d_empty.pdf"), height = 10, width = 16)
cnetplot(DEADvsHC_basalIIIup_GO, showCategory = DEADvsHC_basalIIIup_GO$Description[c(54, 240, 190, 3, 61, 132, 227)], node_label = "category") + guides(size = "none")
# dev.off()
```

```{r GO GSEA}
####read####
DEADvsHC_basalIII_gseaGO <- readRDS(paste0(save.dir, "DEADvsHC_basalIII_gseaGO.rds"))
DEADvsHC_basalII_gseaGO <- readRDS(paste0(save.dir, "DEADvsHC_basalII_gseaGO.rds"))
DEADvsHC_basalI_gseaGO <- readRDS(paste0(save.dir, "DEADvsHC_basalI_gseaGO.rds"))
DEADvsHC_basalIV_gseaGO <- readRDS(paste0(save.dir, "DEADvsHC_basalIV_gseaGO.rds"))
DEADvsHC_spin_gseaGO <- readRDS(paste0(save.dir, "DEADvsHC_spin_gseaGO.rds"))
DEADvsHC_gran_gseaGO <- readRDS(paste0(save.dir, "DEADvsHC_gran_gseaGO.rds"))

#####calculate####
#basal III
DEADvsHC_basalIII_gseaGO <- gseGO(geneList = DEADvsHC_basalIII_gsea, OrgDb = org.Hs.eg.db)
DEADvsHC_basalIII_gseaGO <- setReadable(DEADvsHC_basalIII_gseaGO, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#basal II
DEADvsHC_basalII_gseaGO <- gseGO(geneList = DEADvsHC_basalII_gsea, OrgDb = org.Hs.eg.db)
DEADvsHC_basalII_gseaGO <- setReadable(DEADvsHC_basalII_gseaGO, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#basal I
DEADvsHC_basalI_gseaGO <- gseGO(geneList = DEADvsHC_basalI_gsea, OrgDb = org.Hs.eg.db)
DEADvsHC_basalI_gseaGO <- setReadable(DEADvsHC_basalI_gseaGO, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#basal IV
DEADvsHC_basalIV_gseaGO <- gseGO(geneList = DEADvsHC_basalIV_gsea, OrgDb = org.Hs.eg.db)
DEADvsHC_basalIV_gseaGO <- setReadable(DEADvsHC_basalIV_gseaGO, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#spinosum
DEADvsHC_spin_gseaGO <- gseGO(geneList = DEADvsHC_spin_gsea, OrgDb = org.Hs.eg.db)
DEADvsHC_spin_gseaGO <- setReadable(DEADvsHC_spin_gseaGO, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#granulosum
DEADvsHC_gran_gseaGO <- gseGO(geneList = DEADvsHC_gran_gsea, OrgDb = org.Hs.eg.db)
DEADvsHC_gran_gseaGO <- setReadable(DEADvsHC_gran_gseaGO, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#####save####
# list <- list("GSEA GO ADvsHC basal III " = DEADvsHC_basalIII_gseaGO,
#              "GSEA GO ADvsHC basal II" = DEADvsHC_basalII_gseaGO,
#              "GSEA GO ADvsHC basal I" = DEADvsHC_basalI_gseaGO,
#              "GSEA GO ADvsHC basal IV" = DEADvsHC_basalIV_gseaGO,
#              "GSEA GO ADvsHC spinosum" = DEADvsHC_spin_gseaGO,
#              "GSEA GO ADvsHC granulosum" = DEADvsHC_gran_gseaGO)
# write.xlsx(list, file = paste0(save.dir, "GO GSEA ADvsHC logFC 0.1.xlsx"))
# 
# saveRDS(DEADvsHC_basalIII_gseaGO, paste0(save.dir, "DEADvsHC_basalIII_gseaGO.rds"))
# saveRDS(DEADvsHC_basalII_gseaGO, paste0(save.dir, "DEADvsHC_basalII_gseaGO.rds"))
# saveRDS(DEADvsHC_basalI_gseaGO, paste0(save.dir, "DEADvsHC_basalI_gseaGO.rds"))
# saveRDS(DEADvsHC_basalIV_gseaGO, paste0(save.dir, "DEADvsHC_basalIV_gseaGO.rds"))
# saveRDS(DEADvsHC_spin_gseaGO, paste0(save.dir, "DEADvsHC_spin_gseaGO.rds"))
# saveRDS(DEADvsHC_gran_gseaGO, paste0(save.dir, "DEADvsHC_gran_gseaGO.rds"))

#####visualize####
#Supplemental figure 6a, preparing the facet text
GOgsea_ADvsHC_basalIII_vis <- DEADvsHC_basalIII_gseaGO[c(123, 263, 337, 163, 63, 138, 2, 387, 144, 416, 98, 166, 202, 92, 118, 205, 315, 201, 342, 161, 262, 74, 168)] #be aware, excel has to be sorted with decreasing qvalue

GOgsea_ADvsHC_basalIII_vis$Description <- factor(GOgsea_ADvsHC_basalIII_vis$Description, levels = GOgsea_ADvsHC_basalIII_vis$Description)
GOgsea_ADvsHC_basalIII_vis$category <- c(rep("cytokine production", 6), rep("immune cell activation", 7), rep("antigen\npresentation", 2), rep("cellular\nstress", 2), rep("cell\ndeath", 2), rep("dysregulated\nlineage programs", 4)) 
GOgsea_ADvsHC_basalIII_vis$category <- factor(GOgsea_ADvsHC_basalIII_vis$category, levels = c("cytokine production", "immune cell activation", "antigen\npresentation", "cellular\nstress", "cell\ndeath", "dysregulated\nlineage programs"))

ggplot(GOgsea_ADvsHC_basalIII_vis, aes(x = NES, y = forcats::fct_rev(Description))) +
        geom_point(aes(color = p.adjust), size = 5) + #geom_dotplot stackratio = 1.5, dotsize = 3
        facet_grid("category", switch = "y", scale = "free_y", space = "free_y") +
        scale_color_gradient(high = "#FAFAD2", low = "darkgoldenrod2", guide = guide_colourbar(reverse = TRUE)) +
        theme_bw(base_size = 13) +  scale_y_discrete(labels = scales::label_wrap(38)) +
        ylab(NULL) + xlab("normalized enrichment score") + labs(fill = "adjusted \np value") + 
        scale_x_continuous(limits = c(1, 2.5)) + 
        theme(axis.text.y = element_text(size = 14), strip.text.y = element_text(size = 14), strip.placement = "outside")

#Supplemental figure 6a, ridgeplot
gseacyt <- ridgeplot(DEADvsHC_basalIII_gseaGO, showCategory = DEADvsHC_basalIII_gseaGO$Description[c(123, 263, 337, 163, 63, 138)], orderBy = "Description", decreasing = T) +  
              scale_x_continuous(limits = c(-2, 11)) +
              scale_fill_gradient(high = c("#FAFAD2"), low = c("darkgoldenrod2"))  + labs(fill = "adjusted \np value")
  
gseacell <- ridgeplot(DEADvsHC_basalIII_gseaGO, showCategory = DEADvsHC_basalIII_gseaGO$Description[c(2, 387, 144, 416, 98, 166, 202)], orderBy = "Description") + 
              scale_x_continuous(limits = c(-2, 11)) +
              scale_fill_gradient(high = c("#FAFAD2"), low = c("darkgoldenrod2"))  + labs(fill = "adjusted \np value")

gseaAG <- ridgeplot(DEADvsHC_basalIII_gseaGO, showCategory = DEADvsHC_basalIII_gseaGO$Description[c(92, 118)], orderBy = "Description", decreasing = T) + 
              scale_x_continuous(limits = c(-2, 11)) +
              scale_fill_gradient(high = c("#FAFAD2"), low = c("darkgoldenrod2")) + labs(fill = "adjusted \np value")

gseastress <- ridgeplot(DEADvsHC_basalIII_gseaGO, showCategory = DEADvsHC_basalIII_gseaGO$Description[c(205, 315)]) +
              scale_x_continuous(limits = c(-2, 11)) +
              scale_fill_gradient(high = c("#FAFAD2"), low = c("darkgoldenrod2")) + labs(fill = "adjusted \np value")

gseadeath <- ridgeplot(DEADvsHC_basalIII_gseaGO, showCategory = DEADvsHC_basalIII_gseaGO$Description[c(201, 342)]) + 
              scale_x_continuous(limits = c(-2, 11)) +
              scale_fill_gradient(high = c("#FAFAD2"), low = c("darkgoldenrod2"))  + labs(fill = "adjusted \np value")

gsealineage <- ridgeplot(DEADvsHC_basalIII_gseaGO, showCategory = DEADvsHC_basalIII_gseaGO$Description[c(161, 262, 74, 168)], orderBy = "ID", decreasing = TRUE) + 
              scale_x_continuous(limits = c(-2, 11)) +
              scale_fill_gradient(high = c("#FAFAD2"), low = c("darkgoldenrod2")) + xlab("normalized enrichment score") + labs(fill = "adjusted \np value")

ggpubr::ggarrange(gseacyt, gseacell, gseaAG, gseastress, gseadeath, gsealineage, ncol = 1, align = "v", heights = c(6, 5, 2, 2, 2, 4), common.legend = TRUE, legend = "right") #adding scale_y_discrete(labels = scales::label_wrap(30)) doesn't work

# inflammation:
#   cytokine production: 123, 263, 337, 163, 63, 138
#   immune cell activation: 2, 387, 144, 416, 98, 166, 202
#   AG presentation: 92, 118
# cellular stress: 205 negative regulation of protein maturation, 
#             315 response to oxidative stress
# cell death: 201 positive regulation of programmed cell death, 
#             342 inflammatory cell apoptotic process
# dysregulated lineage programs:  74 regulation of cell-cell adhesion, 168 positive regulation of cell migration, 161 positive regulation of cell population proliferation, 262 epidermal cell differentiation
```

### KEGG pathway enrichment

```{r KEGG over enrichment}
#####read####
DEADvsHC_basalIII_kegg <- readRDS(paste0(save.dir, "DEADvsHC_basalIII_kegg.rds"))
DEADvsHC_basalIIIup_kegg <- readRDS(paste0(save.dir, "DEADvsHC_basalIIIup_kegg.rds"))
DEADvsHC_basalIIIdown_kegg <- readRDS(paste0(save.dir, "DEADvsHC_basalIIIdown_kegg.rds"))
DEADvsHC_basalII_kegg <- readRDS(paste0(save.dir, "DEADvsHC_basalII_kegg.rds"))
DEADvsHC_basalIIup_kegg <- readRDS(paste0(save.dir, "DEADvsHC_basalIIup_kegg.rds"))
DEADvsHC_basalIIdown_kegg <- readRDS(paste0(save.dir, "DEADvsHC_basalIIdown_kegg.rds"))
DEADvsHC_basalI_kegg <- readRDS(paste0(save.dir, "DEADvsHC_basalI_kegg.rds"))
DEADvsHC_basalIup_kegg <- readRDS(paste0(save.dir, "DEADvsHC_basalIup_kegg.rds"))
DEADvsHC_basalIdown_kegg <- readRDS(paste0(save.dir, "DEADvsHC_basalIdown_kegg.rds"))
DEADvsHC_basalIV_kegg <- readRDS(paste0(save.dir, "DEADvsHC_basalIV_kegg.rds"))
DEADvsHC_basalIVup_kegg <- readRDS(paste0(save.dir, "DEADvsHC_basalIVup_kegg.rds"))
DEADvsHC_basalIVdown_kegg <- readRDS(paste0(save.dir, "DEADvsHC_basalIVdown_kegg.rds"))
DEADvsHC_spin_kegg <- readRDS(paste0(save.dir, "DEADvsHC_spin_kegg.rds"))
DEADvsHC_spinup_kegg <- readRDS(paste0(save.dir, "DEADvsHC_spinup_kegg.rds"))
DEADvsHC_spindown_kegg <- readRDS(paste0(save.dir, "DEADvsHC_spindown_kegg.rds"))
DEADvsHC_gran_kegg <- readRDS(paste0(save.dir, "DEADvsHC_gran_kegg.rds"))
DEADvsHC_granup_kegg <- readRDS(paste0(save.dir, "DEADvsHC_granup_kegg.rds"))
DEADvsHC_grandown_kegg <- readRDS(paste0(save.dir, "DEADvsHC_grandown_kegg.rds"))

#####calculate####
DEADvsHC_basalIII_kegg <- enrichKEGG(gene = DEADvsHC_basalIII, universe = background, pvalueCutoff  = 0.05)
DEADvsHC_basalIIIup_kegg <- enrichKEGG(gene = DEADvsHC_basalIIIup, universe = background, pvalueCutoff  = 0.05)
DEADvsHC_basalIIIdown_kegg <- enrichKEGG(gene = DEADvsHC_basalIIIdown, universe = background, pvalueCutoff  = 0.05)

DEADvsHC_basalII_kegg <- enrichKEGG(gene = DEADvsHC_basalII, universe = background, pvalueCutoff  = 0.05)
DEADvsHC_basalIIup_kegg <- enrichKEGG(gene = DEADvsHC_basalIIup, universe = background, pvalueCutoff  = 0.05)
DEADvsHC_basalIIdown_kegg <- enrichKEGG(gene = DEADvsHC_basalIIdown, universe = background, pvalueCutoff  = 0.05)

DEADvsHC_basalI_kegg <- enrichKEGG(gene = DEADvsHC_basalI, universe = background, pvalueCutoff  = 0.05)
DEADvsHC_basalIup_kegg <- enrichKEGG(gene = DEADvsHC_basalIup, universe = background, pvalueCutoff  = 0.05)
DEADvsHC_basalIdown_kegg <- enrichKEGG(gene = DEADvsHC_basalIdown, universe = background, pvalueCutoff  = 0.05)

DEADvsHC_basalIV_kegg <- enrichKEGG(gene = DEADvsHC_basalIV, universe = background, pvalueCutoff  = 0.05)
DEADvsHC_basalIVup_kegg <- enrichKEGG(gene = DEADvsHC_basalIVup, universe = background, pvalueCutoff  = 0.05)
DEADvsHC_basalIVdown_kegg <- enrichKEGG(gene = DEADvsHC_basalIVdown, universe = background, pvalueCutoff  = 0.05)

DEADvsHC_spin_kegg <- enrichKEGG(gene = DEADvsHC_spin, universe = background, pvalueCutoff  = 0.05)
DEADvsHC_spinup_kegg <- enrichKEGG(gene = DEADvsHC_spinup, universe = background, pvalueCutoff  = 0.05)
DEADvsHC_spindown_kegg <- enrichKEGG(gene = DEADvsHC_spindown, universe = background, pvalueCutoff  = 0.05)

DEADvsHC_gran_kegg <- enrichKEGG(gene = DEADvsHC_gran, universe = background, pvalueCutoff  = 0.05)
DEADvsHC_granup_kegg <- enrichKEGG(gene = DEADvsHC_granup, universe = background, pvalueCutoff  = 0.05)
DEADvsHC_grandown_kegg <- enrichKEGG(gene = DEADvsHC_grandown, universe = background, pvalueCutoff  = 0.05)

# make result read.table
DEADvsHC_basalIII_kegg <- setReadable(DEADvsHC_basalIII_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEADvsHC_basalIIIup_kegg <- setReadable(DEADvsHC_basalIIIup_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEADvsHC_basalIIIdown_kegg <- setReadable(DEADvsHC_basalIIIdown_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

DEADvsHC_basalII_kegg <- setReadable(DEADvsHC_basalII_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEADvsHC_basalIIup_kegg <- setReadable(DEADvsHC_basalIIup_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEADvsHC_basalIIdown_kegg <- setReadable(DEADvsHC_basalIIdown_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

DEADvsHC_basalI_kegg <- setReadable(DEADvsHC_basalI_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEADvsHC_basalIup_kegg <- setReadable(DEADvsHC_basalIup_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEADvsHC_basalIdown_kegg <- setReadable(DEADvsHC_basalIdown_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

DEADvsHC_basalIV_kegg <- setReadable(DEADvsHC_basalIV_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEADvsHC_basalIVup_kegg <- setReadable(DEADvsHC_basalIVup_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEADvsHC_basalIVdown_kegg <- setReadable(DEADvsHC_basalIVdown_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

DEADvsHC_spin_kegg <- setReadable(DEADvsHC_spin_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEADvsHC_spinup_kegg <- setReadable(DEADvsHC_spinup_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEADvsHC_spindown_kegg <- setReadable(DEADvsHC_spindown_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

DEADvsHC_gran_kegg <- setReadable(DEADvsHC_gran_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEADvsHC_granup_kegg <- setReadable(DEADvsHC_granup_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
DEADvsHC_grandown_kegg <- setReadable(DEADvsHC_grandown_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

####save####
# list <- list("KEGG DE ADvsHC basal III" = DEADvsHC_basalIII_kegg,
#              "KEGG DE ADvsHC basal IIIup" = DEADvsHC_basalIIIup_kegg,
#              "KEGG DE ADvsHC basal IIIdown" = DEADvsHC_basalIIIdown_kegg,
#              "KEGG DE ADvsHC basal II" = DEADvsHC_basalII_kegg,
#              "KEGG DE ADvsHC basal IIup" = DEADvsHC_basalIIup_kegg,
#              "KEGG DE ADvsHC basal IIdown" = DEADvsHC_basalIIdown_kegg,
#              "KEGG DE ADvsHC basal I" = DEADvsHC_basalI_kegg,
#              "KEGG DE ADvsHC basal Iup" = DEADvsHC_basalIup_kegg,
#              "KEGG DE ADvsHC basal Idown" = DEADvsHC_basalIdown_kegg,
#              "KEGG DE ADvsHC basal IV" = DEADvsHC_basalIV_kegg,
#              "KEGG DE ADvsHC basal IVup" = DEADvsHC_basalIVup_kegg,
#              "KEGG DE ADvsHC basal IVdown" = DEADvsHC_basalIVdown_kegg,
#              "KEGG DE ADvsHC spinosum" = DEADvsHC_spin_kegg,
#              "KEGG DE ADvsHC spinosum up" = DEADvsHC_spinup_kegg,
#              "KEGG DE ADvsHC spinosum down" = DEADvsHC_spindown_kegg,
#              "KEGG DE ADvsHC granulosum" = DEADvsHC_gran_kegg,
#              "KEGG DE ADvsHC granulosum up" = DEADvsHC_granup_kegg,
#              "KEGG DE ADvsHC granulosum down" = DEADvsHC_grandown_kegg)
# write.xlsx(list, file = paste0(save.dir, "KEGG DE ADvsHC.xlsx"))
# 
# saveRDS(DEADvsHC_basalIII_kegg, paste0(save.dir, "DEADvsHC_basalIII_kegg.rds"))
# saveRDS(DEADvsHC_basalIIIup_kegg, paste0(save.dir, "DEADvsHC_basalIIIup_kegg.rds"))
# saveRDS(DEADvsHC_basalIIIdown_kegg, paste0(save.dir, "DEADvsHC_basalIIIdown_kegg.rds"))
# saveRDS(DEADvsHC_basalII_kegg, paste0(save.dir, "DEADvsHC_basalII_kegg.rds"))
# saveRDS(DEADvsHC_basalIIup_kegg, paste0(save.dir, "DEADvsHC_basalIIup_kegg.rds"))
# saveRDS(DEADvsHC_basalIIdown_kegg, paste0(save.dir, "DEADvsHC_basalIIdown_kegg.rds"))
# saveRDS(DEADvsHC_basalI_kegg, paste0(save.dir, "DEADvsHC_basalI_kegg.rds"))
# saveRDS(DEADvsHC_basalIup_kegg, paste0(save.dir, "DEADvsHC_basalIup_kegg.rds"))
# saveRDS(DEADvsHC_basalIdown_kegg, paste0(save.dir, "DEADvsHC_basalIdown_kegg.rds"))
# saveRDS(DEADvsHC_basalIV_kegg, paste0(save.dir, "DEADvsHC_basalIV_kegg.rds"))
# saveRDS(DEADvsHC_basalIVup_kegg, paste0(save.dir, "DEADvsHC_basalIVup_kegg.rds"))
# saveRDS(DEADvsHC_basalIVdown_kegg, paste0(save.dir, "DEADvsHC_basalIVdown_kegg.rds"))
# saveRDS(DEADvsHC_spin_kegg, paste0(save.dir, "DEADvsHC_spin_kegg.rds"))
# saveRDS(DEADvsHC_spinup_kegg, paste0(save.dir, "DEADvsHC_spinup_kegg.rds"))
# saveRDS(DEADvsHC_spindown_kegg, paste0(save.dir, "DEADvsHC_spindown_kegg.rds"))
# saveRDS(DEADvsHC_gran_kegg, paste0(save.dir, "DEADvsHC_gran_kegg.rds"))
# saveRDS(DEADvsHC_granup_kegg, paste0(save.dir, "DEADvsHC_granup_kegg.rds"))
# saveRDS(DEADvsHC_grandown_kegg, paste0(save.dir, "DEADvsHC_grandown_kegg.rds"))

#####visualize####
#Figure 5b
KEGG_ADvsHC_basalIIIup_vis <- DEADvsHC_basalIIIup_kegg[c(62, 7, 1, 22, 19, 5, 33, 47, 31)] 
KEGG_ADvsHC_basalIIIup_vis$Description <- factor(KEGG_ADvsHC_basalIIIup_vis$Description, levels = KEGG_ADvsHC_basalIIIup_vis$Description)
KEGG_ADvsHC_basalIIIup_vis$category <- c(rep("inflammation", 2), rep("cellular stress", 2), rep("cell death", 2), rep("dysregulated\nlineage programs", 3))
KEGG_ADvsHC_basalIIIup_vis$category <- factor(KEGG_ADvsHC_basalIIIup_vis$category, levels = c("inflammation", "cellular stress", "cell death", "dysregulated\nlineage programs"))

ggplot(KEGG_ADvsHC_basalIIIup_vis, aes(x = Count, y = forcats::fct_rev(Description))) +
        geom_col(aes(fill = p.adjust)) +
        facet_grid("category", switch = "y", scale = "free_y", space = "free_y") +
        scale_fill_gradient(high = "#FAFAD2", low = "darkgoldenrod2", guide = guide_colourbar(reverse = TRUE)) +
        theme_bw(base_size = 13) +  scale_y_discrete(labels = scales::label_wrap(30)) +
        ylab(NULL) + xlab("gene count") + labs(fill = "adjusted \np value") +
        theme(axis.text.y = element_text(size = 13), strip.text.y = element_text(size = 13), strip.placement = "outside")

# #visualize genes in pathway
# DEADvsHC_basalIIIup_kegg <- enrichKEGG(gene = DEADvsHC_basalIIIup, universe = background, pvalueCutoff  = 0.05) #visualization only works with entrezID
# 
# #inflammation
# browseKEGG(DEADvsHC_basalIIIup_kegg, 'hsa04659')
# browseKEGG(DEADvsHC_basalIIIup_kegg, 'hsa04657')
# browseKEGG(DEADvsHC_basalIIIup_kegg, 'hsa04621')
# 
# #stress
# browseKEGG(DEADvsHC_basalIIIup_kegg, 'hsa03040')
# browseKEGG(DEADvsHC_basalIIIup_kegg, 'hsa03050')
# browseKEGG(DEADvsHC_basalIIIup_kegg, 'hsa05208')
# browseKEGG(DEADvsHC_basalIIIup_kegg, 'hsa04137')
# 
# #cell death
# browseKEGG(DEADvsHC_basalIIIup_kegg, 'hsa04210')
# browseKEGG(DEADvsHC_basalIIIup_kegg, 'hsa04217')
# 
# #dysregulated lineage programs
# browseKEGG(DEADvsHC_basalIIIup_kegg, 'hsa04010')
# browseKEGG(DEADvsHC_basalIIIup_kegg, 'hsa04810')
# browseKEGG(DEADvsHC_basalIIIup_kegg, 'hsa04520')
```

```{r KEGG GSEA}
####read####
DEADvsHC_basalIII_gseaKEGG <- readRDS(paste0(save.dir, "DEADvsHC_basalIII_gseaKEGG.rds"))
DEADvsHC_basalII_gseaKEGG <- readRDS(paste0(save.dir, "DEADvsHC_basalII_gseaKEGG.rds"))
DEADvsHC_basalI_gseaKEGG <- readRDS(paste0(save.dir, "DEADvsHC_basalI_gseaKEGG.rds"))
DEADvsHC_basalIV_gseaKEGG <- readRDS(paste0(save.dir, "DEADvsHC_basalIV_gseaKEGG.rds"))
DEADvsHC_spin_gseaKEGG <- readRDS(paste0(save.dir, "DEADvsHC_spin_gseaKEGG.rds"))
DEADvsHC_gran_gseaKEGG <- readRDS(paste0(save.dir, "DEADvsHC_gran_gseaKEGG.rds"))

####calculate####
#basal III
DEADvsHC_basalIII_gseaKEGG <- gseKEGG(geneList = DEADvsHC_basalIII_gsea, organism = 'hsa', keyType = "kegg", pvalueCutoff = 0.05)
DEADvsHC_basalIII_gseaKEGG <- setReadable(DEADvsHC_basalIII_gseaKEGG, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#basal II
DEADvsHC_basalII_gseaKEGG <- gseKEGG(geneList = DEADvsHC_basalII_gsea, organism = 'hsa', keyType = "kegg", pvalueCutoff = 0.05)
DEADvsHC_basalII_gseaKEGG <- setReadable(DEADvsHC_basalII_gseaKEGG, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#basal I
DEADvsHC_basalI_gseaKEGG <- gseKEGG(geneList = DEADvsHC_basalI_gsea, organism = 'hsa', keyType = "kegg", pvalueCutoff = 0.05)
DEADvsHC_basalI_gseaKEGG <- setReadable(DEADvsHC_basalI_gseaKEGG, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#basal IV
DEADvsHC_basalIV_gseaKEGG <- gseKEGG(geneList = DEADvsHC_basalIV_gsea, organism = 'hsa', keyType = "kegg", pvalueCutoff = 0.05)
DEADvsHC_basalIV_gseaKEGG <- setReadable(DEADvsHC_basalIV_gseaKEGG, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#spinosum
DEADvsHC_spin_gseaKEGG <- gseKEGG(geneList = DEADvsHC_spin_gsea, organism = 'hsa', keyType = "kegg", pvalueCutoff = 0.05)
DEADvsHC_spin_gseaKEGG <- setReadable(DEADvsHC_spin_gseaKEGG, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

#granulosum
DEADvsHC_gran_gseaKEGG <- gseKEGG(geneList = DEADvsHC_gran_gsea, organism = 'hsa', keyType = "kegg", pvalueCutoff = 0.05)
DEADvsHC_gran_gseaKEGG <- setReadable(DEADvsHC_gran_gseaKEGG, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")

####save####
# list <- list("GSEA KEGG ADvsHC basal III " = DEADvsHC_basalIII_gseaKEGG,
#              "GSEA KEGG ADvsHC basal II" = DEADvsHC_basalII_gseaKEGG,
#              "GSEA KEGG ADvsHC basal I" = DEADvsHC_basalI_gseaKEGG,
#              "GSEA KEGG ADvsHC basal IV" = DEADvsHC_basalIV_gseaKEGG,
#              "GSEA KEGG ADvsHC spinosum" = DEADvsHC_spin_gseaKEGG,
#              "GSEA KEGG ADvsHC granulosum" = DEADvsHC_gran_gseaKEGG)
# write.xlsx(list, file = paste0(save.dir, "KEGG GSEA ADvsHC logFC 0.1.xlsx"))
# 
# saveRDS(DEADvsHC_basalIII_gseaKEGG, paste0(save.dir, "DEADvsHC_basalIII_gseaKEGG.rds"))
# saveRDS(DEADvsHC_basalII_gseaKEGG, paste0(save.dir, "DEADvsHC_basalII_gseaKEGG.rds"))
# saveRDS(DEADvsHC_basalI_gseaKEGG, paste0(save.dir, "DEADvsHC_basalI_gseaKEGG.rds"))
# saveRDS(DEADvsHC_basalIV_gseaKEGG, paste0(save.dir, "DEADvsHC_basalIV_gseaKEGG.rds"))
# saveRDS(DEADvsHC_spin_gseaKEGG, paste0(save.dir, "DEADvsHC_spin_gseaKEGG.rds"))
# saveRDS(DEADvsHC_gran_gseaKEGG, paste0(save.dir, "DEADvsHC_gran_gseaKEGG.rds"))

####visualize####
#visualize genes in pathway
# DEADvsHC_basalIII_gseaKEGG <- gseKEGG(geneList = DEADvsHC_basalIII_gsea, organism = 'hsa', keyType = "kegg", pvalueCutoff = 0.05)
# #works only with entrezID
# #you need to manually add the entrezIDs after the "hsa..." pathway number in the browser url. e.g. https://www.kegg.jp/kegg-bin/show_pathway?hsa04062/6362/6354/6352/6363/6373/6364/5293/9560/53358/2920/6361/3627/7852/2921/3576/4067/653361/6464/4893/58191/5880/6772/5829/10681/6351
# 
# #you have to manually take the entrezIDs from the "core enrichment" column
# # View(DEADvsHC_basalIII_gseaKEGG)
# # DEADvsHC_basalIII_gseaKEGG[ , "core_enrichment"]
# 
# #inflammation
# browseKEGG(DEADvsHC_basalIII_gseaKEGG, 'hsa04062') # https://www.kegg.jp/kegg-bin/show_pathway?hsa04062/6362/6354/6352/6363/6373/6364/5293/9560/53358/2920/6361/3627/7852/2921/3576/4067/653361/6464/4893/58191/5880/6772/5829/10681/6351
# 
# browseKEGG(DEADvsHC_basalIII_gseaKEGG, 'hsa04060') #  https://www.kegg.jp/kegg-bin/show_pathway?hsa04060/6362/6354/6352/27242/3598/51561/6363/6373/6364/9560/3575/2920/7850/4050/6361/3627/3595/8743/3600/50604/7852/2921/58985/8795/3576/3601/285613/3587/7046/3566/51330/9180/7124/8797/58191/8809/3556/84957/3589/51554/90865/3597/3455/6351
# 
# browseKEGG(DEADvsHC_basalIII_gseaKEGG, 'hsa04612') # https://www.kegg.jp/kegg-bin/show_pathway?hsa04612/972/3310/1520/3118/3122/3115/3108/3135/3304/3305/3117/4261/7124/6891/5721/3303/5993/3320/6892/5720/3113/6890
# 
# #cell death
# browseKEGG(DEADvsHC_basalIII_gseaKEGG, 'hsa04668') # https://www.kegg.jp/kegg-bin/show_pathway?hsa04668/4314/6352/6364/5293/2920/64127/4318/3627/7128/3600/3383/2921/330/7124/5600/192111/8809/7424/23118/83786/153090/8986/2919/355/602/1051/7132/843/8837/7187/5608/5606/10454/6868/7186
# 
# browseKEGG(DEADvsHC_basalIII_gseaKEGG, 'hsa04064') # https://www.kegg.jp/kegg-bin/show_pathway?hsa04064/6363/9560/2920/4050/7128/29760/5328/3383/2921/23643/3576/4067/330/7124/5971/8915/6351/23118/79092/2919/29775/148022/7132/5335/8837/7187/10912/10454/4791/7186/1647/4055/9020/10673 
# 
# #Supplemental figure 6b, visualizing the basal III HC vs AD differentially expressed genes (!ATTENTION: CXCL14 is downregulated)
# https://www.kegg.jp/kegg-bin/show_pathway?hsa04060/10850/7850/50604/9547/2920/3566/3556/58191/6362/6364/3597/6354/3460/3601/3459/53832/9235/2921/3600/146433/3572/3576/3455
```

### Progeny

AD and HC need to be analyzed together so that the pathway activities are normalized and centered together.

```{r progeny together}
#Figure 3b
Idents(rojahn_kc) <-  paste0(rojahn_kc$state,"_",rojahn_kc$svm_annotation)

cell_order <- c(paste0(levels(rojahn_kc$state)[1], "_", levels(rojahn_kc$svm_annotation)),
                paste0(levels(rojahn_kc$state)[2], "_", levels(rojahn_kc$svm_annotation)))
# cell_order <- c("HC_basal III", "AD_basal III")
pathway_order <- c("JAK-STAT", "TNFa", "NFkB", "WNT", "Estrogen", "MAPK", "EGFR", "PI3K", "TGFb", "Trail", "VEGF", "Androgen", "Hypoxia", "p53")

progeny_heatmap(rojahn_kc, nfeatures = 500, cell_order = cell_order, clustering = F, pathway_order = pathway_order)  
```

```{r progeny together - only show basal III}
#Figure 5c
Idents(rojahn_kc) <-  paste0(rojahn_kc$state,"_",rojahn_kc$svm_annotation)

cell_order <- c("HC_basal III", "AD_basal III")
pathway_order <- c("JAK-STAT", "TNFa", "NFkB", "TGFb", "WNT", "MAPK", "EGFR", "VEGF", "Estrogen", "PI3K", "Trail",  "Hypoxia", "p53", "Androgen")

progeny_heatmap(rojahn_kc, nfeatures = 500, cell_order = cell_order, clustering = F, pathway_order = pathway_order)  
```

other comparisons: 

a) One could compare basal III cells HC vs AD directly.
-> We chose not to do that because it takes the activities (HC) basal III cells out of context. All pathways will show up either up- or downregulated in a direct comparison, e.g. the Trail pathway will look upregulated in HC which could be misleading since trail activities are much higher in HC stratum spinosum or granulosum

```{r progeny basal III only}
# rojahn_kc_basalIII <- rojahn_kc[ , rojahn_kc$svm_annotation == "basal III"]
# Idents(rojahn_kc_basalIII) <-  paste0(rojahn_kc_basalIII$state)
# 
# cell_order <- c(levels(rojahn_kc_basalIII$state))
# 
# progeny_heatmap(rojahn_kc_basalIII, nfeatures = 500, cell_order = cell_order, clustering = F, pathway_order = pathway_order)
```

b) Since the scaling seems to be linear, one could directly compare them by subtracting them from each other. 

```{r progeny function comparison}
#result only interpretable if equal amount of AD vs HC conditions!
progeny_comparison = function(object, nfeatures = 100, cell_order = x, clustering = F, pathway_order = c("JAK-STAT", "TNFa", "NFkB", "VEGF",  "TGFb", "Androgen", "PI3K", "EGFR", "MAPK", "Estrogen", "p53",  "WNT", "Hypoxia", "Trail")){

  require(progeny)
  require(pheatmap)
  require(tidyr)
  require(tibble)
  
  #data frame delineating cell annotations
  cell_annotation <- data.frame(cell = names(Idents(object)),
                                annotation = as.character(Idents(object)), stringsAsFactors = FALSE)
  
  #calculation progeny, added as an assay to the seurat object 
  object <- progeny(object, scale = F, top = nfeatures, perm = 1, return_assay = TRUE)
  # object@assays$progeny
  
  #We can now directly apply Seurat functions in our Progeny scores, like scaling the pathway activity scores
  object <- Seurat::ScaleData(object, assay = "progeny")
  
  #We transform Progeny scores into a data frame to better handling the results
  progeny_scores <- as.data.frame(t(object@assays$progeny$scale.data)) %>%
    tibble::rownames_to_column("cell") %>%
    tidyr::pivot_longer(names_to = "pathway", values_to = "activity", c(-"cell")) #transforms the table into a longer table format - assigning a separate row for every pathways + coordinating activity per cell
  
  #add information of cell annotation
  progeny_scores <- inner_join(progeny_scores, cell_annotation, by = c("cell"))
  
  #summarize progeny scores by population
  progeny_summarized <- progeny_scores %>%
    group_by(pathway, annotation) %>%
    summarise(average = mean(activity), std = sd(activity))
  
  #prepare dataframe for plotting
  progeny_summarized <- progeny_summarized %>% 
    dplyr::select(-std) %>%
    tidyr::spread(pathway, average) %>%
    data.frame(row.names = "annotation", check.names = FALSE, stringsAsFactors = FALSE)
  
  progeny_summarized <- progeny_summarized[cell_order, ]
  progeny_summarized <- progeny_summarized[, pathway_order]
  
  nr = nrow(progeny_summarized) / 2
  df1 = progeny_summarized[1:nr, ]
  df2 = progeny_summarized[(nr + 1):c(2 * nr), ]
  progeny_summarized_relative = df1 - df2 #relative activity between two variables
  rownames(progeny_summarized_relative) <- gsub("([a-zA-Z]+_)", "", rownames(df1)) 
  
  #we prepare the plot
  paletteLength <- 100
  myColor <-  colorRampPalette(c("darkblue", "white", "indianred"))(paletteLength)
  
  progenyBreaks = c(seq(min(progeny_summarized_relative), 0, length.out = ceiling(paletteLength / 2) + 1),
                    seq(max(progeny_summarized_relative) / paletteLength, max(progeny_summarized_relative),  length.out = floor(paletteLength / 2))
                    )
  #we plot
  progeny_hmap <- pheatmap::pheatmap(t(progeny_summarized_relative), 
                          cluster_cols = F,
                          cluster_rows = clustering, row_order = cell_order,  
                          color = myColor, breaks = progenyBreaks, border_color = NA,
                          main = c("Pathway activity AD relative to HC"),
                          angle_col = c("45")) #fontsize = 12, fontsize_row = 9 
  return(progeny_hmap)
}

#AD needs to be together and before HC, order within AD / HC can be as desired
cell_order <- c(paste0(levels(rojahn_kc$state)[2], "_", levels(rojahn_kc$svm_annotation)),
                paste0(levels(rojahn_kc$state)[1], "_", levels(rojahn_kc$svm_annotation)))
# cell_order <- c("AD_basal III", "AD_basal I", "AD_basal II", "AD_basal IV", "AD_stratum spinosum", "AD_stratum granulosum", "HC_basal III", "HC_basal I", "HC_basal II", "HC_basal IV", "HC_stratum spinosum", "HC_stratum granulosum")

pathway_order <- c("JAK-STAT", "TNFa", "NFkB", "WNT", "Estrogen", "MAPK", "EGFR", "PI3K", "TGFb", "Trail", "VEGF", "Androgen", "Hypoxia", "p53")

Idents(rojahn_kc) <- paste0(rojahn_kc$state, "_", rojahn_kc$svm_annotation)
progeny_comparison(rojahn_kc, nfeatures = 500, cell_order = cell_order, clustering = F, pathway_order = pathway_order)
```

# Pseudotime analysis

Using slingshot on the non-integrated data provides a more accurate representation of the cell trajectory(!!) when comparing HC vs AD. However, when looking at individual genes, pseudotime performed on the integrated umap but looking at non-integrated gene expression ensures having the same trajectory of AD and HC. This is what we do when we perform trajectory interference on integrated data, the tradeseq infered gene expression analysis is using non-integrated counts again. 
Conclusion: show pseudotime timeline derived from non-integrated umap. show pseudotime heatmap of genes derived from integrated umap (gene expression will remain from non-integrated RNA assay)

## Compare pseudotime across keratinocyte subtypes (non-integrated UMAP)

```{r clustering}
DefaultAssay(rojahn_kc) <- "RNA"
rojahn_kc <- NormalizeData(rojahn_kc)
rojahn_kc <- seuratpipeline(rojahn_kc, nfeat = 2000, res = 0.5, dims = 11)

DimPlot(rojahn_kc, group.by = "sample")
DimPlot(rojahn_kc, group.by = "state")
DimPlot(rojahn_kc, group.by = "svm_annotation")
```

```{r pseudotime together}
DefaultAssay(rojahn_kc) <- "RNA"

#remove possible previous iterations
if(exists("rojahn_kc$slingPseudotime_1")){
  rojahn_kc$slingPseudotime_1 <- NULL
}

if(exists("rojahn_kc$slingPseudotime_2")){
  rojahn_kc$slingPseudotime_2 <- NULL
}

rojahn_kc_sce <- as.SingleCellExperiment(rojahn_kc, assay = "RNA") #transform seurat to SingleCellExperiment object, needed as input for slingshot 

rojahn_kc_sce <- slingshot(rojahn_kc_sce, reducedDim = 'UMAP') #strong recommendation to specify a starting cluster, clusterLabels = "svm_annotation", start.clus = "basal III", end.clus = "stratum granulosum"

#plot pseudotime on UMAP
plot(reducedDims(rojahn_kc_sce)$UMAP,
     col = colors_subtypes[as.factor(rojahn_kc_sce$svm_annotation)], pch = 20, asp = 0.5) + lines(SlingshotDataSet(rojahn_kc_sce))

#plot pseudotime 1
pseudotime <- rojahn_kc_sce@colData[ , c("svm_annotation", "slingPseudotime_1", "state", "sample")]
pseudotime <- as.data.frame(pseudotime)
pseudotime <- pseudotime[!is.na(pseudotime$slingPseudotime_1), ]
pseudotime$svm_annotation <- factor(pseudotime$svm_annotation, levels = c("basal III", "basal I", "basal II", "basal IV", "stratum spinosum", "stratum granulosum"))

#annotation of cell types
ggplot(pseudotime[pseudotime$state == "HC", ],
       aes(x = svm_annotation, y = slingPseudotime_1, fill = state, color = state)) +
    geom_dotplot(binwidth = 0.3, stackdir = "centerwhole", binaxis = "y", stackratio = 0.1) + coord_flip() +
    scale_fill_manual(values = colors_subtypes) + scale_color_manual(values = colors_state) + 
    theme_classic() + labs(y = NULL) + theme(legend.position = "none") 

ggplot(pseudotime[pseudotime$state == "AD", ],
       aes(x = svm_annotation, y = slingPseudotime_1, fill = state, color = state)) +
    geom_dotplot(binwidth = 0.3, stackdir = "centerwhole", binaxis = "y", stackratio = 0.1) + coord_flip() +
    scale_fill_manual(values = colors_subtypes) + scale_color_manual(values = colors_state) + 
    theme_classic() + labs(y = NULL) + theme(legend.position = "none") 
```

In this case, i think it is better to do the analysis for HC and AD separately. Because it is based on a shared UMAP calculation, I can still use just "half" of the UMAP for a separate HC and AD calculation and can still compare the results quantitatively. But I reduce the chance that HC and AD trajectories get mixed and that cells get intertwined in the trajectory of the wrong cell state. 

```{r slingshot healthy}
DefaultAssay(rojahn_kc) <- "RNA"

#remove possible previous iterations
if(exists("rojahn_kc$slingPseudotime_1")){
  rojahn_kc$slingPseudotime_1 <- NULL
}

if(exists("rojahn_kc$slingPseudotime_2")){
  rojahn_kc$slingPseudotime_2 <- NULL
}

rojahn_kc_hc <- subset(rojahn_kc, state == "HC")
rojahn_kc_hc_sce <- as.SingleCellExperiment(rojahn_kc_hc, assay = "RNA") #transform seurat to SingleCellExperiment object, needed as input for slingshot 

rojahn_kc_hc_sce <- slingshot(rojahn_kc_hc_sce, reducedDim = 'UMAP') #strong recommendation to specify a starting cluster, clusterLabels = "svm_annotation", start.clus = "basal III", end.clus = "stratum granulosum"

#plot pseudotime on UMAP
plot(reducedDims(rojahn_kc_hc_sce)$UMAP,
     col = colors_subtypes[as.factor(rojahn_kc_hc_sce$svm_annotation)], pch = 20, asp = 0.5) + lines(SlingshotDataSet(rojahn_kc_hc_sce))

#transfer trajectories back to the seurat object
rojahn_kc_hc$slingPseudotime <- rojahn_kc_hc_sce$slingPseudotime_1

#Supplemental figure 5a HC
FeaturePlot(rojahn_kc_hc,features = c("slingPseudotime")) + scale_color_gradient(low = "yellow", high ="red", na.value = "grey", limits = c(0, 24)) + theme(plot.title = element_blank()) + labs(color = "pseudo-\ntime")

#plot pseudotime 1
pseudotime_HC <- rojahn_kc_hc_sce@colData[ , c("svm_annotation", "slingPseudotime_1", "state", "sample")]
pseudotime_HC <- as.data.frame(pseudotime_HC)
pseudotime_HC <- pseudotime_HC[!is.na(pseudotime_HC$slingPseudotime_1), ]
pseudotime_HC$svm_annotation <- factor(pseudotime_HC$svm_annotation, levels = c("basal III", "basal I", "basal II", "basal IV", "stratum spinosum", "stratum granulosum"))

#annotation of cell types, Figure 4a HC
ggplot(pseudotime_HC,
       aes(x = svm_annotation, y = slingPseudotime_1, fill = svm_annotation, color = svm_annotation)) +
    geom_dotplot(binwidth = 0.3, stackdir = "centerwhole", binaxis = "y", stackratio = 0.1) + coord_flip() +
    scale_fill_manual(values = colors_subtypes) + scale_color_manual(values = colors_subtypes) + 
    scale_y_continuous(limits = c(0, 24)) +
    theme_classic() + xlab("HC") + labs(y = NULL) + theme(legend.position = "none") 

#annotation of sample
ggplot(pseudotime_HC,
       aes(x = svm_annotation, y = slingPseudotime_1, fill = sample, color = sample)) +
    geom_dotplot(binwidth = 0.3, stackdir = "centerwhole", binaxis = "y", stackratio = 0.1) + coord_flip() +
    scale_fill_manual(values = colors_samples) + scale_color_manual(values = colors_samples) + 
   scale_y_continuous(limits = c(0, 24)) +
    theme_classic() + xlab("HC") + labs(y = NULL)
```

```{r slingshot AD}
DefaultAssay(rojahn_kc) <- "RNA"

#remove possible previous iterations
if(exists("rojahn_kc$slingPseudotime_1")){
  rojahn_kc$slingPseudotime_1 <- NULL
}

if(exists("rojahn_kc$slingPseudotime_2")){
  rojahn_kc$slingPseudotime_2 <- NULL
}

rojahn_kc_ad <- subset(rojahn_kc, state == "AD")
rojahn_kc_ad_sce <- as.SingleCellExperiment(rojahn_kc_ad, assay = "RNA") #transform seurat to SingleCellExperiment object, needed as input for slingshot 

rojahn_kc_ad_sce <- slingshot(rojahn_kc_ad_sce, reducedDim = 'UMAP') #strong recommendation to specify a starting cluster

#plot pseudotime on UMAP
plot(reducedDims(rojahn_kc_ad_sce)$UMAP,
     col = colors_subtypes[as.factor(rojahn_kc_ad_sce$svm_annotation)], pch = 20, asp = 0.5) + lines(SlingshotDataSet(rojahn_kc_ad_sce))

#transfer trajectories back to the seurat object
rojahn_kc_ad$slingPseudotime <- rojahn_kc_ad_sce$slingPseudotime_1
rojahn_kc_ad$slingPseudotime_1_mod <- max(rojahn_kc_ad$slingPseudotime) - rojahn_kc_ad$slingPseudotime

#Supplemental figure 5a AD
FeaturePlot(rojahn_kc_ad,features = c("slingPseudotime_1_mod")) + scale_color_gradient(low = "yellow", high ="red", na.value = "grey", limits = c(0, 24)) + theme(plot.title = element_blank()) + labs(color = "pseudo-\ntime")

#plot pseudotime 1
pseudotime_AD <- rojahn_kc_ad_sce@colData[ , c("svm_annotation", "slingPseudotime_1", "state", "sample")]
pseudotime_AD <- as.data.frame(pseudotime_AD)
pseudotime_AD <- pseudotime_AD[!is.na(pseudotime_AD$slingPseudotime_1), ]
pseudotime_AD$slingPseudotime_1_mod <- max(pseudotime_AD$slingPseudotime_1) - pseudotime_AD$slingPseudotime_1
pseudotime_AD$svm_annotation <- factor(pseudotime_AD$svm_annotation, levels = c("basal III", "basal I", "basal II", "basal IV", "stratum spinosum", "stratum granulosum"))

#annotation of cell types, Figure 4a AD
ggplot(pseudotime_AD,
       aes(x = svm_annotation, y = slingPseudotime_1_mod, fill = svm_annotation, color = svm_annotation)) +
    geom_dotplot(binwidth = 0.3, stackdir = "centerwhole", binaxis = "y", stackratio = 0.1) + coord_flip() +
    scale_fill_manual(values = colors_subtypes) + scale_color_manual(values = colors_subtypes) + 
    scale_y_continuous(limits = c(0, 24)) +
    theme_classic() + xlab("AD") + labs(y = NULL) + theme(legend.position = "none") 

#annotation of sample
ggplot(pseudotime_AD,
       aes(x = svm_annotation, y = slingPseudotime_1_mod, fill = sample, color = sample)) +
    geom_dotplot(binwidth = 0.3, stackdir = "centerwhole", binaxis = "y", stackratio = 0.1) + coord_flip() +
    scale_fill_manual(values = colors_samples) + scale_color_manual(values = colors_samples) + 
    scale_y_continuous(limits = c(0, 24)) +
    theme_classic() + xlab("AD") + labs(y = NULL)
```

```{r plot pseudotime on UMAP}
DefaultAssay(rojahn_kc) <- "RNA"

rojahn_kc$cell <- colnames(rojahn_kc)
pseudotime_HC$cell <- rownames(pseudotime_HC)
pseudotime_AD$cell <- rownames(pseudotime_AD)

#transfer pseudotime into seurat object
rojahn_kc$slingPseudotime_HC <- NA
rojahn_kc$slingPseudotime_AD <- NA

Idents(rojahn_kc) <- rojahn_kc$cell
for (i in rojahn_kc$cell){
  rojahn_kc$slingPseudotime_HC[i] <- pseudotime_HC[i, "slingPseudotime_1"]
  rojahn_kc$slingPseudotime_AD[i] <- pseudotime_AD[i, "slingPseudotime_1"]
}

rojahn_kc$slingPseudotime_AD_mod <- max(rojahn_kc$slingPseudotime_AD, na.rm = T) - rojahn_kc$slingPseudotime_AD

#combine pseudotimes for visualization, take HC for HC cells or ADmod for AD cells
rojahn_kc$slingPseudotime_mod <- ifelse(!is.na(rojahn_kc$slingPseudotime_HC), rojahn_kc$slingPseudotime_HC, rojahn_kc$slingPseudotime_AD_mod)

DefaultAssay(rojahn_kc) <- "RNA"

FeaturePlot(rojahn_kc, features = c("slingPseudotime_mod")) + scale_color_gradient(low = "yellow", high ="red", na.value = "grey") + labs(color = "pseudo-\ntime") + theme(plot.title = element_blank()) 

#plot pseudotime on UMAP
rojahn_kc_sce <- as.SingleCellExperiment(rojahn_kc, assay = "RNA") 

plot(reducedDims(rojahn_kc_sce)$UMAP,
     col = colors_subtypes[as.factor(rojahn_kc_sce$svm_annotation)], pch = 20, asp = 0.5) + lines(SlingshotDataSet(rojahn_kc_hc_sce)) + 
lines(SlingshotDataSet(rojahn_kc_ad_sce))
```

## Comparing gene expression alongside the pseudotime trajectory (integrated UMAP)

We now want to look at gene expression during the differentiation trajectory. For this we first have to order HC and AD cells alongside a shared differentiation trajectory, therefore using the integrated UMAP. Afterwards we will visualize smoothed out non-integrated gene expression along the trajectory to compare HC and AD differentiation. 

### Create pseudotime trajectory (integrated UMAP)

```{r calculate pseudotime}
#remove possible previous iterations
if(exists("rojahn_kc$slingPseudotime")){
rojahn_kc$slingPseudotime_1 <- NULL
rojahn_kc$slingPseudotime_2 <- NULL
}

DefaultAssay(rojahn_kc) <- "integrated"
rojahn_kc <- seuratpipeline(rojahn_kc, nfeat = 5000, res = 0.3, dims = 11) 

#perform slingshot with the sce object
rojahn_kc_sce <- as.SingleCellExperiment(rojahn_kc, assay = "RNA") 
rojahn_kc_sce <- slingshot(rojahn_kc_sce, reducedDim = 'UMAP') 

#Plot pseudotime on UMAP
plot(reducedDims(rojahn_kc_sce)$UMAP, col = colors_subtypes[as.factor(rojahn_kc_sce$svm_annotation)], pch = 20,  asp = 0.5)
lines(SlingshotDataSet(rojahn_kc_sce))
# plotGeneCount(curve = SlingshotDataSet(rojahn_kc_sce), clusters = rojahn_kc_sce$svm_annotation) # other command for drawing trajectory on UMAP

#transfer trajectories back to the seurat object
rojahn_kc$slingPseudotime <- rojahn_kc_sce$slingPseudotime_1

#Supplemental figure 5b
FeaturePlot(rojahn_kc,features = c("slingPseudotime")) + scale_color_gradient(low = "yellow", high ="red", na.value = "grey") + theme(plot.title = element_blank()) + labs(color = "pseudo-\ntime")

#visualize trajectories in HC and AC
rojahn_kc_hc <- subset(rojahn_kc, state == "HC")
rojahn_kc_ad <- subset(rojahn_kc, state == "AD")

FeaturePlot(rojahn_kc_hc,features = c("slingPseudotime")) + ggtitle("HC pseudotime") + scale_color_gradient(low = "yellow", high ="red", na.value = "grey") 
FeaturePlot(rojahn_kc_ad,features = c("slingPseudotime")) + ggtitle("AD pseudotime") + scale_color_gradient(low = "yellow", high ="red", na.value = "grey")
```

```{r visualize pseudotime}
#plot pseudotime HC
pseudotime_int <- rojahn_kc_sce@colData[, c("svm_annotation", "slingPseudotime_1", "state")]
pseudotime_int$svm_annotation <- factor(pseudotime_int$svm_annotation, levels = c("basal III", "basal I", "basal II", "basal IV", "stratum spinosum", "stratum granulosum"))
pseudotime_HC_int <- as.data.frame(pseudotime_int[pseudotime_int$state == "HC", ])
pseudotime_HC_int <- pseudotime_HC_int[!is.na(pseudotime_HC_int$slingPseudotime_1), ]

ggplot(pseudotime_HC_int,
       aes(x = svm_annotation, y = slingPseudotime_1, fill = svm_annotation, color = svm_annotation)) +
    geom_dotplot(binwidth = 0.3, stackdir = "centerwhole", binaxis = "y", stackratio = 0.1) + coord_flip() +
    scale_fill_manual(values = colors_subtypes) + scale_color_manual(values = colors_subtypes) + 
    scale_y_continuous(limits = c(0, 26), breaks = c(0, 5, 10, 15, 20, 25)) +
    theme_classic() + xlab("HC") + labs(y = NULL) + theme(legend.position = "none") 


#plot pseudotime AD
pseudotime_AD_int <- as.data.frame(pseudotime_int[pseudotime_int$state == "AD", ])
pseudotime_AD_int <- pseudotime_AD_int[!is.na(pseudotime_AD_int$slingPseudotime_1), ]

ggplot(pseudotime_AD_int,
       aes(x = svm_annotation, y = slingPseudotime_1, fill = svm_annotation, color = svm_annotation)) +
    geom_dotplot(binwidth = 0.3, stackdir = "centerwhole", binaxis = "y", stackratio = 0.1) + coord_flip() +
    scale_fill_manual(values = colors_subtypes) + scale_color_manual(values = colors_subtypes) + 
    scale_y_continuous(limits = c(0, 26), breaks = c(0, 5, 10, 15, 20, 25)) +
    theme_classic() + xlab("AD") + labs(y = NULL) + theme(legend.position = "none") 
```

As expected, we do not see much difference between HC and AD anymore because they are integrated. The purpose here is not to compare trajectories but to use the shared differentiation trajectory to visualize non-integrated gene expression along the trajectory. 

### Calculate smoothened gene expression

Smooting out gene expression helps accentuate changes and without the heatmap looks more uniform, making the interpretation more difficult. 
We therefore use tradeseq to infer a smoothed gene expression along the pseudotime trajectory. This is done by fitting the data to a non-binomial generalized additive model. We first have to choose the right configuration, in this case by using evaluateK to help select a good number of knots to fit the model on. In general, it is best to only genes we will also visualize later, however that broke the function code here so we use the complete count table. 

```{r configure model}
icMat <- readRDS(paste0(save.dir, "icMat.rds"))

#figure out best number of k to fit the NB-GAM
# icMat <- evaluateK(counts = rojahn_kc_sce@assays@data$counts, sds = SlingshotDataSet(rojahn_kc_sce)) #suggested: genes = GeneExpVar

# saveRDS(icMat, paste0(save.dir, "icMat.rds"))
```

Now we use a negative binomial general additive model to infer the smoothed gene expression of the highly variable genes with the knots we have determined before. 

```{r calculate model}
rojahn_kc_sce_gam <- readRDS(paste0(save.dir, "rojahn_kc_sce_gam.rds"))

#calculating a negative binomial generalized additive model to interfere gene expression along the (slingshot calculated) trajectory
# rojahn_kc_sce_gam <- fitGAM(counts = as.matrix(rojahn_kc@assays$RNA$counts), sds = SlingshotDataSet(rojahn_kc_sce), nknots = 7, parallel = T) 

# statistical test to check if gene expression changes across trajectory
diffex_trajectory <- associationTest(rojahn_kc_sce_gam)
diffex_trajectory$gene <- rownames(diffex_trajectory)

# saveRDS(rojahn_kc_sce_gam, paste0(save.dir, "rojahn_kc_sce_gam.rds"))
```

### Select genes to visualize 

Now we draw a heatmap to visualize gene expression across the trajectory. We use only the highly variable + highly variable genes for this which are also differential expressed along the pseudotime (differentiation) trajectory.

```{r select highly expressed, variable and differential genes}
#select highly expressed genes
counts <- as.data.frame(rojahn_kc@assays$RNA$counts)

#calculate log10 transformed counts for all genes
gene_expr <- data.frame(gene_expr = rowSums(counts)) #calculates the genes total counts across all cells
gene_expr$gene <- rownames(gene_expr)
gene_expr$log10_counts <- log10(gene_expr$gene_expr + 1) #log10 transform that number
gene_expr$percent_cell <- apply(counts, 1, function(x){length(which(x != 0)) / ncol(counts)}) #obtain percentage of cells expressing the gene

hist(gene_expr$log10_counts)
hist(gene_expr$percent_cell)
summary(gene_expr$log10_counts)
summary(gene_expr$percent_cell)

#set count and % cells expressing thresholds and indicate them by assigned color
gene_expr$highly_expressed <- "no"
gene_expr[gene_expr$log10_counts > 2.8 & gene_expr$percent_cell > 0.15, ]$highly_expressed <- "yes"
color <- c("black", "red")
names(color) <- c("no","yes")

ggplot(data = gene_expr, aes(x = log10_counts, y = percent_cell, col = highly_expressed)) + geom_point() +
  scale_colour_manual(values = color) +
  geom_vline(xintercept = 2.8, col = "red") +
  geom_hline(yintercept = 0.15, col = "red") +
  theme_light()

GeneHighExpr <- rownames(gene_expr)[gene_expr$highly_expressed == "yes"]

#select highly variable genes
rojahn_kc <- FindVariableFeatures(rojahn_kc)
GeneHighVari <- rojahn_kc@assays$RNA@meta.data$var.features[!is.na(rojahn_kc@assays$RNA@meta.data$var.features)]

#select genes that are DE during trajectory
GeneDiffexTraject <- diffex_trajectory[diffex_trajectory$pvalue < 0.05 & !is.na(diffex_trajectory$pvalue), "gene"]
```

```{r intersect and add genes to visualize}
#intersect
pseudotime_genes <- intersect(intersect(GeneHighExpr, GeneHighVari), GeneDiffexTraject)

#add genes to annotate
genes_annotate <- c("KRT5", "KRT14", "KRT15", "MCM3", "MKI67", "TP63", "KRT1", "KRT10", "KRT6", "IVL", "ASS1") #select interesting genes to include by hand
heatmap_genes <- c(pseudotime_genes, genes_annotate)
```

### Draw the heatmap

First in healthy epidermis

```{r heatmap healthy}
rojahn_kc_hc <- subset(rojahn_kc, state == "HC")

heatmap_hc <- as.data.frame(rojahn_kc_hc@assays$RNA$counts)
heatmap_hc <- heatmap_hc[rownames(heatmap_hc) %in% heatmap_genes, ] # filter for desired genes

#order cells in pseudotime
pseudtime_cells <- as.data.frame(rojahn_kc_hc@meta.data[ , c("slingPseudotime", "svm_annotation")]) #obtain pseudotime
heatmap_hc <- heatmap_hc[ , rownames(pseudtime_cells)] #sanity check: remove cells not included in pseudotime
heatmap_hc <- heatmap_hc[ , order(pseudtime_cells$slingPseudotime)] # order by pseudotime

#change gene counts to z score and smooths gene expression
heatmap_hc <- t(apply(heatmap_hc, 1, function(x) scale(x))) #1 specifies rows, scale calculates the z score
heatmap_hc <- t(apply(heatmap_hc, 1, function(x){zoo::rollmean(x, 100, align = "center")})) #computes a rolling mean to the rows, this smooths out the outliers
# tried with 20 and 50 but it produces a more patchy heatmap and the expression pattern stay the same 
heatmap_hc <- na.omit(heatmap_hc)
heatmap_hc <- as.data.frame(heatmap_hc)

#annotate cell types
svm_annotation_hc <- pseudtime_cells[order(pseudtime_cells$slingPseudotime), ]$svm_annotation
svm_annotation_hc <- svm_annotation_hc[50 : (ncol(heatmap_hc) + 49)] #centered rolling mean -> first 50 cells don't have a direct output 

annot_celltype_hc <- HeatmapAnnotation(celltype = svm_annotation_hc, col = list(celltype = colors_subtypes))

# annotate genes of interest
annot_genes <- HeatmapAnnotation(
  GeneOfInterest = anno_mark(at = match(genes_annotate, rownames(heatmap_hc)),
                             labels = genes_annotate),
  which = "row"
  ) #show_annotation_name = T
 
heatmap_trajectory_hc <- readRDS(paste0(save.dir, "heatmap_trajectory_hc.rds"))
heatmap_trajectory_hc

# heatmap_trajectory_hc <- draw(Heatmap(heatmap_hc, top_annotation = annot_celltype_hc, # right_annotation = annot_genes, 
#                                       col = colorRamp2(c(min(heatmap_hc), 0, max(heatmap_hc)), c("blue", "lightyellow", "red")),
#                                       name = "z-score", 
#                                       km = 7, row_km_repeats = 1, row_dend_reorder = F, cluster_rows = T, clustering_method_rows = "single", cluster_row_slices = F, #clustering
#                                       row_title_rot = 0, show_row_names = F, show_column_names = F,
#                                       cluster_columns  = F))

# saveRDS(heatmap_trajectory_hc, paste0(save.dir, "heatmap_trajectory_hc.rds"))
```

And now in AD epidermis

```{r heatmap AD}
rojahn_kc_ad <- subset(rojahn_kc, state == "AD")

heatmap_ad <- as.data.frame(rojahn_kc_ad@assays$RNA$counts)
heatmap_ad <- heatmap_ad[rownames(heatmap_ad) %in% heatmap_genes, ] # filter for desired genes

#order cells in pseudotime
pseudtime_cells <- as.data.frame(rojahn_kc_ad@meta.data[ , c("slingPseudotime","svm_annotation")]) #obtain pseudotime
heatmap_ad <- heatmap_ad[ , rownames(pseudtime_cells)] #sanity check: remove cells not included in pseudotime
heatmap_ad <- heatmap_ad[ , order(pseudtime_cells$slingPseudotime)] # order by pseudotime

#change gene counts to z score and smooths gene expression
heatmap_ad <- t(apply(heatmap_ad, 1, function(x) scale(x))) #1 specifies rows, scale calculates the z score
heatmap_ad <- t(apply(heatmap_ad, 1, function(x){zoo::rollmean(x, 100, align = "center")})) #computes a rolling mean to the rows, this smooths out the outliers
# tried with 20 and 50 but it produces a more patchy heatmap and the expression pattern stay the same 
heatmap_ad <- na.omit(heatmap_ad)
heatmap_ad <- as.data.frame(heatmap_ad)

#annotate cell types
svm_annotation_ad <- pseudtime_cells[order(pseudtime_cells$slingPseudotime), ]$svm_annotation
svm_annotation_ad <- svm_annotation_ad[50 : (ncol(heatmap_ad) + 49)] #centered rolling mean -> first 50 cells don't have a direct output 

annot_celltype_ad <- HeatmapAnnotation(celltype = svm_annotation_ad, col = list(celltype = colors_subtypes))

# annotate genes of interest
annot_genes <- HeatmapAnnotation(
  GeneOfInterest = anno_mark(at = match(genes_annotate, rownames(heatmap_ad)),
                             labels = genes_annotate),
  which = "row"
  ) 

#draw and save heatmap
heatmap_trajectory_ad <- readRDS(paste0(save.dir, "heatmap_trajectory_ad.rds"))
heatmap_trajectory_ad

# heatmap_trajectory_ad <- draw(Heatmap(heatmap_ad, top_annotation = annot_celltype_ad, right_annotation = annot_genes,
#                                       col = colorRamp2(c(min(heatmap_ad), 0, max(heatmap_ad)), c("blue", "lightyellow", "red")),
#                                       name = "z-score",
#                                       km = 7, row_km_repeats = 1, row_dend_reorder = F, cluster_rows = T, clustering_method_rows = "single", cluster_row_slices = F, #clustering
#                                       row_title_rot = 0, show_row_names = F, show_column_names = F,
#                                       cluster_columns  = F))

# saveRDS(heatmap_trajectory_ad, paste0(save.dir, "heatmap_trajectory_ad.rds"))
```

And now both together in a side-by-side heatmap 

```{r heatmap HC and AD}
#match per gene ad and hc dataframe
shared_genes <- intersect(rownames(heatmap_hc), rownames(heatmap_ad)) #only use genes present in both heatmaps
heatmap_hcad <- cbind(heatmap_hc[shared_genes, ], heatmap_ad[shared_genes, ])

#annotate cell types
state_state <- c(rep("HC", length(svm_annotation_hc)), rep("AD", length(svm_annotation_ad)))
state_state <- factor(state_state, levels = c("HC", "AD"))

annot_celltype_hcad <- HeatmapAnnotation(
  "tissue state" = state_state,
  "cell type" = c(svm_annotation_hc, svm_annotation_ad),
  col = list("cell type" = colors_subtypes, 
             "tissue state" = colors_state)
  )
# saveRDS(annot_celltype_hcad, paste0(save.dir, "annot_celltype_hcad.rds"))

#annotate genes of interest
annot_genes <- HeatmapAnnotation(
  GeneOfInterest = anno_mark(at = match(genes_annotate, rownames(heatmap_hcad)),
                             labels = genes_annotate),
  which = "row"
  ) 

#draw and save the heatmap
heatmap_trajectory_hcad <- readRDS(paste0(save.dir, "heatmap_trajectory_hcad.rds"))
heatmap_trajectory_hcad

# heatmap_trajectory_hcad <- draw(Heatmap(heatmap_hcad, right_annotation = annot_genes, top_annotation = annot_celltype_hcad,
#                                       col = colorRamp2(c(min(heatmap_ad), 0, max(heatmap_ad)), c("blue", "lightyellow", "red")),
#                                       name = "z-score",
#                                       km = 7, row_km_repeats = 1, row_dend_reorder = F, cluster_rows = T, clustering_method_rows = "single", cluster_row_slices = F, show_row_dend = F, #clustering
#                                       row_title_rot = 0, show_row_names = F, show_column_names = F,
#                                       cluster_columns  = F))
# saveRDS(heatmap_trajectory_hcad, paste0(save.dir, "heatmap_trajectory_hcad.rds"))
```

While we see some clear changes in the expression of late differentiation genes in AD, the patterns of early differentiation genes that I saw in the AD heatmap are less visible here. I therefore decided to extract early, mid and late differentiation genes and visualize them in separate heatmaps to  increase the clustering resolution. 

```{r isolate clusters}
summary(row_order(heatmap_trajectory_hcad))
c1 <- heatmap_hcad[row_order(heatmap_trajectory_hcad)[[1]], ]
c2 <- heatmap_hcad[row_order(heatmap_trajectory_hcad)[[2]], ]
c3 <- heatmap_hcad[row_order(heatmap_trajectory_hcad)[[3]], ]
c4 <- heatmap_hcad[row_order(heatmap_trajectory_hcad)[[4]], ]
c5 <- heatmap_hcad[row_order(heatmap_trajectory_hcad)[[5]], ]
c6 <- heatmap_hcad[row_order(heatmap_trajectory_hcad)[[6]], ]
c7 <- heatmap_hcad[row_order(heatmap_trajectory_hcad)[[7]], ]

heatmap_hcad_early <- rbind(c3, c4, c6)
heatmap_hcad_late <- rbind(c1, c2, c5) 
heatmap_hcad_cc <- c7
```

Reorder early differentiation genes

```{r heatmap early differentiation}
#annotate genes of interest
annot_genes <- HeatmapAnnotation(
  GeneOfInterest = anno_mark(at = match(genes_annotate, rownames(heatmap_hcad_early)),
                             labels = genes_annotate),
  which = "row"
  ) 

#draw and save the heatmap
heatmap_trajectory_hcad_early <- readRDS(paste0(save.dir, "heatmap_trajectory_hcad_early.rds"))
heatmap_trajectory_hcad_early

# heatmap_trajectory_hcad_early <- draw(Heatmap(heatmap_hcad_early,
#                                                 right_annotation = annot_genes, top_annotation = annot_celltype_hcad,
#                                                 col = colorRamp2(c(min(heatmap_ad), 0, max(heatmap_ad)), c("blue", "lightyellow", "red")),
#                                                 km = 5, row_km_repeats = 1, cluster_row_slices = T, #clustering
#                                                 show_row_names = F, show_row_dend = F, row_title_rot = 0,
#                                                 cluster_columns = F, show_column_names = F))
# saveRDS(heatmap_trajectory_hcad_early, paste0(save.dir, "heatmap_trajectory_hcad_early.rds"))
        
HCearly_ADlate_late <- heatmap_hcad_early[row_order(heatmap_trajectory_hcad_early)[[1]], ]
HCearly_ADearly <- heatmap_hcad_early[row_order(heatmap_trajectory_hcad_early)[[3]], ]
HCearly_ADlate_1 <- heatmap_hcad_early[row_order(heatmap_trajectory_hcad_early)[[4]], ]
HCearly_ADlate_2 <- heatmap_hcad_early[row_order(heatmap_trajectory_hcad_early)[[5]], ]

HCearly_ADlate <- c(rownames(HCearly_ADlate_1), rownames(HCearly_ADlate_2))
HCearly_ADlate <- HCearly_ADlate[! HCearly_ADlate %in% c("HOPX", "S100A10", "IGFBP3", "KRT5", "KRT14")]
HCearly_ADlate <- c(HCearly_ADlate, c("S100A10", "IGFBP3", "KRT5", "KRT14"))

manual_order_early <- c(rownames(HCearly_ADearly), "SLC30A10", "ANKRD30B", HCearly_ADlate, "ADH1C", "ANGPTL7", rownames(HCearly_ADlate_late))
heatmap_manual_early <- heatmap_hcad[rownames(heatmap_hcad) %in% manual_order_early, ]
heatmap_manual_early <- heatmap_manual_early[manual_order_early, ]

heatmap_trajectory_hcad_reorder_early <- draw(Heatmap(heatmap_manual_early, name = c("gene ex-\npression"),
                                                right_annotation = annot_genes, top_annotation = annot_celltype_hcad,
                                                col = colorRamp2(c(min(heatmap_ad), 0, max(heatmap_ad)), c("blue", "lightyellow", "red")), na_col = "white",
                                                cluster_rows = F, #clustering # km = 8, 
                                                show_row_names = F, show_row_dend = F, row_title_rot = 0,
                                                cluster_columns = F, show_column_names = F))
```

Reorder late differentiation genes

```{r heatmap late differentiation}
#annotate genes of interest
annot_genes <- HeatmapAnnotation(
  GeneOfInterest = anno_mark(at = match(genes_annotate, rownames(heatmap_hcad_late)),
                             labels = genes_annotate),
  which = "row"
  ) 

#draw and save the heatmap
heatmap_trajectory_hcad_late <- readRDS(paste0(save.dir, "heatmap_trajectory_hcad_late.rds"))
heatmap_trajectory_hcad_late

# heatmap_trajectory_hcad_late <- draw(Heatmap(heatmap_hcad_late,
#                                                 right_annotation = annot_genes, top_annotation = annot_celltype_hcad,
#                                                 col = colorRamp2(c(min(heatmap_ad), 0, max(heatmap_ad)), c("blue", "lightyellow", "red")),
#                                                 km = 5, row_km_repeats = 1, cluster_row_slices = T, #clustering
#                                                 show_row_names = F, show_row_dend = F, row_title_rot = 0,
#                                                 cluster_columns = F, show_column_names = F))
# saveRDS(heatmap_trajectory_hcad_late, paste0(save.dir, "heatmap_trajectory_hcad_late.rds"))
        
HCweak_ADweaker <- heatmap_hcad_late[row_order(heatmap_trajectory_hcad_late)[[1]], ]
HCstrong_ADweak_2 <- heatmap_hcad_late[row_order(heatmap_trajectory_hcad_late)[[2]], ]
HCweak_ADstrong <- heatmap_hcad_late[row_order(heatmap_trajectory_hcad_late)[[3]], ]
HCstrong_ADweak_1 <- heatmap_hcad_late[row_order(heatmap_trajectory_hcad_late)[[5]], ]

manual_order_late <- c(rownames(HCstrong_ADweak_1), rownames(HCstrong_ADweak_2), "JAKMIP3", "LINC00605", rownames(HCweak_ADweaker), "PLPPR4", "TMEM262", rownames(HCweak_ADstrong))
heatmap_manual_late <- heatmap_hcad[rownames(heatmap_hcad) %in% manual_order_late, ]
heatmap_manual_late <- heatmap_manual_late[manual_order_late, ]

heatmap_trajectory_hcad_reorder_late <- draw(Heatmap(heatmap_manual_late, name = c("gene ex-\npression"),
                                                right_annotation = annot_genes, top_annotation = annot_celltype_hcad,
                                                col = colorRamp2(c(min(heatmap_ad), 0, max(heatmap_ad)), c("blue", "lightyellow", "red")), na_col = "white",
                                                cluster_rows = F, #clustering # km = 8, 
                                                show_row_names = F, show_row_dend = F, row_title_rot = 0,
                                                cluster_columns = F, show_column_names = F))
```
```{r heatmap combined}
heatmap_manual_complete <- readRDS(paste0(save.dir, "heatmap_manual_complete.rds"))

# manual_order_complete <- c(manual_order_early, "SYT6", "PRSS2", rownames(heatmap_hcad_cc), "TTPA", "LINC00605", manual_order_late)
# 
# heatmap_manual_complete <- heatmap_hcad[rownames(heatmap_hcad) %in% manual_order_complete, ]
# heatmap_manual_complete <- heatmap_manual_complete[manual_order_complete, ]
# saveRDS(heatmap_manual_complete, paste0(save.dir, "heatmap_manual_complete.rds"))

#annotate cell types
annot_celltype_hcad <- readRDS(paste0(save.dir, "annot_celltype_hcad.rds"))

#annotate genes
genes_annotate <- c("KRT15", "PDGFA", 
                    "ASS1", "COL17A1", "ITGA3", #maybe take out ITGA3
                    "BCL10", "NFKBIZ", "IRS2", 
                    "MCM3", "MKI67",
                    "KRT1", "KRT10", "CLDN4",
                    "IVL", "OVOL1", "NEAT1",
                    "KRT6A", "KRT6B", "KRT6C") 
 
annot_genes <- HeatmapAnnotation(
  GeneOfInterest = anno_mark(at = match(genes_annotate, rownames(heatmap_manual_complete)),
                             labels = genes_annotate),
  which = "row"
  ) 

#draw heatmap of manual order, Figure 4b
heatmap_trajectory_hcad_reorder <- draw(Heatmap(heatmap_manual_complete, name = c("gene ex-\npression"),
                                                right_annotation = annot_genes, top_annotation = annot_celltype_hcad,
                                                col = colorRamp2(c(min(na.omit(heatmap_manual_complete)), 0, max(na.omit(heatmap_manual_complete))),
                                                                 c("blue", "lightyellow", "red")), 
                                                na_col = "white",
                                                cluster_rows = F, show_row_names = F, show_row_dend = F, row_title_rot = 0,
                                                cluster_columns = F, show_column_names = F))
# saveRDS(heatmap_trajectory_hcad_reorder, paste0(save.dir, "heatmap_trajectory_hcad_reorder.rds"))
```

### Perform gene over-enrichment analysis of the clustered genes

```{r prepare input}
background_symbol <- names(rowSums(rojahn_kc@assays$RNA$counts) > 3)

#annotate all expressed background genes
background <-  mapIds(
  org.Hs.eg.db,
  keys = background_symbol,
  column = "ENTREZID",
  keytype = "SYMBOL",
  multiVals = "first"
)

#annotate DE genes in the differential expression data frame
Diff_HCearly_ADearly <- readRDS(paste0(save.dir, "Diff_HCearly_ADearly.rds"))
Diff_HCearly_ADlate <- readRDS(paste0(save.dir, "Diff_HCearly_ADlate.rds"))
Diff_HCearly_ADlate_late <- readRDS(paste0(save.dir, "Diff_HCearly_ADlate_late.rds"))
Diff_HC_AD_CC <- readRDS(paste0(save.dir, "Diff_HC_AD_CC.rds"))
Diff_HCstrong_ADweak <- readRDS(paste0(save.dir, "Diff_HCstrong_ADweak.rds"))
Diff_HCweak_ADweaker <- readRDS(paste0(save.dir, "Diff_HCweak_ADweaker.rds"))
Diff_HCweak_ADstrong <- readRDS(paste0(save.dir, "Diff_HCweak_ADstrong.rds"))

# Diff_HCearly_ADearly <- bitr(rownames(HCearly_ADearly), fromType = "SYMBOL", toType = c("ENTREZID"), OrgDb = org.Hs.eg.db)
# Diff_HCearly_ADlate <- bitr(HCearly_ADlate, fromType = "SYMBOL", toType = c("ENTREZID"), OrgDb = org.Hs.eg.db)
# Diff_HCearly_ADlate_late <- bitr(rownames(HCearly_ADlate_late), fromType = "SYMBOL", toType = c("ENTREZID"), OrgDb = org.Hs.eg.db)
# Diff_HC_AD_CC <- bitr(rownames(heatmap_hcad_cc), fromType = "SYMBOL", toType = c("ENTREZID"), OrgDb = org.Hs.eg.db)
# Diff_HCstrong_ADweak <- bitr(c(rownames(HCstrong_ADweak_1), rownames(HCstrong_ADweak_2)), fromType = "SYMBOL", toType = c("ENTREZID"), OrgDb = org.Hs.eg.db)
# Diff_HCweak_ADweaker <- bitr(rownames(HCweak_ADweaker), fromType = "SYMBOL", toType = c("ENTREZID"), OrgDb = org.Hs.eg.db)
# Diff_HCweak_ADstrong <- bitr(rownames(HCweak_ADstrong), fromType = "SYMBOL", toType = c("ENTREZID"), OrgDb = org.Hs.eg.db)
# 
# saveRDS(Diff_HCearly_ADearly, paste0(save.dir, "Diff_HCearly_ADearly.rds"))
# saveRDS(Diff_HCearly_ADlate, paste0(save.dir, "Diff_HCearly_ADlate.rds"))
# saveRDS(Diff_HCearly_ADlate_late, paste0(save.dir, "Diff_HCearly_ADlate_late.rds"))
# saveRDS(Diff_HC_AD_CC, paste0(save.dir, "Diff_HC_AD_CC.rds"))
# saveRDS(Diff_HCstrong_ADweak, paste0(save.dir, "Diff_HCstrong_ADweak.rds"))
# saveRDS(Diff_HCweak_ADweaker, paste0(save.dir, "Diff_HCweak_ADweaker.rds"))
# saveRDS(Diff_HCweak_ADstrong, paste0(save.dir, "Diff_HCweak_ADstrong.rds"))

# save in excel 
# list <- list("Cluster 1" = Diff_HCearly_ADearly,
#              "Cluster 2" = Diff_HCearly_ADlate,
#              "Cluster 3" = Diff_HCearly_ADlate_late,
#              "Cluster 4" = Diff_HC_AD_CC,
#              "Cluster 5" = Diff_HCstrong_ADweak,
#              "Cluster 6" = Diff_HCweak_ADweaker,
#              "Cluster 7" = Diff_HCweak_ADstrong)
# write.xlsx(list, file = paste0(save.dir, "Trajectory heatmap HC AD.xlsx"))
```

```{r GO over enrichment heatmap HC-AD}
#read
HCearly_ADearly_GO <- readRDS(paste0(save.dir, "HCearly_ADearly_GO.rds"))
HCearly_ADlate_GO <- readRDS(paste0(save.dir, "HCearly_ADlate_GO.rds"))
HCearly_ADlate_late_GO <- readRDS(paste0(save.dir, "HCearly_ADlate_late_GO.rds"))
HC_AD_CC_GO <- readRDS(paste0(save.dir, "HC_AD_CC_GO.rds"))
HCstrong_ADweak_GO <- readRDS(paste0(save.dir, "HCstrong_ADweak_GO.rds"))
HCweak_ADweaker_GO <- readRDS(paste0(save.dir, "HCweak_ADweaker_GO.rds"))
HCweak_ADstrong_GO <- readRDS(paste0(save.dir, "HCweak_ADstrong_GO.rds"))

#calculate
background_symbol <- names(rowSums(rojahn_kc@assays$RNA$counts) > 3)

HCearly_ADearly_GO <- enrichGO(gene = Diff_HCearly_ADearly$ENTREZID, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "ALL", pvalueCutoff = 0.1, readable = T) 
HCearly_ADearly_GOdf <- as.data.frame(HCearly_ADearly_GO)
View(HCearly_ADearly_GOdf)

HCearly_ADlate_GO <- enrichGO(gene = Diff_HCearly_ADlate$ENTREZID, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "ALL", pvalueCutoff = 0.1, readable = T)
HCearly_ADlate_GOdf <- as.data.frame(HCearly_ADlate_GO)
View(HCearly_ADlate_GOdf)

HCearly_ADlate_late_GO <- enrichGO(gene = Diff_HCearly_ADlate_late$ENTREZID, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "ALL", pvalueCutoff = 0.1, readable = T)
HCearly_ADlate_late_GOdf <- as.data.frame(HCearly_ADlate_late_GO)
View(HCearly_ADlate_late_GOdf)

HC_AD_CC_GO <- enrichGO(gene = Diff_HC_AD_CC$ENTREZID, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "ALL", pvalueCutoff = 0.1, readable = T)
HC_AD_CC_GOdf <- as.data.frame(HC_AD_CC_GO)
View(HC_AD_CC_GOdf)

HCstrong_ADweak_GO <- enrichGO(gene = Diff_HCstrong_ADweak$ENTREZID, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "ALL", pvalueCutoff = 0.1, readable = T)
HCstrong_ADweak_GOdf <- as.data.frame(HCstrong_ADweak_GO)
View(HCstrong_ADweak_GOdf)

HCweak_ADweaker_GO <- enrichGO(gene = Diff_HCweak_ADweaker$ENTREZID, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "ALL", pvalueCutoff = 0.1, readable = T)
HCweak_ADweaker_GOdf <- as.data.frame(HCweak_ADweaker_GO)
View(HCweak_ADweaker_GOdf)

HCweak_ADstrong_GO <- enrichGO(gene = Diff_HCweak_ADstrong$ENTREZID, OrgDb = org.Hs.eg.db, universe = background, keyType = 'ENTREZID', ont = "ALL", pvalueCutoff = 0.1, readable = T)
HCweak_ADstrong_GOdf <- as.data.frame(HCweak_ADstrong_GO)
View(HCweak_ADstrong_GOdf)

#save
# list <- list("GO cluster 1" = HCearly_ADearly_GO,
#              "GO cluster 2" = HCearly_ADlate_GO,
#              "GO cluster 3" = HCearly_ADlate_late_GO,
#              "GO cluster 4" = HC_AD_CC_GO,
#              "GO cluster 5" = HCstrong_ADweak_GO,
#              "GO cluster 6" = HCweak_ADweaker_GO,
#              "GO cluster 7" = HCweak_ADstrong_GO)
# write.xlsx(list, file = paste0(save.dir, "GO trajectory heatmap HC AD.xlsx"))

# saveRDS(HCearly_ADearly_GO, paste0(save.dir, "HCearly_ADearly_GO.rds"))
# saveRDS(HCearly_ADlate_GO, paste0(save.dir, "HCearly_ADlate_GO.rds"))
# saveRDS(HCearly_ADlate_late_GO, paste0(save.dir, "HCearly_ADlate_late_GO.rds"))
# saveRDS(HC_AD_CC_GO, paste0(save.dir, "HC_AD_CC_GO.rds"))
# saveRDS(HCstrong_ADweak_GO, paste0(save.dir, "HCstrong_ADweak_GO.rds"))
# saveRDS(HCweak_ADweaker_GO, paste0(save.dir, "HCweak_ADweaker_GO.rds"))
# saveRDS(HCweak_ADstrong_GO, paste0(save.dir, "HCweak_ADstrong_GO.rds"))

HCAD_trajectory_GO <- HCearly_ADearly_GO[c(3, 4, 24)]
HCAD_trajectory_GO <- rbind(HCAD_trajectory_GO, HCearly_ADlate_GO[c(2, 5)])
HCAD_trajectory_GO <- rbind(HCAD_trajectory_GO, HCearly_ADlate_late_GO[c(4, 14, 35, 90)])
HCAD_trajectory_GO <- rbind(HCAD_trajectory_GO, HC_AD_CC_GO[c(3, 6)])
HCAD_trajectory_GO <- rbind(HCAD_trajectory_GO, HCstrong_ADweak_GO[c(2, 8)])
HCAD_trajectory_GO <- rbind(HCAD_trajectory_GO, HC_AD_CC_GO[c(3)])
HCAD_trajectory_GO <- rbind(HCAD_trajectory_GO, HCweak_ADstrong_GO[c(1)])

HCAD_trajectory_GO$celltype <- c(rep("cluster 1", 3),
                                 rep("cluster 2", 2),
                                 rep("cluster 3", 4),
                                 rep("cluster 4", 2),
                                 rep("cluster 5", 2),
                                 rep("cluster 6", 1),
                                 rep("cluster 7", 1))
HCAD_trajectory_GO$celltype <- factor(HCAD_trajectory_GO$celltype, levels = c("cluster 1", "cluster 2", "cluster 3", "cluster 4", "cluster 5", "cluster 6", "cluster 7"))

#Figure 4c
ggplot(HCAD_trajectory_GO, aes(x = Count, y = Description)) +
        geom_col(aes(fill = p.adjust)) +
        facet_grid("celltype", scale = "free_y", space = "free_y") + #switch = "y",
        scale_fill_gradient(high = "#FAFAD2", low = "darkgoldenrod2", guide = guide_colourbar(reverse = TRUE)) +
        theme_bw(base_size = 13) + scale_y_discrete(labels = scales::label_wrap(30)) +
        ylab(NULL) + xlab("gene count") + labs(fill = "adjusted \np value") +
        theme(axis.text.y = element_text(size = 13), strip.text.y = element_text(size = 13))
```

```{r KEGG over enrichment heatmap HC-AD}
HCearly_ADearly_kegg <- enrichKEGG(gene = Diff_HCearly_ADearly$ENTREZID, universe = background, pvalueCutoff  = 0.05)
HCearly_ADearly_kegg <- setReadable(HCearly_ADearly_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
View(as.data.frame(HCearly_ADearly_kegg@result))

HCearly_ADlate_kegg <- enrichKEGG(gene = Diff_HCearly_ADlate$ENTREZID, universe = background, pvalueCutoff  = 0.05)
HCearly_ADlate_kegg <- setReadable(HCearly_ADlate_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
View(as.data.frame(HCearly_ADlate_kegg@result))

HCearly_ADlate_late_kegg <- enrichKEGG(gene = Diff_HCearly_ADlate_late$ENTREZID, universe = background, pvalueCutoff  = 0.05)
HCearly_ADlate_late_kegg <- setReadable(HCearly_ADlate_late_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
View(as.data.frame(HCearly_ADlate_late_kegg@result))

HC_AD_CC_kegg <- enrichKEGG(gene = Diff_HC_AD_CC$ENTREZID, universe = background, pvalueCutoff  = 0.05)
HC_AD_CC_kegg <- setReadable(HC_AD_CC_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
View(as.data.frame(HC_AD_CC_kegg@result))

HCstrong_ADweak_kegg <- enrichKEGG(gene = Diff_HCstrong_ADweak$ENTREZID, universe = background, pvalueCutoff  = 0.05)
HCstrong_ADweak_kegg <- setReadable(HCstrong_ADweak_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
View(as.data.frame(HCstrong_ADweak_kegg@result))

HCweak_ADweaker_kegg <- enrichKEGG(gene = Diff_HCweak_ADweaker$ENTREZID, universe = background, pvalueCutoff  = 0.05)
HCweak_ADweaker_kegg <- setReadable(HCweak_ADweaker_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
View(as.data.frame(HCweak_ADweaker_kegg@result))

HCweak_ADstrong_kegg <- enrichKEGG(gene = Diff_HCweak_ADstrong$ENTREZID, universe = background, pvalueCutoff  = 0.05)
HCweak_ADstrong_kegg <- setReadable(HCweak_ADstrong_kegg, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
View(as.data.frame(HCweak_ADstrong_kegg@result))

#save
# list <- list("KEGG cluster 1" = HCearly_ADearly_kegg,
#              "KEGG cluster 2" = HCearly_ADlate_kegg,
#              "KEGG cluster 3" = HCearly_ADlate_late_kegg,
#              "KEGG cluster 4" = HC_AD_CC_kegg,
#              "KEGG cluster 5" = HCstrong_ADweak_kegg,
#              "KEGG cluster 6" = HCweak_ADweaker_kegg,
#              "KEGG cluster 7" = HCweak_ADstrong_kegg)
# write.xlsx(list, file = paste0(save.dir, "KEGG trajectory heatmap HC AD.xlsx"))
```